

<!DOCTYPE html>
<html class="writer-html5" lang="en">
<head>
  <meta charset="utf-8" /><meta name="viewport" content="width=device-width, initial-scale=1" />

  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>Basic use 1: Mouse to human &mdash; TransBrain 0.1 documentation</title>
      <link rel="stylesheet" type="text/css" href="../../../_static/pygments.css?v=03e43079" />
      <link rel="stylesheet" type="text/css" href="../../../_static/css/theme.css?v=e59714d7" />
      <link rel="stylesheet" type="text/css" href="../../../_static/nbsphinx-code-cells.css?v=2aa19091" />

  
      <script src="../../../_static/jquery.js?v=5d32c60e"></script>
      <script src="../../../_static/_sphinx_javascript_frameworks_compat.js?v=2cd50e6c"></script>
      <script data-url_root="../../../" id="documentation_options" src="../../../_static/documentation_options.js?v=e031e9a9"></script>
      <script src="../../../_static/doctools.js?v=888ff710"></script>
      <script src="../../../_static/sphinx_highlight.js?v=4825356b"></script>
      <script crossorigin="anonymous" integrity="sha256-Ae2Vz/4ePdIu6ZyI/5ZGsYnb+m0JlOmKPjt6XZ9JJkA=" src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.4/require.min.js"></script>
      <script>window.MathJax = {"tex": {"inlineMath": [["$", "$"], ["\\(", "\\)"]], "processEscapes": true}, "options": {"ignoreHtmlClass": "tex2jax_ignore|mathjax_ignore|document", "processHtmlClass": "tex2jax_process|mathjax_process|math|output_area"}}</script>
      <script defer="defer" src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-mml-chtml.js"></script>
    <script src="../../../_static/js/theme.js"></script>
    <link rel="index" title="Index" href="../../../genindex.html" />
    <link rel="search" title="Search" href="../../../search.html" />
    <link rel="next" title="Basic use 2: Human to mouse" href="basic_use_human_to_mouse.html" />
    <link rel="prev" title="Online Mapping" href="../../../online_mapping.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >

          
          
          <a href="../../../index.html" class="icon icon-home">
            TransBrain
              <img src="../../../_static/logo_final.png" class="logo" alt="Logo"/>
          </a>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../../../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" aria-label="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <p class="caption" role="heading"><span class="caption-text">Setup &amp; Usage</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../../../installation.html">Installation</a><ul>
<li class="toctree-l2"><a class="reference internal" href="../../../installation.html#installation-options">Installation Options</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../installation.html#python-dependencies">Python Dependencies</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../installation.html#unit-test">Unit Test</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="../../../user.html">User Guide</a><ul>
<li class="toctree-l2"><a class="reference internal" href="../../../user.html#id2">Online Mapping</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../user.html#id3">General Use</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../user.html#id4">Advanced Use</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="../../../atlas.html">Supported Atlases</a><ul>
<li class="toctree-l2"><a class="reference internal" href="../../../atlas.html#human-brain-atlases">Human Brain Atlases</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../atlas.html#mouse-brain-atlas">Mouse Brain Atlas</a></li>
</ul>
</li>
</ul>
<p class="caption" role="heading"><span class="caption-text">Tutorials</span></p>
<ul class="current">
<li class="toctree-l1"><a class="reference internal" href="../../../methods/method_glossory.html">Method Glossory</a><ul>
<li class="toctree-l2"><a class="reference internal" href="../../../methods/method_glossory.html#datasets">Datasets</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../methods/method_glossory.html#methods-or-concepts">Methods or Concepts</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="../../../online_mapping.html">Online Mapping</a></li>
<li class="toctree-l1 current"><a class="current reference internal" href="#">Basic use 1: Mouse to human</a><ul>
<li class="toctree-l2"><a class="reference internal" href="#Step-1:-Prepare-input-data">Step 1: Prepare input data</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#Visualizing-mouse-phenotype-data">Visualizing mouse phenotype data</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#Step-2:-Initialization">Step 2: Initialization</a></li>
<li class="toctree-l2"><a class="reference internal" href="#Step-3:-Mapping">Step 3: Mapping</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="basic_use_human_to_mouse.html">Basic use 2: Human to mouse</a><ul>
<li class="toctree-l2"><a class="reference internal" href="basic_use_human_to_mouse.html#Step-1:-Prepare-input-data">Step 1: Prepare input data</a><ul>
<li class="toctree-l3"><a class="reference internal" href="basic_use_human_to_mouse.html#Visualizing-human-phenotype-data">Visualizing human phenotype data</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="basic_use_human_to_mouse.html#Step-2:-Initialization">Step 2: Initialization</a></li>
<li class="toctree-l2"><a class="reference internal" href="basic_use_human_to_mouse.html#Step-3:-Mapping">Step 3: Mapping</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="../translation/gradient_spectrum/fMRI_gradients.html">Case 1: Analyzing the conservation of resting-fMRI gradients</a><ul>
<li class="toctree-l2"><a class="reference internal" href="../translation/gradient_spectrum/fMRI_gradients.html#Introduction">Introduction</a></li>
<li class="toctree-l2"><a class="reference internal" href="../translation/gradient_spectrum/fMRI_gradients.html#Mapping-mouse-gradient-to-human">Mapping mouse gradient to human</a></li>
<li class="toctree-l2"><a class="reference internal" href="../translation/gradient_spectrum/fMRI_gradients.html#Compare-with-human-Gradient">Compare with human Gradient</a></li>
<li class="toctree-l2"><a class="reference internal" href="../translation/gradient_spectrum/fMRI_gradients.html#Cross-species-correlation">Cross-species correlation</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="../translation/optogenetic_annotation/optogenetic_circuits.html">Case 2: Annotating optogenetic circuits</a><ul>
<li class="toctree-l2"><a class="reference internal" href="../translation/optogenetic_annotation/optogenetic_circuits.html#Introduction">Introduction</a><ul>
<li class="toctree-l3"><a class="reference internal" href="../translation/optogenetic_annotation/optogenetic_circuits.html#Data">Data</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="../translation/optogenetic_annotation/optogenetic_circuits.html#Translating-mouse-stimulated-pattern-to-human">Translating mouse stimulated pattern to human</a></li>
<li class="toctree-l2"><a class="reference internal" href="../translation/optogenetic_annotation/optogenetic_circuits.html#Annotating-using-NeuroSynth.">Annotating using NeuroSynth.</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="../translation/autism_mutation/gene_mutations.html">Case 3: Linking gene mutations to imaging phenotype deviations</a><ul>
<li class="toctree-l2"><a class="reference internal" href="../translation/autism_mutation/gene_mutations.html#Introduction">Introduction</a><ul>
<li class="toctree-l3"><a class="reference internal" href="../translation/autism_mutation/gene_mutations.html#Data">Data</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="../translation/autism_mutation/gene_mutations.html#Translating-mouse-mutation-patterns-to-human">Translating mouse mutation patterns to human</a></li>
<li class="toctree-l2"><a class="reference internal" href="../translation/autism_mutation/gene_mutations.html#Calculate-individual-gene-risk-score">Calculate individual gene risk score</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="../../../building.html">Advanced Use</a><ul>
<li class="toctree-l2"><a class="reference internal" href="../../../pipeline/run_pipeline/integration.html">Transcriptional Data Integration</a><ul>
<li class="toctree-l3"><a class="reference internal" href="../../../pipeline/run_pipeline/integration.html#Introduction">Introduction</a><ul>
<li class="toctree-l4"><a class="reference internal" href="../../../pipeline/run_pipeline/integration.html#AHBA-data">AHBA data</a></li>
<li class="toctree-l4"><a class="reference internal" href="../../../pipeline/run_pipeline/integration.html#Human-single-nucleus-data">Human single-nucleus data</a></li>
<li class="toctree-l4"><a class="reference internal" href="../../../pipeline/run_pipeline/integration.html#Mouse-spatial-transcriptomic-data">Mouse spatial transcriptomic data</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="../../../pipeline/run_pipeline/integration.html#Selection-of-cortical-stable-genes">Selection of cortical stable genes</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../../pipeline/run_pipeline/integration.html#Selection-of-subcortical-stable-genes">Selection of subcortical stable genes</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../../pipeline/run_pipeline/integration.html#Selection-of-all-stable-genes">Selection of all stable genes</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../../pipeline/run_pipeline/integration.html#Thresholds-of-cortex">Thresholds of cortex</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../../pipeline/run_pipeline/integration.html#Thresholds-of-subcortex">Thresholds of subcortex</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../../pipeline/run_pipeline/integration.html#Thresholds-of-all">Thresholds of all</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="../../../pipeline/run_pipeline/dnn.html">DNN Model Training</a><ul>
<li class="toctree-l3"><a class="reference internal" href="../../../pipeline/run_pipeline/dnn.html#Introduction">Introduction</a><ul>
<li class="toctree-l4"><a class="reference internal" href="../../../pipeline/run_pipeline/dnn.html#Model-architecture">Model architecture</a></li>
<li class="toctree-l4"><a class="reference internal" href="../../../pipeline/run_pipeline/dnn.html#Region-specific-embeddings">Region-specific embeddings</a></li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="../../../pipeline/run_pipeline/graph_embedding.html">Graph-based Random Walk</a><ul>
<li class="toctree-l3"><a class="reference internal" href="../../../pipeline/run_pipeline/graph_embedding.html#Introduction">Introduction</a><ul>
<li class="toctree-l4"><a class="reference internal" href="../../../pipeline/run_pipeline/graph_embedding.html#Human-diffusion-based-tractography">Human diffusion-based tractography</a></li>
<li class="toctree-l4"><a class="reference internal" href="../../../pipeline/run_pipeline/graph_embedding.html#Mouse-Tracer">Mouse Tracer</a></li>
<li class="toctree-l4"><a class="reference internal" href="../../../pipeline/run_pipeline/graph_embedding.html#Embedding-space-generated-from-Human-Mouse-graph">Embedding space generated from Human-Mouse graph</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="../../../pipeline/run_pipeline/graph_embedding.html#Define-human-and-mouse-ROIs">Define human and mouse ROIs</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../../pipeline/run_pipeline/graph_embedding.html#Load-anatomical-hierachy-constraint-information">Load anatomical hierachy constraint information</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="../../../methods/mapping.html">Bidirectional Mapping</a></li>
</ul>
</li>
</ul>
<p class="caption" role="heading"><span class="caption-text">API</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../../../api/external_api.html">API Reference</a><ul>
<li class="toctree-l2"><a class="reference internal" href="../../../api/fetch_api.html">Fetching Atlases</a><ul>
<li class="toctree-l3"><a class="reference internal" href="../../../api/fetch_api.html#transbrain.atlas.fetch_human_atlas"><code class="docutils literal notranslate"><span class="pre">fetch_human_atlas()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="../../../api/fetch_api.html#transbrain.atlas.fetch_mouse_atlas"><code class="docutils literal notranslate"><span class="pre">fetch_mouse_atlas()</span></code></a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="../../../api/base_api.html">Base Functions</a><ul>
<li class="toctree-l3"><a class="reference internal" href="../../../api/base_api.html#transbrain.base.get_region_phenotypes"><code class="docutils literal notranslate"><span class="pre">get_region_phenotypes()</span></code></a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="../../../api/mapping_api.html">Mapping Functions</a><ul>
<li class="toctree-l3"><a class="reference internal" href="../../../api/mapping_api.html#transbrain.trans.SpeciesTrans"><code class="docutils literal notranslate"><span class="pre">SpeciesTrans</span></code></a><ul>
<li class="toctree-l4"><a class="reference internal" href="../../../api/mapping_api.html#transbrain.trans.SpeciesTrans.atlas_type"><code class="docutils literal notranslate"><span class="pre">SpeciesTrans.atlas_type</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="../../../api/mapping_api.html#transbrain.trans.SpeciesTrans.regions"><code class="docutils literal notranslate"><span class="pre">SpeciesTrans.regions</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="../../../api/mapping_api.html#transbrain.trans.SpeciesTrans.embeddings"><code class="docutils literal notranslate"><span class="pre">SpeciesTrans.embeddings</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="../../../api/mapping_api.html#transbrain.trans.SpeciesTrans._load_embeddings"><code class="docutils literal notranslate"><span class="pre">SpeciesTrans._load_embeddings()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="../../../api/mapping_api.html#transbrain.trans.SpeciesTrans._dual_mapping"><code class="docutils literal notranslate"><span class="pre">SpeciesTrans._dual_mapping()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="../../../api/mapping_api.html#transbrain.trans.SpeciesTrans._translate"><code class="docutils literal notranslate"><span class="pre">SpeciesTrans._translate()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="../../../api/mapping_api.html#transbrain.trans.SpeciesTrans.mouse_to_human"><code class="docutils literal notranslate"><span class="pre">SpeciesTrans.mouse_to_human()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="../../../api/mapping_api.html#transbrain.trans.SpeciesTrans.human_to_mouse"><code class="docutils literal notranslate"><span class="pre">SpeciesTrans.human_to_mouse()</span></code></a></li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="../../../api/vis_api.html">Visualization Functions</a><ul>
<li class="toctree-l3"><a class="reference internal" href="../../../api/vis_api.html#transbrain.vis.map_phenotype_to_nifti"><code class="docutils literal notranslate"><span class="pre">map_phenotype_to_nifti()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="../../../api/vis_api.html#transbrain.vis.plot_mouse_phenotype"><code class="docutils literal notranslate"><span class="pre">plot_mouse_phenotype()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="../../../api/vis_api.html#transbrain.vis.plot_human_phenotype"><code class="docutils literal notranslate"><span class="pre">plot_human_phenotype()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="../../../api/vis_api.html#transbrain.surf.plot_surface_phenotype"><code class="docutils literal notranslate"><span class="pre">plot_surface_phenotype()</span></code></a></li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="../../../api/utils.html">Development Functions</a><ul>
<li class="toctree-l2"><a class="reference internal" href="../../../api/data_process_api.html">Transcriptional Data Process</a><ul>
<li class="toctree-l3"><a class="reference internal" href="../../../api/data_process_api.html#pipeline.code.dpreprocess.sn_.Preprocess_Sn"><code class="docutils literal notranslate"><span class="pre">Preprocess_Sn</span></code></a><ul>
<li class="toctree-l4"><a class="reference internal" href="../../../api/data_process_api.html#pipeline.code.dpreprocess.sn_.Preprocess_Sn.dataset_qc"><code class="docutils literal notranslate"><span class="pre">Preprocess_Sn.dataset_qc()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="../../../api/data_process_api.html#pipeline.code.dpreprocess.sn_.Preprocess_Sn.dataset_normalized"><code class="docutils literal notranslate"><span class="pre">Preprocess_Sn.dataset_normalized()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="../../../api/data_process_api.html#pipeline.code.dpreprocess.sn_.Preprocess_Sn.dataset_common_genes"><code class="docutils literal notranslate"><span class="pre">Preprocess_Sn.dataset_common_genes()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="../../../api/data_process_api.html#pipeline.code.dpreprocess.sn_.Preprocess_Sn.dataset_batch_correct"><code class="docutils literal notranslate"><span class="pre">Preprocess_Sn.dataset_batch_correct()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="../../../api/data_process_api.html#pipeline.code.dpreprocess.sn_.Preprocess_Sn.dataset_smooth"><code class="docutils literal notranslate"><span class="pre">Preprocess_Sn.dataset_smooth()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="../../../api/data_process_api.html#pipeline.code.dpreprocess.sn_.Preprocess_Sn.dataset_scaled"><code class="docutils literal notranslate"><span class="pre">Preprocess_Sn.dataset_scaled()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="../../../api/data_process_api.html#pipeline.code.dpreprocess.sn_.Preprocess_Sn.preprocess_pipeline"><code class="docutils literal notranslate"><span class="pre">Preprocess_Sn.preprocess_pipeline()</span></code></a></li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="../../../api/data_integration_api.html">Data Integration</a><ul>
<li class="toctree-l3"><a class="reference internal" href="../../../api/data_integration_api.html#pipeline.code.dintegrate.utils.stable_gene_filter"><code class="docutils literal notranslate"><span class="pre">stable_gene_filter()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="../../../api/data_integration_api.html#pipeline.code.dintegrate.utils.calculate_thresholds"><code class="docutils literal notranslate"><span class="pre">calculate_thresholds()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="../../../api/data_integration_api.html#pipeline.code.dintegrate.main.DataIntegrator"><code class="docutils literal notranslate"><span class="pre">DataIntegrator</span></code></a><ul>
<li class="toctree-l4"><a class="reference internal" href="../../../api/data_integration_api.html#pipeline.code.dintegrate.main.DataIntegrator.__init__"><code class="docutils literal notranslate"><span class="pre">DataIntegrator.__init__()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="../../../api/data_integration_api.html#pipeline.code.dintegrate.main.DataIntegrator.run"><code class="docutils literal notranslate"><span class="pre">DataIntegrator.run()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="../../../api/data_integration_api.html#pipeline.code.dintegrate.main.DataIntegrator.pool_construction"><code class="docutils literal notranslate"><span class="pre">DataIntegrator.pool_construction()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="../../../api/data_integration_api.html#pipeline.code.dintegrate.main.DataIntegrator.clean_pool"><code class="docutils literal notranslate"><span class="pre">DataIntegrator.clean_pool()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="../../../api/data_integration_api.html#pipeline.code.dintegrate.main.DataIntegrator.weighted_average"><code class="docutils literal notranslate"><span class="pre">DataIntegrator.weighted_average()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="../../../api/data_integration_api.html#pipeline.code.dintegrate.main.DataIntegrator.downsample_data"><code class="docutils literal notranslate"><span class="pre">DataIntegrator.downsample_data()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="../../../api/data_integration_api.html#pipeline.code.dintegrate.main.DataIntegrator.generate_integrated_anndata"><code class="docutils literal notranslate"><span class="pre">DataIntegrator.generate_integrated_anndata()</span></code></a></li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="../../../api/dnn_api.html">DNN</a><ul>
<li class="toctree-l3"><a class="reference internal" href="../../../api/dnn_api.html#pipeline.code.dnn.analyzer.calculate_rank"><code class="docutils literal notranslate"><span class="pre">calculate_rank()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="../../../api/dnn_api.html#pipeline.code.dnn.analyzer.transform_human_mouse"><code class="docutils literal notranslate"><span class="pre">transform_human_mouse()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="../../../api/dnn_api.html#pipeline.code.dnn.analyzer.independent_test"><code class="docutils literal notranslate"><span class="pre">independent_test()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="../../../api/dnn_api.html#pipeline.code.dnn.pipeline.DnnPipeline"><code class="docutils literal notranslate"><span class="pre">DnnPipeline</span></code></a><ul>
<li class="toctree-l4"><a class="reference internal" href="../../../api/dnn_api.html#pipeline.code.dnn.pipeline.DnnPipeline.cfg"><code class="docutils literal notranslate"><span class="pre">DnnPipeline.cfg</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="../../../api/dnn_api.html#pipeline.code.dnn.pipeline.DnnPipeline.logger"><code class="docutils literal notranslate"><span class="pre">DnnPipeline.logger</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="../../../api/dnn_api.html#pipeline.code.dnn.pipeline.DnnPipeline.execute"><code class="docutils literal notranslate"><span class="pre">DnnPipeline.execute()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="../../../api/dnn_api.html#pipeline.code.dnn.pipeline.DnnPipeline.train"><code class="docutils literal notranslate"><span class="pre">DnnPipeline.train()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="../../../api/dnn_api.html#pipeline.code.dnn.pipeline.DnnPipeline.trans"><code class="docutils literal notranslate"><span class="pre">DnnPipeline.trans()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="../../../api/dnn_api.html#pipeline.code.dnn.pipeline.DnnPipeline.test"><code class="docutils literal notranslate"><span class="pre">DnnPipeline.test()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="../../../api/dnn_api.html#pipeline.code.dnn.pipeline.DnnPipeline._process_single_dataset"><code class="docutils literal notranslate"><span class="pre">DnnPipeline._process_single_dataset()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="../../../api/dnn_api.html#pipeline.code.dnn.pipeline.DnnPipeline._train_single_model"><code class="docutils literal notranslate"><span class="pre">DnnPipeline._train_single_model()</span></code></a></li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="../../../api/graph_walk_api.html">Graph Walk</a><ul>
<li class="toctree-l3"><a class="reference internal" href="../../../api/graph_walk_api.html#pipeline.code.graph_walk.run.generate_embeddings"><code class="docutils literal notranslate"><span class="pre">generate_embeddings()</span></code></a></li>
</ul>
</li>
</ul>
</li>
</ul>
<p class="caption" role="heading"><span class="caption-text">Other</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../../../faq.html">Frequently Asked Questions</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../citing.html">Citing TransBrain</a></li>
</ul>

        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../../../index.html">TransBrain</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="../../../index.html" class="icon icon-home" aria-label="Home"></a></li>
      <li class="breadcrumb-item active">Basic use 1: Mouse to human</li>
      <li class="wy-breadcrumbs-aside">
            <a href="../../../_sources/tutorials/notebooks/basic_usage/basic_use_mouse_to_human.ipynb.txt" rel="nofollow"> View page source</a>
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  <section id="Basic-use-1:-Mouse-to-human">
<h1>Basic use 1: Mouse to human<a class="headerlink" href="#Basic-use-1:-Mouse-to-human" title="Permalink to this heading"></a></h1>
<p>Here is a brief demonstration on how to use TransBrain to map the phenptype from mouse to human. The example data used in this tutorial can be download from <code class="docutils literal notranslate"><span class="pre">exampledata</span></code> directory of our <a class="reference external" href="https://github.com/ibpshangzheng/transbrain/tree/main/transbrain/exampledata">GitHub repository</a>.</p>
<section id="Step-1:-Prepare-input-data">
<h2>Step 1: Prepare input data<a class="headerlink" href="#Step-1:-Prepare-input-data" title="Permalink to this heading"></a></h2>
<p>There are two ways to input data. One is that you already have region-level phenotype data as a <code class="docutils literal notranslate"><span class="pre">CSV</span></code> table which follows the format and region order in the <a class="reference external" href="https://github.com/ibpshangzheng/transbrain/tree/main/transbrain/exampledata">provided template file</a>. The DataFrame contains two columns: <code class="docutils literal notranslate"><span class="pre">['Anatomical</span> <span class="pre">Name',</span> <span class="pre">'Phenotype']</span></code>.</p>
<p>Here we load the example mouse data table.</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[1]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span><span class="w"> </span><span class="nn">pandas</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">pd</span>

<span class="c1"># Example mouse data</span>
<span class="n">mouse_phenotype</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="s1">&#39;../../../transbrain/exampledata/mouse/mouse_all_example_data.csv&#39;</span><span class="p">,</span><span class="n">index_col</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[2]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">mouse_phenotype</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[2]:
</pre></div>
</div>
<div class="output_area rendered_html docutils container">
<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>Phenotype</th>
    </tr>
    <tr>
      <th>Anatomical Name</th>
      <th></th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>ACAd</th>
      <td>0.604528</td>
    </tr>
    <tr>
      <th>ACAv</th>
      <td>0.369476</td>
    </tr>
    <tr>
      <th>PL</th>
      <td>1.139296</td>
    </tr>
    <tr>
      <th>ILA</th>
      <td>0.539155</td>
    </tr>
    <tr>
      <th>ORBl</th>
      <td>0.000000</td>
    </tr>
    <tr>
      <th>...</th>
      <td>...</td>
    </tr>
    <tr>
      <th>MTN</th>
      <td>4.848956</td>
    </tr>
    <tr>
      <th>ILM</th>
      <td>5.223949</td>
    </tr>
    <tr>
      <th>GENv</th>
      <td>0.211631</td>
    </tr>
    <tr>
      <th>EPI</th>
      <td>3.944129</td>
    </tr>
    <tr>
      <th>RT</th>
      <td>1.150901</td>
    </tr>
  </tbody>
</table>
<p>68 rows × 1 columns</p>
</div></div>
</div>
<p>Or if you data is volumetric data in <code class="docutils literal notranslate"><span class="pre">.nii</span></code> or <code class="docutils literal notranslate"><span class="pre">.nii.gz</span></code> format that has been aligned to the mouse atlas space. You can use the following functions to extract region-level data. When fetcing atlas, you need to choose the region_type parameter from {‘cortex’, ‘subcortex’, ‘all’}, which controls the regions to include in returned region names and info table.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[3]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span><span class="w"> </span><span class="nn">transbrain</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">tb</span>

<span class="n">help</span><span class="p">(</span><span class="n">tb</span><span class="o">.</span><span class="n">atlas</span><span class="o">.</span><span class="n">fetch_mouse_atlas</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Help on function fetch_mouse_atlas in module transbrain.atlas:

fetch_mouse_atlas(region_type: Literal[&#39;cortex&#39;, &#39;subcortex&#39;, &#39;all&#39;] = &#39;all&#39;)
    Fetch the mouse atlas data and related information.

    This function loads the labeled mouse brain atlas image along with its corresponding ROI (region of interest) metadata table.

    Parameters
    ----------
    region_type : {&#39;cortex&#39;, &#39;subcortex&#39;, &#39;all&#39;}, optional
        Which regions to include in returned region names and info table: cortical, subcortical, or all.
        Default is &#39;all&#39;.

    Returns
    -------
    dict
        A dictionary with the following keys:

        - &#39;atlas&#39; : nibabel.Nifti1Image
            The loaded NIfTI image of the mouse atlas.
        - &#39;atlas_data&#39; : np.ndarray
            The atlas volume data as a NumPy array.
        - &#39;region_info&#39; : np.ndarray
            A list of anatomical region names, extracted from the &#39;Anatomical Name&#39; column of the label file.
        - &#39;info_table&#39; : pandas.DataFrame
            The full region-of-interest (ROI) information table including additional metadata.

    Raises
    ------
    FileNotFoundError
        If the atlas image file or the ROI label file is not found.

</pre></div></div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[4]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1">#Fetch mouse atlas</span>
<span class="n">mouse_atlas</span> <span class="o">=</span> <span class="n">tb</span><span class="o">.</span><span class="n">atlas</span><span class="o">.</span><span class="n">fetch_mouse_atlas</span><span class="p">(</span><span class="n">region_type</span><span class="o">=</span><span class="s1">&#39;all&#39;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[5]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">mouse_atlas</span><span class="o">.</span><span class="n">keys</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[5]:
</pre></div>
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
dict_keys([&#39;atlas&#39;, &#39;atlas_data&#39;, &#39;region_info&#39;, &#39;info_table&#39;])
</pre></div></div>
</div>
<p>The first two items in the returned dictionary are the NIfTI file of the atlas and the array in it, while the next two items store the brain region names and a table containing label values.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[6]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">mouse_atlas</span><span class="p">[</span><span class="s1">&#39;info_table&#39;</span><span class="p">]</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[6]:
</pre></div>
</div>
<div class="output_area rendered_html docutils container">
<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>Anatomical Name</th>
      <th>Atlas Index</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>ACAd</td>
      <td>1</td>
    </tr>
    <tr>
      <th>1</th>
      <td>ACAv</td>
      <td>2</td>
    </tr>
    <tr>
      <th>2</th>
      <td>AId</td>
      <td>3</td>
    </tr>
    <tr>
      <th>3</th>
      <td>AIp</td>
      <td>4</td>
    </tr>
    <tr>
      <th>4</th>
      <td>AIv</td>
      <td>5</td>
    </tr>
    <tr>
      <th>...</th>
      <td>...</td>
      <td>...</td>
    </tr>
    <tr>
      <th>63</th>
      <td>MTN</td>
      <td>64</td>
    </tr>
    <tr>
      <th>64</th>
      <td>ILM</td>
      <td>65</td>
    </tr>
    <tr>
      <th>65</th>
      <td>GENv</td>
      <td>66</td>
    </tr>
    <tr>
      <th>66</th>
      <td>EPI</td>
      <td>67</td>
    </tr>
    <tr>
      <th>67</th>
      <td>RT</td>
      <td>68</td>
    </tr>
  </tbody>
</table>
<p>68 rows × 2 columns</p>
</div></div>
</div>
<p>Next, we use the <code class="docutils literal notranslate"><span class="pre">get_region_phenotypes</span></code> function to calculate regional phenotype data from the image. In addition to the phenotype image and the atlas dictionary, you also need to specify the atlas name and the regions to be extracted.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[7]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">help</span><span class="p">(</span><span class="n">tb</span><span class="o">.</span><span class="n">base</span><span class="o">.</span><span class="n">get_region_phenotypes</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Help on function get_region_phenotypes in module transbrain.base:

get_region_phenotypes(phenotype_nii_path: str, atlas_dict: dict, atlas_type: Literal[&#39;bn&#39;, &#39;dk&#39;, &#39;aal&#39;, &#39;mouse&#39;] = &#39;bn&#39;, region_type: Literal[&#39;cortex&#39;, &#39;subcortex&#39;, &#39;all&#39;] = &#39;all&#39;, method: str = &#39;mean&#39;, resample: bool = True, label_column: str = &#39;Atlas Index&#39;, region_column: str = &#39;Anatomical Name&#39;) -&gt; pandas.core.frame.DataFrame
    Calculate region-wise phenotype values using a specified brain atlas.

    This function extracts regional statistics (mean or sum) from a phenotype NIfTI image
    based on a chosen human or mouse brain atlas. The atlas can be automatically
    resampled to match the phenotype image resolution if needed.

    Parameters
    ----------
    phenotype_nii_path : str
        Path to the input phenotype NIfTI file. Should be in MNI space for human atlases,
        or Allen CCFv3 space for mouse atlas.
    atlas_dict : dict
        A dictionary containing the following keys:

            - &#39;atlas&#39;: The loaded Mouse atlas image.
            - &#39;atlas_data&#39;: The atlas data as a numpy array.
            - &#39;region_info&#39;: A list of anatomical names for the specified regions.
            - &#39;info_table&#39;: The full ROI information table.

    atlas_type : {&#39;bn&#39;, &#39;dk&#39;, &#39;aal&#39;,&#39;mouse}, optional
        The type of atlas. Must be one of:
        - &#39;bn&#39;  : Brainnetome Atlas
        - &#39;dk&#39;  : Desikan-Killiany Atlas
        - &#39;aal&#39; : Automated Anatomical Labeling (AAL) Atlas
        - &#39;mouse&#39; : Allen Mouse CCFv3 atlas
        Default is &#39;bn&#39;.

    region_type : {&#39;cortex&#39;, &#39;subcortex&#39;, &#39;all&#39;}, optional
        Which regions to include in returned region names and info table: cortical, subcortical, or all. Default is &#39;all&#39;.
    method : {&#39;mean&#39;, &#39;sum&#39;}, optional
        Method for aggregating voxel values within each region. Default is &#39;mean&#39;.
    resample : bool, optional
        If True, resample the atlas to match the shape and resolution of the phenotype image.
        Default is True.
    label_column : str, optional
        Name of the column in the atlas label CSV that contains numeric label indices.
        Default is &#39;Atlas Index&#39;.
    region_column : str, optional
        Name of the column in the atlas label CSV that contains region names.
        Default is &#39;Anatomical Name&#39;.

    Returns
    -------
    pandas.DataFrame
        A DataFrame with aggregated phenotype values (mean or sum) for each region, , indexed by brain region name.

</pre></div></div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[8]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Get phenotypes in Mouse atlas used in TransBrain</span>
<span class="n">phenotype_nii_path</span> <span class="o">=</span> <span class="p">(</span><span class="s1">&#39;../../../transbrain/exampledata/mouse/mouse_example_phenotype_data.nii.gz&#39;</span><span class="p">)</span>
<span class="n">mouse_phenotype_extracted</span> <span class="o">=</span> <span class="n">tb</span><span class="o">.</span><span class="n">base</span><span class="o">.</span><span class="n">get_region_phenotypes</span><span class="p">(</span><span class="n">phenotype_nii_path</span><span class="p">,</span> <span class="n">atlas_dict</span> <span class="o">=</span> <span class="n">mouse_atlas</span><span class="p">,</span> <span class="n">atlas_type</span><span class="o">=</span><span class="s1">&#39;mouse&#39;</span><span class="p">,</span><span class="n">region_type</span><span class="o">=</span><span class="s1">&#39;all&#39;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[9]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">mouse_phenotype_extracted</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[9]:
</pre></div>
</div>
<div class="output_area rendered_html docutils container">
<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>Phenotype</th>
    </tr>
    <tr>
      <th>Anatomical Name</th>
      <th></th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>ACAd</th>
      <td>0.604520</td>
    </tr>
    <tr>
      <th>ACAv</th>
      <td>0.369486</td>
    </tr>
    <tr>
      <th>PL</th>
      <td>1.139259</td>
    </tr>
    <tr>
      <th>ILA</th>
      <td>0.539203</td>
    </tr>
    <tr>
      <th>ORBl</th>
      <td>0.000038</td>
    </tr>
    <tr>
      <th>...</th>
      <td>...</td>
    </tr>
    <tr>
      <th>MTN</th>
      <td>4.848961</td>
    </tr>
    <tr>
      <th>ILM</th>
      <td>5.223915</td>
    </tr>
    <tr>
      <th>GENv</th>
      <td>0.211644</td>
    </tr>
    <tr>
      <th>EPI</th>
      <td>3.944127</td>
    </tr>
    <tr>
      <th>RT</th>
      <td>1.150919</td>
    </tr>
  </tbody>
</table>
<p>68 rows × 1 columns</p>
</div></div>
</div>
<section id="Visualizing-mouse-phenotype-data">
<h3>Visualizing mouse phenotype data<a class="headerlink" href="#Visualizing-mouse-phenotype-data" title="Permalink to this heading"></a></h3>
<p>Using visualization function of TransBrain to view the data.</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[10]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span><span class="w"> </span><span class="nn">transbrain.vis</span><span class="w"> </span><span class="kn">import</span> <span class="n">map_phenotype_to_nifti</span><span class="p">,</span> <span class="n">plot_mouse_phenotype</span>

<span class="c1"># map the region-level phenotype data to an image</span>
<span class="n">phenotype_img</span> <span class="o">=</span> <span class="n">map_phenotype_to_nifti</span><span class="p">(</span><span class="n">mouse_phenotype</span><span class="p">,</span> <span class="n">mouse_atlas</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[11]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># view the image</span>
<span class="n">plot_mouse_phenotype</span><span class="p">(</span><span class="n">phenotype_img</span><span class="p">,</span> <span class="n">normalize_img</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">symmetric_cbar</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span><span class="n">vmax</span><span class="o">=</span><span class="mi">2</span><span class="p">,</span> <span class="n">threshold</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../../../_images/tutorials_notebooks_basic_usage_basic_use_mouse_to_human_20_0.png" src="../../../_images/tutorials_notebooks_basic_usage_basic_use_mouse_to_human_20_0.png" />
</div>
</div>
</section>
</section>
<section id="Step-2:-Initialization">
<h2>Step 2: Initialization<a class="headerlink" href="#Step-2:-Initialization" title="Permalink to this heading"></a></h2>
<p>Before starting the mapping, you need to initialize TransBrain by creating an instance of the <code class="docutils literal notranslate"><span class="pre">SpeciesTrans</span></code> class.</p>
<p>You need to specify which target human brain atlas you want to map to. You can choose <code class="docutils literal notranslate"><span class="pre">atlas_type</span></code> from <code class="docutils literal notranslate"><span class="pre">['bn',</span> <span class="pre">'dk',</span> <span class="pre">'aal']</span></code>. During this process, TransBrain will automatically load the necessary information for the mapping, including the brain region name list and the graph embedding matrix for the selected atlas.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[12]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1">#Initialize TransBrain for bn atlas</span>
<span class="n">Transformer</span> <span class="o">=</span> <span class="n">tb</span><span class="o">.</span><span class="n">trans</span><span class="o">.</span><span class="n">SpeciesTrans</span><span class="p">(</span><span class="s1">&#39;bn&#39;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
INFO:root:Initialized for bn atlas.
</pre></div></div>
</div>
</section>
<section id="Step-3:-Mapping">
<h2>Step 3: Mapping<a class="headerlink" href="#Step-3:-Mapping" title="Permalink to this heading"></a></h2>
<p>Now, we can map the mouse phenotype DataFrame (either loaded or extracted) to human brain, you need to choose the regions to map and whether to normalize the data. This step is enabled by default. If your data has already been normalized, you can set it to False.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[13]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">help</span><span class="p">(</span><span class="n">Transformer</span><span class="o">.</span><span class="n">mouse_to_human</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Help on method mouse_to_human in module transbrain.trans:

mouse_to_human(phenotype: pandas.core.frame.DataFrame, region_type: Literal[&#39;cortex&#39;, &#39;subcortex&#39;, &#39;all&#39;] = &#39;cortex&#39;, normalize: bool = True, restore: bool = False) -&gt; pandas.core.frame.DataFrame method of transbrain.trans.SpeciesTrans instance
    Translate mouse phenotype to human.

    Parameters
    ----------
    phenotype : pd.DataFrame
        Mouse phenotype DataFrame (regions × phenotypes).
    region_type : {&#39;cortex&#39;, &#39;subcortex&#39;, &#39;all&#39;}, optional
        The brain region type to translate. Default is &#39;cortex&#39;.
    normalize : bool, optional
        Whether to normalize data before translation. Default is True.
    restore : bool, optional
        Whether to restore values back to original scale after translation. Only used if normalize is True.
        Please enable this parameter with caution, unless you are certain that the distributions of this phenotype are consistent between the two species.

    Returns
    -------
    pd.DataFrame
        Translated human phenotype DataFrame (regions × phenotypes).

</pre></div></div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[14]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Example from mouse to human</span>
<span class="n">mouse_phenotype_in_human</span> <span class="o">=</span> <span class="n">Transformer</span><span class="o">.</span><span class="n">mouse_to_human</span><span class="p">(</span>
                            <span class="n">mouse_phenotype</span><span class="p">,</span>
                            <span class="n">region_type</span><span class="o">=</span><span class="s1">&#39;all&#39;</span><span class="p">,</span>
                            <span class="n">normalize</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
INFO:root:Successfully translated mouse all phenotypes to human.
</pre></div></div>
</div>
<p>Once the mapping is complete, you will see a success message. Let’s take a look at the mapped data. Now, you got the DataFrame that stores the phenotype data corresponding to the Brainnetome atlas.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[15]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">mouse_phenotype_in_human</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[15]:
</pre></div>
</div>
<div class="output_area rendered_html docutils container">
<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>Phenotype</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>A8m</th>
      <td>-0.036860</td>
    </tr>
    <tr>
      <th>A8dl</th>
      <td>-0.037950</td>
    </tr>
    <tr>
      <th>A9l</th>
      <td>-0.048863</td>
    </tr>
    <tr>
      <th>A6dl</th>
      <td>-0.023963</td>
    </tr>
    <tr>
      <th>A6m</th>
      <td>-0.005832</td>
    </tr>
    <tr>
      <th>...</th>
      <td>...</td>
    </tr>
    <tr>
      <th>Otha</th>
      <td>0.286257</td>
    </tr>
    <tr>
      <th>mPFtha</th>
      <td>0.568260</td>
    </tr>
    <tr>
      <th>lPFtha</th>
      <td>0.224642</td>
    </tr>
    <tr>
      <th>rTtha</th>
      <td>0.732512</td>
    </tr>
    <tr>
      <th>PPtha</th>
      <td>0.469093</td>
    </tr>
  </tbody>
</table>
<p>127 rows × 1 columns</p>
</div></div>
</div>
<p>You can also visualize what the mapped data looks like on human brain.</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[16]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1">#fetch human BN atlas</span>
<span class="n">human_atlas</span> <span class="o">=</span> <span class="n">tb</span><span class="o">.</span><span class="n">atlas</span><span class="o">.</span><span class="n">fetch_human_atlas</span><span class="p">(</span><span class="n">atlas_type</span><span class="o">=</span><span class="s1">&#39;bn&#39;</span><span class="p">,</span><span class="n">region_type</span><span class="o">=</span><span class="s1">&#39;all&#39;</span><span class="p">)</span>

<span class="c1"># map the region-level phenotype data to an image</span>
<span class="n">mapped_phenotype_img</span> <span class="o">=</span> <span class="n">map_phenotype_to_nifti</span><span class="p">(</span><span class="n">mouse_phenotype_in_human</span><span class="p">,</span> <span class="n">human_atlas</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[17]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span><span class="w"> </span><span class="nn">transbrain.vis</span><span class="w"> </span><span class="kn">import</span> <span class="n">plot_human_phenotype</span>

<span class="c1"># view the image</span>
<span class="n">plot_human_phenotype</span><span class="p">(</span><span class="n">mapped_phenotype_img</span><span class="p">,</span> <span class="n">normalize_img</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">cut_coords</span><span class="o">=</span><span class="nb">range</span><span class="p">(</span><span class="mi">10</span><span class="p">,</span> <span class="mi">60</span><span class="p">,</span> <span class="mi">10</span><span class="p">),</span> <span class="n">vmax</span><span class="o">=</span><span class="mi">2</span><span class="p">,</span> <span class="n">symmetric_cbar</span> <span class="o">=</span> <span class="kc">True</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
&lt;Figure size 3000x300 with 0 Axes&gt;
</pre></div></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../../../_images/tutorials_notebooks_basic_usage_basic_use_mouse_to_human_32_1.png" src="../../../_images/tutorials_notebooks_basic_usage_basic_use_mouse_to_human_32_1.png" />
</div>
</div>
</section>
</section>


           </div>
          </div>
          <footer><div class="rst-footer-buttons" role="navigation" aria-label="Footer">
        <a href="../../../online_mapping.html" class="btn btn-neutral float-left" title="Online Mapping" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
        <a href="basic_use_human_to_mouse.html" class="btn btn-neutral float-right" title="Basic use 2: Human to mouse" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right" aria-hidden="true"></span></a>
    </div>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright 2025, TransBrain.</p>
  </div>

  Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    provided by <a href="https://readthedocs.org">Read the Docs</a>.
   

</footer>
        </div>
      </div>
    </section>
  </div>
  <script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>