

<!DOCTYPE html>
<html class="writer-html5" lang="en">
<head>
  <meta charset="utf-8" />
  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>pipeline.code.dnn.analyzer &mdash; TransBrain 0.1 documentation</title>
      <link rel="stylesheet" type="text/css" href="../../../../_static/pygments.css?v=03e43079" />
      <link rel="stylesheet" type="text/css" href="../../../../_static/css/theme.css?v=e59714d7" />

  
      <script src="../../../../_static/jquery.js?v=5d32c60e"></script>
      <script src="../../../../_static/_sphinx_javascript_frameworks_compat.js?v=2cd50e6c"></script>
      <script data-url_root="../../../../" id="documentation_options" src="../../../../_static/documentation_options.js?v=e031e9a9"></script>
      <script src="../../../../_static/doctools.js?v=888ff710"></script>
      <script src="../../../../_static/sphinx_highlight.js?v=4825356b"></script>
      <script crossorigin="anonymous" integrity="sha256-Ae2Vz/4ePdIu6ZyI/5ZGsYnb+m0JlOmKPjt6XZ9JJkA=" src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.4/require.min.js"></script>
    <script src="../../../../_static/js/theme.js"></script>
    <link rel="index" title="Index" href="../../../../genindex.html" />
    <link rel="search" title="Search" href="../../../../search.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >

          
          
          <a href="../../../../index.html" class="icon icon-home">
            TransBrain
              <img src="../../../../_static/logo_final.png" class="logo" alt="Logo"/>
          </a>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../../../../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" aria-label="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <p class="caption" role="heading"><span class="caption-text">Setup &amp; Usage</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../../../../installation.html">Installation</a><ul>
<li class="toctree-l2"><a class="reference internal" href="../../../../installation.html#installation-options">Installation Options</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../../installation.html#python-dependencies">Python Dependencies</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../../installation.html#unit-test">Unit Test</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="../../../../user.html">User Guide</a><ul>
<li class="toctree-l2"><a class="reference internal" href="../../../../user.html#id2">Online Mapping</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../../user.html#id3">General Use</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../../user.html#id4">Advanced Use</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="../../../../atlas.html">Supported Atlases</a><ul>
<li class="toctree-l2"><a class="reference internal" href="../../../../atlas.html#human-brain-atlases">Human Brain Atlases</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../../atlas.html#mouse-brain-atlas">Mouse Brain Atlas</a></li>
</ul>
</li>
</ul>
<p class="caption" role="heading"><span class="caption-text">Tutorials</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../../../../methods/method_glossory.html">Method Glossory</a><ul>
<li class="toctree-l2"><a class="reference internal" href="../../../../methods/method_glossory.html#datasets">Datasets</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../../methods/method_glossory.html#methods-or-concepts">Methods or Concepts</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="../../../../online_mapping.html">Online Mapping</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../../tutorials/notebooks/basic_usage/basic_use_mouse_to_human.html">Basic use 1: Mouse to human</a><ul>
<li class="toctree-l2"><a class="reference internal" href="../../../../tutorials/notebooks/basic_usage/basic_use_mouse_to_human.html#Step-1:-Prepare-input-data">Step 1: Prepare input data</a><ul>
<li class="toctree-l3"><a class="reference internal" href="../../../../tutorials/notebooks/basic_usage/basic_use_mouse_to_human.html#Visualizing-mouse-phenotype-data">Visualizing mouse phenotype data</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="../../../../tutorials/notebooks/basic_usage/basic_use_mouse_to_human.html#Step-2:-Initialization">Step 2: Initialization</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../../tutorials/notebooks/basic_usage/basic_use_mouse_to_human.html#Step-3:-Mapping">Step 3: Mapping</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="../../../../tutorials/notebooks/basic_usage/basic_use_human_to_mouse.html">Basic use 2: Human to mouse</a><ul>
<li class="toctree-l2"><a class="reference internal" href="../../../../tutorials/notebooks/basic_usage/basic_use_human_to_mouse.html#Step-1:-Prepare-input-data">Step 1: Prepare input data</a><ul>
<li class="toctree-l3"><a class="reference internal" href="../../../../tutorials/notebooks/basic_usage/basic_use_human_to_mouse.html#Visualizing-human-phenotype-data">Visualizing human phenotype data</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="../../../../tutorials/notebooks/basic_usage/basic_use_human_to_mouse.html#Step-2:-Initialization">Step 2: Initialization</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../../tutorials/notebooks/basic_usage/basic_use_human_to_mouse.html#Step-3:-Mapping">Step 3: Mapping</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="../../../../tutorials/notebooks/translation/gradient_spectrum/fMRI_gradients.html">Case 1: Analyzing the conservation of resting-fMRI gradients</a><ul>
<li class="toctree-l2"><a class="reference internal" href="../../../../tutorials/notebooks/translation/gradient_spectrum/fMRI_gradients.html#Introduction">Introduction</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../../tutorials/notebooks/translation/gradient_spectrum/fMRI_gradients.html#Mapping-mouse-gradient-to-human">Mapping mouse gradient to human</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../../tutorials/notebooks/translation/gradient_spectrum/fMRI_gradients.html#Compare-with-human-Gradient">Compare with human Gradient</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../../tutorials/notebooks/translation/gradient_spectrum/fMRI_gradients.html#Cross-species-correlation">Cross-species correlation</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="../../../../tutorials/notebooks/translation/optogenetic_annotation/optogenetic_circuits.html">Case 2: Annotating optogenetic circuits</a><ul>
<li class="toctree-l2"><a class="reference internal" href="../../../../tutorials/notebooks/translation/optogenetic_annotation/optogenetic_circuits.html#Introduction">Introduction</a><ul>
<li class="toctree-l3"><a class="reference internal" href="../../../../tutorials/notebooks/translation/optogenetic_annotation/optogenetic_circuits.html#Data">Data</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="../../../../tutorials/notebooks/translation/optogenetic_annotation/optogenetic_circuits.html#Translating-mouse-stimulated-pattern-to-human">Translating mouse stimulated pattern to human</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../../tutorials/notebooks/translation/optogenetic_annotation/optogenetic_circuits.html#Annotating-using-NeuroSynth.">Annotating using NeuroSynth.</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="../../../../tutorials/notebooks/translation/autism_mutation/gene_mutations.html">Case 3: Linking gene mutations to imaging phenotype deviations</a><ul>
<li class="toctree-l2"><a class="reference internal" href="../../../../tutorials/notebooks/translation/autism_mutation/gene_mutations.html#Introduction">Introduction</a><ul>
<li class="toctree-l3"><a class="reference internal" href="../../../../tutorials/notebooks/translation/autism_mutation/gene_mutations.html#Data">Data</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="../../../../tutorials/notebooks/translation/autism_mutation/gene_mutations.html#Translating-mouse-mutation-patterns-to-human">Translating mouse mutation patterns to human</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../../tutorials/notebooks/translation/autism_mutation/gene_mutations.html#Calculate-individual-gene-risk-score">Calculate individual gene risk score</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="../../../../building.html">Advanced Use</a><ul>
<li class="toctree-l2"><a class="reference internal" href="../../../../pipeline/run_pipeline/integration.html">Transcriptional Data Integration</a><ul>
<li class="toctree-l3"><a class="reference internal" href="../../../../pipeline/run_pipeline/integration.html#Introduction">Introduction</a><ul>
<li class="toctree-l4"><a class="reference internal" href="../../../../pipeline/run_pipeline/integration.html#AHBA-data">AHBA data</a></li>
<li class="toctree-l4"><a class="reference internal" href="../../../../pipeline/run_pipeline/integration.html#Human-single-nucleus-data">Human single-nucleus data</a></li>
<li class="toctree-l4"><a class="reference internal" href="../../../../pipeline/run_pipeline/integration.html#Mouse-spatial-transcriptomic-data">Mouse spatial transcriptomic data</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="../../../../pipeline/run_pipeline/integration.html#Selection-of-cortical-stable-genes">Selection of cortical stable genes</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../../../pipeline/run_pipeline/integration.html#Selection-of-subcortical-stable-genes">Selection of subcortical stable genes</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../../../pipeline/run_pipeline/integration.html#Selection-of-all-stable-genes">Selection of all stable genes</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../../../pipeline/run_pipeline/integration.html#Thresholds-of-cortex">Thresholds of cortex</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../../../pipeline/run_pipeline/integration.html#Thresholds-of-subcortex">Thresholds of subcortex</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../../../pipeline/run_pipeline/integration.html#Thresholds-of-all">Thresholds of all</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="../../../../pipeline/run_pipeline/dnn.html">DNN Model Training</a><ul>
<li class="toctree-l3"><a class="reference internal" href="../../../../pipeline/run_pipeline/dnn.html#Introduction">Introduction</a><ul>
<li class="toctree-l4"><a class="reference internal" href="../../../../pipeline/run_pipeline/dnn.html#Model-architecture">Model architecture</a></li>
<li class="toctree-l4"><a class="reference internal" href="../../../../pipeline/run_pipeline/dnn.html#Region-specific-embeddings">Region-specific embeddings</a></li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="../../../../pipeline/run_pipeline/graph_embedding.html">Graph-based Random Walk</a><ul>
<li class="toctree-l3"><a class="reference internal" href="../../../../pipeline/run_pipeline/graph_embedding.html#Introduction">Introduction</a><ul>
<li class="toctree-l4"><a class="reference internal" href="../../../../pipeline/run_pipeline/graph_embedding.html#Human-diffusion-based-tractography">Human diffusion-based tractography</a></li>
<li class="toctree-l4"><a class="reference internal" href="../../../../pipeline/run_pipeline/graph_embedding.html#Mouse-Tracer">Mouse Tracer</a></li>
<li class="toctree-l4"><a class="reference internal" href="../../../../pipeline/run_pipeline/graph_embedding.html#Embedding-space-generated-from-Human-Mouse-graph">Embedding space generated from Human-Mouse graph</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="../../../../pipeline/run_pipeline/graph_embedding.html#Define-human-and-mouse-ROIs">Define human and mouse ROIs</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../../../pipeline/run_pipeline/graph_embedding.html#Load-anatomical-hierachy-constraint-information">Load anatomical hierachy constraint information</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="../../../../methods/mapping.html">Bidirectional Mapping</a></li>
</ul>
</li>
</ul>
<p class="caption" role="heading"><span class="caption-text">API</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../../../../api/external_api.html">API Reference</a><ul>
<li class="toctree-l2"><a class="reference internal" href="../../../../api/fetch_api.html">Fetching Atlases</a><ul>
<li class="toctree-l3"><a class="reference internal" href="../../../../api/fetch_api.html#transbrain.atlas.fetch_human_atlas"><code class="docutils literal notranslate"><span class="pre">fetch_human_atlas()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="../../../../api/fetch_api.html#transbrain.atlas.fetch_mouse_atlas"><code class="docutils literal notranslate"><span class="pre">fetch_mouse_atlas()</span></code></a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="../../../../api/base_api.html">Base Functions</a><ul>
<li class="toctree-l3"><a class="reference internal" href="../../../../api/base_api.html#transbrain.base.get_region_phenotypes"><code class="docutils literal notranslate"><span class="pre">get_region_phenotypes()</span></code></a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="../../../../api/mapping_api.html">Mapping Functions</a><ul>
<li class="toctree-l3"><a class="reference internal" href="../../../../api/mapping_api.html#transbrain.trans.SpeciesTrans"><code class="docutils literal notranslate"><span class="pre">SpeciesTrans</span></code></a><ul>
<li class="toctree-l4"><a class="reference internal" href="../../../../api/mapping_api.html#transbrain.trans.SpeciesTrans.atlas_type"><code class="docutils literal notranslate"><span class="pre">SpeciesTrans.atlas_type</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="../../../../api/mapping_api.html#transbrain.trans.SpeciesTrans.regions"><code class="docutils literal notranslate"><span class="pre">SpeciesTrans.regions</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="../../../../api/mapping_api.html#transbrain.trans.SpeciesTrans.embeddings"><code class="docutils literal notranslate"><span class="pre">SpeciesTrans.embeddings</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="../../../../api/mapping_api.html#transbrain.trans.SpeciesTrans._load_embeddings"><code class="docutils literal notranslate"><span class="pre">SpeciesTrans._load_embeddings()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="../../../../api/mapping_api.html#transbrain.trans.SpeciesTrans._dual_mapping"><code class="docutils literal notranslate"><span class="pre">SpeciesTrans._dual_mapping()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="../../../../api/mapping_api.html#transbrain.trans.SpeciesTrans._translate"><code class="docutils literal notranslate"><span class="pre">SpeciesTrans._translate()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="../../../../api/mapping_api.html#transbrain.trans.SpeciesTrans.mouse_to_human"><code class="docutils literal notranslate"><span class="pre">SpeciesTrans.mouse_to_human()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="../../../../api/mapping_api.html#transbrain.trans.SpeciesTrans.human_to_mouse"><code class="docutils literal notranslate"><span class="pre">SpeciesTrans.human_to_mouse()</span></code></a></li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="../../../../api/vis_api.html">Visualization Functions</a><ul>
<li class="toctree-l3"><a class="reference internal" href="../../../../api/vis_api.html#transbrain.vis.map_phenotype_to_nifti"><code class="docutils literal notranslate"><span class="pre">map_phenotype_to_nifti()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="../../../../api/vis_api.html#transbrain.vis.plot_mouse_phenotype"><code class="docutils literal notranslate"><span class="pre">plot_mouse_phenotype()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="../../../../api/vis_api.html#transbrain.vis.plot_human_phenotype"><code class="docutils literal notranslate"><span class="pre">plot_human_phenotype()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="../../../../api/vis_api.html#transbrain.surf.plot_surface_phenotype"><code class="docutils literal notranslate"><span class="pre">plot_surface_phenotype()</span></code></a></li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="../../../../api/utils.html">Development Functions</a><ul>
<li class="toctree-l2"><a class="reference internal" href="../../../../api/data_process_api.html">Transcriptional Data Process</a><ul>
<li class="toctree-l3"><a class="reference internal" href="../../../../api/data_process_api.html#pipeline.code.dpreprocess.sn_.Preprocess_Sn"><code class="docutils literal notranslate"><span class="pre">Preprocess_Sn</span></code></a><ul>
<li class="toctree-l4"><a class="reference internal" href="../../../../api/data_process_api.html#pipeline.code.dpreprocess.sn_.Preprocess_Sn.dataset_qc"><code class="docutils literal notranslate"><span class="pre">Preprocess_Sn.dataset_qc()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="../../../../api/data_process_api.html#pipeline.code.dpreprocess.sn_.Preprocess_Sn.dataset_normalized"><code class="docutils literal notranslate"><span class="pre">Preprocess_Sn.dataset_normalized()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="../../../../api/data_process_api.html#pipeline.code.dpreprocess.sn_.Preprocess_Sn.dataset_common_genes"><code class="docutils literal notranslate"><span class="pre">Preprocess_Sn.dataset_common_genes()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="../../../../api/data_process_api.html#pipeline.code.dpreprocess.sn_.Preprocess_Sn.dataset_batch_correct"><code class="docutils literal notranslate"><span class="pre">Preprocess_Sn.dataset_batch_correct()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="../../../../api/data_process_api.html#pipeline.code.dpreprocess.sn_.Preprocess_Sn.dataset_smooth"><code class="docutils literal notranslate"><span class="pre">Preprocess_Sn.dataset_smooth()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="../../../../api/data_process_api.html#pipeline.code.dpreprocess.sn_.Preprocess_Sn.dataset_scaled"><code class="docutils literal notranslate"><span class="pre">Preprocess_Sn.dataset_scaled()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="../../../../api/data_process_api.html#pipeline.code.dpreprocess.sn_.Preprocess_Sn.preprocess_pipeline"><code class="docutils literal notranslate"><span class="pre">Preprocess_Sn.preprocess_pipeline()</span></code></a></li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="../../../../api/data_integration_api.html">Data Integration</a><ul>
<li class="toctree-l3"><a class="reference internal" href="../../../../api/data_integration_api.html#pipeline.code.dintegrate.utils.stable_gene_filter"><code class="docutils literal notranslate"><span class="pre">stable_gene_filter()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="../../../../api/data_integration_api.html#pipeline.code.dintegrate.utils.calculate_thresholds"><code class="docutils literal notranslate"><span class="pre">calculate_thresholds()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="../../../../api/data_integration_api.html#pipeline.code.dintegrate.main.DataIntegrator"><code class="docutils literal notranslate"><span class="pre">DataIntegrator</span></code></a><ul>
<li class="toctree-l4"><a class="reference internal" href="../../../../api/data_integration_api.html#pipeline.code.dintegrate.main.DataIntegrator.__init__"><code class="docutils literal notranslate"><span class="pre">DataIntegrator.__init__()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="../../../../api/data_integration_api.html#pipeline.code.dintegrate.main.DataIntegrator.run"><code class="docutils literal notranslate"><span class="pre">DataIntegrator.run()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="../../../../api/data_integration_api.html#pipeline.code.dintegrate.main.DataIntegrator.pool_construction"><code class="docutils literal notranslate"><span class="pre">DataIntegrator.pool_construction()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="../../../../api/data_integration_api.html#pipeline.code.dintegrate.main.DataIntegrator.clean_pool"><code class="docutils literal notranslate"><span class="pre">DataIntegrator.clean_pool()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="../../../../api/data_integration_api.html#pipeline.code.dintegrate.main.DataIntegrator.weighted_average"><code class="docutils literal notranslate"><span class="pre">DataIntegrator.weighted_average()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="../../../../api/data_integration_api.html#pipeline.code.dintegrate.main.DataIntegrator.downsample_data"><code class="docutils literal notranslate"><span class="pre">DataIntegrator.downsample_data()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="../../../../api/data_integration_api.html#pipeline.code.dintegrate.main.DataIntegrator.generate_integrated_anndata"><code class="docutils literal notranslate"><span class="pre">DataIntegrator.generate_integrated_anndata()</span></code></a></li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="../../../../api/dnn_api.html">DNN</a><ul>
<li class="toctree-l3"><a class="reference internal" href="../../../../api/dnn_api.html#pipeline.code.dnn.analyzer.calculate_rank"><code class="docutils literal notranslate"><span class="pre">calculate_rank()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="../../../../api/dnn_api.html#pipeline.code.dnn.analyzer.transform_human_mouse"><code class="docutils literal notranslate"><span class="pre">transform_human_mouse()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="../../../../api/dnn_api.html#pipeline.code.dnn.analyzer.independent_test"><code class="docutils literal notranslate"><span class="pre">independent_test()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="../../../../api/dnn_api.html#pipeline.code.dnn.pipeline.DnnPipeline"><code class="docutils literal notranslate"><span class="pre">DnnPipeline</span></code></a><ul>
<li class="toctree-l4"><a class="reference internal" href="../../../../api/dnn_api.html#pipeline.code.dnn.pipeline.DnnPipeline.cfg"><code class="docutils literal notranslate"><span class="pre">DnnPipeline.cfg</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="../../../../api/dnn_api.html#pipeline.code.dnn.pipeline.DnnPipeline.logger"><code class="docutils literal notranslate"><span class="pre">DnnPipeline.logger</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="../../../../api/dnn_api.html#pipeline.code.dnn.pipeline.DnnPipeline.execute"><code class="docutils literal notranslate"><span class="pre">DnnPipeline.execute()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="../../../../api/dnn_api.html#pipeline.code.dnn.pipeline.DnnPipeline.train"><code class="docutils literal notranslate"><span class="pre">DnnPipeline.train()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="../../../../api/dnn_api.html#pipeline.code.dnn.pipeline.DnnPipeline.trans"><code class="docutils literal notranslate"><span class="pre">DnnPipeline.trans()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="../../../../api/dnn_api.html#pipeline.code.dnn.pipeline.DnnPipeline.test"><code class="docutils literal notranslate"><span class="pre">DnnPipeline.test()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="../../../../api/dnn_api.html#pipeline.code.dnn.pipeline.DnnPipeline._process_single_dataset"><code class="docutils literal notranslate"><span class="pre">DnnPipeline._process_single_dataset()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="../../../../api/dnn_api.html#pipeline.code.dnn.pipeline.DnnPipeline._train_single_model"><code class="docutils literal notranslate"><span class="pre">DnnPipeline._train_single_model()</span></code></a></li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="../../../../api/graph_walk_api.html">Graph Walk</a><ul>
<li class="toctree-l3"><a class="reference internal" href="../../../../api/graph_walk_api.html#pipeline.code.graph_walk.run.generate_embeddings"><code class="docutils literal notranslate"><span class="pre">generate_embeddings()</span></code></a></li>
</ul>
</li>
</ul>
</li>
</ul>
<p class="caption" role="heading"><span class="caption-text">Other</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../../../../faq.html">Frequently Asked Questions</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../../citing.html">Citing TransBrain</a></li>
</ul>

        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../../../../index.html">TransBrain</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="../../../../index.html" class="icon icon-home" aria-label="Home"></a></li>
          <li class="breadcrumb-item"><a href="../../../index.html">Module code</a></li>
      <li class="breadcrumb-item active">pipeline.code.dnn.analyzer</li>
      <li class="wy-breadcrumbs-aside">
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  <h1>Source code for pipeline.code.dnn.analyzer</h1><div class="highlight"><pre>
<span></span><span class="kn">import</span><span class="w"> </span><span class="nn">os</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">torch</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">numpy</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">np</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">pandas</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">pd</span>
<span class="kn">from</span><span class="w"> </span><span class="nn">pathlib</span><span class="w"> </span><span class="kn">import</span> <span class="n">Path</span>
<span class="kn">from</span><span class="w"> </span><span class="nn">collections</span><span class="w"> </span><span class="kn">import</span> <span class="n">defaultdict</span>
<span class="kn">from</span><span class="w"> </span><span class="nn">pipeline.code.dnn.dataloader</span><span class="w"> </span><span class="kn">import</span> <span class="n">create_dataloader</span>
<span class="kn">from</span><span class="w"> </span><span class="nn">pipeline.code.dnn.dataloader</span><span class="w"> </span><span class="kn">import</span> <span class="n">generate_matrix</span>
<span class="kn">from</span><span class="w"> </span><span class="nn">pipeline.code.dnn.config.logger</span><span class="w"> </span><span class="kn">import</span> <span class="n">get_logger</span>
<span class="kn">from</span><span class="w"> </span><span class="nn">pipeline.code.dnn.models</span><span class="w"> </span><span class="kn">import</span> <span class="n">ClassifierModule</span>
<span class="kn">from</span><span class="w"> </span><span class="nn">sklearn.metrics</span><span class="w"> </span><span class="kn">import</span> <span class="n">confusion_matrix</span>


<span class="k">def</span><span class="w"> </span><span class="nf">train_test_split</span><span class="p">(</span><span class="n">dataset</span><span class="p">,</span> <span class="n">valid_size</span><span class="p">,</span> <span class="n">sample_id</span><span class="p">,</span> <span class="n">all_id</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Stratified train-validation split based on brain regions.</span>

<span class="sd">    This function performs a stratified train-validation split by brain region.</span>

<span class="sd">    Parameters</span>
<span class="sd">    ----------</span>
<span class="sd">    dataset : ndarray of shape (n_samples, n_features + 1)</span>
<span class="sd">        Feature matrix concatenated with the target labels in the last column.</span>
<span class="sd">    valid_size : float</span>
<span class="sd">        Proportion (between 0 and 1) of samples per brain region to allocate to the validation set.</span>
<span class="sd">    sample_id : set of str</span>
<span class="sd">        Collection of sample identifiers in the format &#39;region_sampleX&#39;.</span>
<span class="sd">    all_id : list of str</span>
<span class="sd">        Full identifiers in the format &#39;region_sampleX_scellY&#39;, corresponding to each row in `dataset`.</span>

<span class="sd">    Returns</span>
<span class="sd">    -------</span>
<span class="sd">    train : ndarray</span>
<span class="sd">        Training set after splitting.</span>
<span class="sd">    val : ndarray</span>
<span class="sd">        Validation set after splitting.</span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="n">brain_sample_map</span> <span class="o">=</span> <span class="n">defaultdict</span><span class="p">(</span><span class="nb">set</span><span class="p">)</span>  
    <span class="n">sample_allid_map</span> <span class="o">=</span> <span class="n">defaultdict</span><span class="p">(</span><span class="nb">list</span><span class="p">)</span>  
    
    <span class="k">for</span> <span class="n">idx</span><span class="p">,</span> <span class="n">full_id</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">all_id</span><span class="p">):</span>
        <span class="n">parts</span> <span class="o">=</span> <span class="n">full_id</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39;_&#39;</span><span class="p">)</span>
        <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">parts</span><span class="p">)</span> <span class="o">&lt;</span> <span class="mi">2</span><span class="p">:</span>
            <span class="k">continue</span>  
        
        <span class="n">brain_region</span> <span class="o">=</span> <span class="n">parts</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
        <span class="n">sample_key</span> <span class="o">=</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">parts</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="si">}</span><span class="s2">_</span><span class="si">{</span><span class="n">parts</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span><span class="si">}</span><span class="s2">&quot;</span>
        
        <span class="k">if</span> <span class="n">sample_key</span> <span class="ow">in</span> <span class="n">sample_id</span><span class="p">:</span>
            <span class="n">brain_sample_map</span><span class="p">[</span><span class="n">brain_region</span><span class="p">]</span><span class="o">.</span><span class="n">add</span><span class="p">(</span><span class="n">sample_key</span><span class="p">)</span>
            <span class="n">sample_allid_map</span><span class="p">[</span><span class="n">sample_key</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">idx</span><span class="p">)</span>
    
    <span class="n">val_indices</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="k">for</span> <span class="n">brain_region</span><span class="p">,</span> <span class="n">samples</span> <span class="ow">in</span> <span class="n">brain_sample_map</span><span class="o">.</span><span class="n">items</span><span class="p">():</span>
        <span class="n">sample_list</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="n">samples</span><span class="p">)</span>
        <span class="n">n_samples</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="n">sample_list</span><span class="p">)</span>

        <span class="n">n_val</span> <span class="o">=</span> <span class="nb">max</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="nb">int</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">round</span><span class="p">(</span><span class="n">n_samples</span> <span class="o">*</span> <span class="n">valid_size</span><span class="p">)))</span>

        <span class="n">selected_samples</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">random</span><span class="o">.</span><span class="n">choice</span><span class="p">(</span><span class="n">sample_list</span><span class="p">,</span> <span class="n">size</span><span class="o">=</span><span class="n">n_val</span><span class="p">,</span> <span class="n">replace</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
        
        <span class="k">for</span> <span class="n">sample</span> <span class="ow">in</span> <span class="n">selected_samples</span><span class="p">:</span>
            <span class="n">val_indices</span><span class="o">.</span><span class="n">extend</span><span class="p">(</span><span class="n">sample_allid_map</span><span class="p">[</span><span class="n">sample</span><span class="p">])</span>
    
    <span class="n">mask</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">zeros</span><span class="p">(</span><span class="n">dataset</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">0</span><span class="p">],</span> <span class="n">dtype</span><span class="o">=</span><span class="nb">bool</span><span class="p">)</span>
    <span class="n">mask</span><span class="p">[</span><span class="n">val_indices</span><span class="p">]</span> <span class="o">=</span> <span class="kc">True</span>

    <span class="n">train</span> <span class="o">=</span> <span class="n">dataset</span><span class="p">[</span><span class="o">~</span><span class="n">mask</span><span class="p">]</span> 
    <span class="n">val</span> <span class="o">=</span> <span class="n">dataset</span><span class="p">[</span><span class="n">mask</span><span class="p">]</span>
    
    <span class="k">return</span> <span class="n">train</span><span class="p">,</span> <span class="n">val</span>

<span class="k">def</span><span class="w"> </span><span class="nf">generate_average_data</span><span class="p">(</span><span class="n">data_path</span><span class="p">,</span><span class="n">params</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Generate and process transcriptional dataset with region averaging.</span>

<span class="sd">    This function loads transcriptional data and averages the features across</span>
<span class="sd">    brain regions. Each region&#39;s data is collapsed into a single feature vector</span>
<span class="sd">    representing the mean across all samples from that region.</span>

<span class="sd">    Parameters</span>
<span class="sd">    ----------</span>
<span class="sd">    data_path : str</span>
<span class="sd">        Path to the dataset file containing transcriptional features and metadata.</span>
<span class="sd">    params : object</span>
<span class="sd">        Configuration object containing options for data loading and preprocessing.</span>

<span class="sd">    Returns</span>
<span class="sd">    -------</span>
<span class="sd">    averaged_data : ndarray of shape (n_regions, n_features + 1)</span>
<span class="sd">        Matrix where each row represents a brain region. The last column contains</span>
<span class="sd">        the region label, and preceding columns are the averaged features.</span>
<span class="sd">    unique_regions : ndarray</span>
<span class="sd">        Array of unique region identifiers.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">features</span><span class="p">,</span> <span class="n">labels</span><span class="p">,</span> <span class="n">region_ids</span><span class="p">,</span> <span class="n">_</span><span class="p">,</span> <span class="n">_</span> <span class="o">=</span> <span class="n">generate_matrix</span><span class="p">(</span><span class="n">data_path</span><span class="p">,</span> <span class="n">params</span><span class="p">)</span>

    <span class="n">combined_data</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">hstack</span><span class="p">([</span><span class="n">features</span><span class="p">,</span> <span class="n">labels</span><span class="o">.</span><span class="n">reshape</span><span class="p">(</span><span class="o">-</span><span class="mi">1</span><span class="p">,</span><span class="mi">1</span><span class="p">)])</span>
        
    <span class="n">unique_regions</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">unique</span><span class="p">(</span><span class="n">combined_data</span><span class="p">[:,</span> <span class="o">-</span><span class="mi">1</span><span class="p">])</span>
    <span class="n">averaged_data</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">([</span>
        <span class="n">np</span><span class="o">.</span><span class="n">append</span><span class="p">(</span> 
            <span class="n">np</span><span class="o">.</span><span class="n">mean</span><span class="p">(</span><span class="n">combined_data</span><span class="p">[</span><span class="n">combined_data</span><span class="p">[:,</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span> <span class="o">==</span> <span class="n">region</span><span class="p">,</span> <span class="p">:</span><span class="o">-</span><span class="mi">1</span><span class="p">],</span> <span class="n">axis</span><span class="o">=</span><span class="mi">0</span><span class="p">),</span>
            <span class="n">region</span>
        <span class="p">)</span> <span class="k">for</span> <span class="n">region</span> <span class="ow">in</span> <span class="n">unique_regions</span><span class="p">])</span>
        
    <span class="k">return</span> <span class="n">averaged_data</span><span class="p">,</span> <span class="n">np</span><span class="o">.</span><span class="n">unique</span><span class="p">(</span><span class="n">region_ids</span><span class="p">)</span>

<span class="k">def</span><span class="w"> </span><span class="nf">extract_embeddings</span><span class="p">(</span><span class="n">trans_data</span><span class="p">,</span> <span class="n">nn_model</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Extract DNN embeddings.</span>

<span class="sd">    This function passes the data through a trained neural network model</span>
<span class="sd">    to extract latent feature embeddings.</span>

<span class="sd">    Parameters</span>
<span class="sd">    ----------</span>
<span class="sd">    trans_data : numpy.ndarray </span>
<span class="sd">        Processed data to extract embeddings from.</span>
<span class="sd">    nn_model : torch.nn.Module</span>
<span class="sd">        A trained neural network model used to extract embeddings. </span>

<span class="sd">    Returns</span>
<span class="sd">    -------</span>
<span class="sd">    embeddings : ndarray of shape (n_samples, embedding_dim)</span>
<span class="sd">        Concatenated latent embeddings extracted from the model.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">data_loader</span> <span class="o">=</span> <span class="n">create_dataloader</span><span class="p">(</span><span class="n">dataset</span><span class="o">=</span><span class="n">trans_data</span><span class="p">,</span><span class="n">batch_size</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span><span class="n">shuffle</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span><span class="n">drop_last</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
        
    <span class="n">embeddings</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="k">with</span> <span class="n">torch</span><span class="o">.</span><span class="n">no_grad</span><span class="p">():</span>
        <span class="k">for</span> <span class="n">inputs</span><span class="p">,</span> <span class="n">_</span> <span class="ow">in</span> <span class="n">data_loader</span><span class="p">:</span>
            <span class="n">inputs</span> <span class="o">=</span> <span class="n">inputs</span><span class="o">.</span><span class="n">cuda</span><span class="p">()</span><span class="o">.</span><span class="n">float</span><span class="p">()</span>
            <span class="k">try</span><span class="p">:</span>
                <span class="n">_</span><span class="p">,</span><span class="n">latent_emb</span> <span class="o">=</span> <span class="n">nn_model</span><span class="p">(</span><span class="n">inputs</span><span class="p">)</span>
                <span class="n">embeddings</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">latent_emb</span><span class="o">.</span><span class="n">cpu</span><span class="p">()</span><span class="o">.</span><span class="n">numpy</span><span class="p">())</span>
            <span class="k">except</span><span class="p">:</span>
                <span class="n">latent_emb</span> <span class="o">=</span> <span class="n">nn_model</span><span class="p">(</span><span class="n">inputs</span><span class="p">)</span>
                <span class="n">embeddings</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">latent_emb</span><span class="o">.</span><span class="n">cpu</span><span class="p">()</span><span class="o">.</span><span class="n">numpy</span><span class="p">())</span>
                
    <span class="k">return</span> <span class="n">np</span><span class="o">.</span><span class="n">concatenate</span><span class="p">(</span><span class="n">embeddings</span><span class="p">)</span>

<span class="k">def</span><span class="w"> </span><span class="nf">create_aligned_dataframe</span><span class="p">(</span><span class="n">embeddings</span><span class="p">,</span> <span class="n">region_ids</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Create region-aligned feature dataframe&quot;&quot;&quot;</span>
    <span class="k">return</span> <span class="p">(</span>
            <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span><span class="n">embeddings</span><span class="p">,</span> <span class="n">columns</span><span class="o">=</span><span class="nb">map</span><span class="p">(</span><span class="nb">str</span><span class="p">,</span> <span class="nb">range</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="n">embeddings</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span><span class="o">+</span><span class="mi">1</span><span class="p">)))</span>
            <span class="o">.</span><span class="n">assign</span><span class="p">(</span><span class="n">region_id</span><span class="o">=</span><span class="n">region_ids</span><span class="p">)</span>
            <span class="o">.</span><span class="n">set_index</span><span class="p">(</span><span class="s1">&#39;region_id&#39;</span><span class="p">)</span>
            <span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">region_ids</span><span class="p">]</span> 
        <span class="p">)</span>

<span class="k">def</span><span class="w"> </span><span class="nf">calculate_homology_rank</span><span class="p">(</span><span class="n">df_corr</span><span class="p">,</span><span class="n">mouse_regions</span><span class="p">,</span> <span class="n">human_regions</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Calculate average rank of true homologous human regions in mouse-to-human correlation.</span>

<span class="sd">    For each mouse region, the function retrieves the correlation values with all human regions</span>
<span class="sd">    from the correlation DataFrame `df_corr`, ranks them in descending order, and finds the</span>
<span class="sd">    rank position of the true homologous human region. It returns the average of these ranks.</span>

<span class="sd">    Parameters</span>
<span class="sd">    ----------</span>
<span class="sd">    df_corr : pandas.DataFrame</span>
<span class="sd">        A square DataFrame of shape (n_human_regions, n_mouse_regions), where each column</span>
<span class="sd">        corresponds to a mouse region and each row to a human region. The values represent</span>
<span class="sd">        similarity or correlation scores.</span>
<span class="sd">    mouse_regions : list of str</span>
<span class="sd">        List of mouse region names corresponding to the columns of `df_corr`.</span>
<span class="sd">    human_regions : list of str</span>
<span class="sd">        List of prior homologous human region names, same length as `mouse_regions`.</span>
<span class="sd">        Each entry corresponds to the homolog of the mouse region at the same index.</span>

<span class="sd">    Returns</span>
<span class="sd">    -------</span>
<span class="sd">    float</span>
<span class="sd">        The mean rank of the prior homologous human regions in descending correlation order</span>
<span class="sd">        for each mouse region. A lower value indicates better alignment.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">region_pairs</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="nb">zip</span><span class="p">(</span><span class="n">mouse_regions</span><span class="p">,</span> <span class="n">human_regions</span><span class="p">))</span>

    <span class="n">ranks</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="k">for</span> <span class="n">mouse_reg</span><span class="p">,</span> <span class="n">human_reg</span> <span class="ow">in</span> <span class="n">region_pairs</span><span class="p">:</span>
        <span class="n">sorted_series</span> <span class="o">=</span> <span class="n">df_corr</span><span class="p">[</span><span class="n">mouse_reg</span><span class="p">]</span><span class="o">.</span><span class="n">sort_values</span><span class="p">(</span><span class="n">ascending</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
        <span class="n">rank</span> <span class="o">=</span> <span class="n">sorted_series</span><span class="o">.</span><span class="n">index</span><span class="o">.</span><span class="n">get_loc</span><span class="p">(</span><span class="n">human_reg</span><span class="p">)</span> <span class="o">+</span> <span class="mi">1</span>
        <span class="n">ranks</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">rank</span><span class="p">)</span>

    <span class="k">return</span> <span class="n">np</span><span class="o">.</span><span class="n">mean</span><span class="p">(</span><span class="n">ranks</span><span class="p">)</span>

<div class="viewcode-block" id="calculate_rank"><a class="viewcode-back" href="../../../../api/dnn_api.html#pipeline.code.dnn.analyzer.calculate_rank">[docs]</a><span class="k">def</span><span class="w"> </span><span class="nf">calculate_rank</span><span class="p">(</span><span class="n">params</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">model</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span><span class="n">epoch</span><span class="o">=</span><span class="mi">0</span><span class="p">,</span><span class="n">model_dir</span><span class="o">=</span><span class="kc">None</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Calculate cross-species embedding correlation ranks between human and mouse regions.</span>

<span class="sd">    This function extracts embeddings for human and mouse transcriptional data using</span>
<span class="sd">    a trained neural network model, computes the correlation matrix between region-level</span>
<span class="sd">    embeddings, and calculates the average rank of prior homologous human regions for each</span>
<span class="sd">    mouse region.</span>

<span class="sd">    Parameters</span>
<span class="sd">    ----------</span>
<span class="sd">    params : object</span>
<span class="sd">        Configuration object containing paths to human/mouse transcriptional data and</span>
<span class="sd">        homology mapping file, typically accessed via `params.data_files`.</span>
<span class="sd">    model : torch.nn.Module</span>
<span class="sd">        Trained deep learning model used to extract embeddings from the input data.</span>
<span class="sd">    epoch : int, optional</span>
<span class="sd">        Epoch index used.</span>
<span class="sd">    model_dir : str, optional</span>
<span class="sd">        Directory path to save the embedding results.</span>

<span class="sd">    Returns</span>
<span class="sd">    -------</span>
<span class="sd">    float</span>
<span class="sd">        The average rank of prior homologous human regions for each mouse region in terms</span>
<span class="sd">        of correlation with the extracted embeddings. Lower values indicate better cross-species</span>
<span class="sd">        alignment.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    
    <span class="n">human_data</span><span class="p">,</span> <span class="n">human_regions</span> <span class="o">=</span> <span class="n">generate_average_data</span><span class="p">(</span><span class="n">params</span><span class="o">.</span><span class="n">data_files</span><span class="p">[</span><span class="s1">&#39;human_trans_path&#39;</span><span class="p">],</span><span class="n">params</span><span class="p">)</span>
    <span class="n">mouse_data</span><span class="p">,</span> <span class="n">mouse_regions</span> <span class="o">=</span> <span class="n">generate_average_data</span><span class="p">(</span><span class="n">params</span><span class="o">.</span><span class="n">data_files</span><span class="p">[</span><span class="s1">&#39;mouse_trans_path&#39;</span><span class="p">],</span><span class="n">params</span><span class="p">)</span>

    <span class="n">human_embeddings</span> <span class="o">=</span> <span class="n">extract_embeddings</span><span class="p">(</span><span class="n">human_data</span><span class="p">,</span> <span class="n">model</span><span class="o">.</span><span class="n">eval</span><span class="p">())</span>
    <span class="n">mouse_embeddings</span> <span class="o">=</span> <span class="n">extract_embeddings</span><span class="p">(</span><span class="n">mouse_data</span><span class="p">,</span> <span class="n">model</span><span class="o">.</span><span class="n">eval</span><span class="p">())</span>

    <span class="n">human_df</span> <span class="o">=</span> <span class="n">create_aligned_dataframe</span><span class="p">(</span><span class="n">human_embeddings</span><span class="p">,</span> <span class="n">human_regions</span><span class="p">)</span>
    <span class="n">mouse_df</span> <span class="o">=</span> <span class="n">create_aligned_dataframe</span><span class="p">(</span><span class="n">mouse_embeddings</span><span class="p">,</span> <span class="n">mouse_regions</span><span class="p">)</span>

    <span class="n">save_path</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">model_dir</span><span class="p">,</span><span class="s1">&#39;Results&#39;</span><span class="p">)</span>
    <span class="n">os</span><span class="o">.</span><span class="n">makedirs</span><span class="p">(</span><span class="n">save_path</span><span class="p">,</span><span class="n">exist_ok</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
    <span class="n">human_df</span><span class="o">.</span><span class="n">to_csv</span><span class="p">(</span><span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">save_path</span><span class="p">,</span><span class="s1">&#39;human_embedding_epoch</span><span class="si">{}</span><span class="s1">.csv&#39;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">epoch</span><span class="p">)))</span>
    <span class="n">mouse_df</span><span class="o">.</span><span class="n">to_csv</span><span class="p">(</span><span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">save_path</span><span class="p">,</span><span class="s1">&#39;mouse_embedding_epoch</span><span class="si">{}</span><span class="s1">.csv&#39;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">epoch</span><span class="p">)))</span>

    <span class="n">homology_map</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="n">params</span><span class="o">.</span><span class="n">data_files</span><span class="p">[</span><span class="s1">&#39;human_mouse_path&#39;</span><span class="p">])</span>
    <span class="n">correlation_matrix</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">corrcoef</span><span class="p">(</span><span class="n">human_df</span><span class="p">,</span> <span class="n">mouse_df</span><span class="p">)[</span><span class="nb">len</span><span class="p">(</span><span class="n">human_df</span><span class="p">):,</span> <span class="p">:</span><span class="nb">len</span><span class="p">(</span><span class="n">human_df</span><span class="p">)]</span>
    <span class="n">correlation_df</span><span class="o">=</span><span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span><span class="n">correlation_matrix</span><span class="o">.</span><span class="n">T</span><span class="p">,</span><span class="n">index</span><span class="o">=</span><span class="n">human_df</span><span class="o">.</span><span class="n">index</span><span class="p">,</span><span class="n">columns</span><span class="o">=</span><span class="n">mouse_df</span><span class="o">.</span><span class="n">index</span><span class="p">)</span>

    <span class="k">return</span> <span class="n">calculate_homology_rank</span><span class="p">(</span><span class="n">correlation_df</span><span class="p">,</span><span class="n">mouse_regions</span><span class="o">=</span><span class="n">homology_map</span><span class="p">[</span><span class="s1">&#39;mouse_region&#39;</span><span class="p">],</span><span class="n">human_regions</span><span class="o">=</span><span class="n">homology_map</span><span class="p">[</span><span class="s1">&#39;human_region&#39;</span><span class="p">])</span></div>


<div class="viewcode-block" id="transform_human_mouse"><a class="viewcode-back" href="../../../../api/dnn_api.html#pipeline.code.dnn.analyzer.transform_human_mouse">[docs]</a><span class="k">def</span><span class="w"> </span><span class="nf">transform_human_mouse</span><span class="p">(</span><span class="n">params</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Transform human and mouse data using trained embedding models.</span>

<span class="sd">    This function loads trained neural network models for a number of repeated experiments, </span>
<span class="sd">    transforms human and mouse transcriptional data into embeddings, and saves them as CSV files.</span>

<span class="sd">    Parameters</span>
<span class="sd">    ----------</span>
<span class="sd">    params : object</span>
<span class="sd">        Configuration object containing:</span>
<span class="sd">            - data_files: dictionary with paths to human/mouse data and model save directories</span>
<span class="sd">            - repeat_n: number of repeated models</span>
<span class="sd">            - model hyperparameters: input_units, hidden_units1/2/3, output_units</span>
<span class="sd">            - use_bestrank / use_bestvalid: flags for model selection</span>
<span class="sd">    Returns</span>
<span class="sd">    -------</span>
<span class="sd">    None</span>
<span class="sd">        Saves the generated embedding CSV files.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">logger</span> <span class="o">=</span> <span class="n">get_logger</span><span class="p">()</span>
    <span class="n">human_data</span><span class="p">,</span> <span class="n">human_regions</span> <span class="o">=</span> <span class="n">generate_average_data</span><span class="p">(</span><span class="n">params</span><span class="o">.</span><span class="n">data_files</span><span class="p">[</span><span class="s1">&#39;human_trans_path&#39;</span><span class="p">],</span><span class="n">params</span><span class="p">)</span>
    <span class="n">mouse_data</span><span class="p">,</span> <span class="n">mouse_regions</span> <span class="o">=</span> <span class="n">generate_average_data</span><span class="p">(</span><span class="n">params</span><span class="o">.</span><span class="n">data_files</span><span class="p">[</span><span class="s1">&#39;mouse_trans_path&#39;</span><span class="p">],</span><span class="n">params</span><span class="p">)</span>

    <span class="n">data_label</span> <span class="o">=</span> <span class="n">Path</span><span class="p">(</span><span class="n">params</span><span class="o">.</span><span class="n">data_files</span><span class="p">[</span><span class="s1">&#39;data_path&#39;</span><span class="p">])</span><span class="o">.</span><span class="n">stem</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s2">&quot;_&quot;</span><span class="p">)[</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span>

    <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">params</span><span class="o">.</span><span class="n">repeat_n</span><span class="p">):</span>

        <span class="n">net</span> <span class="o">=</span> <span class="n">ClassifierModule</span><span class="p">(</span><span class="n">input_units</span> <span class="o">=</span> <span class="n">params</span><span class="o">.</span><span class="n">input_units</span><span class="p">,</span>
                    <span class="n">output_units</span> <span class="o">=</span> <span class="n">params</span><span class="o">.</span><span class="n">output_units</span><span class="p">,</span>
                    <span class="n">hidden_units1</span> <span class="o">=</span> <span class="n">params</span><span class="o">.</span><span class="n">hidden_units1</span><span class="p">,</span>
                    <span class="n">hidden_units2</span> <span class="o">=</span> <span class="n">params</span><span class="o">.</span><span class="n">hidden_units2</span><span class="p">,</span>
                    <span class="n">hidden_units3</span> <span class="o">=</span> <span class="n">params</span><span class="o">.</span><span class="n">hidden_units3</span><span class="p">,</span>
                    <span class="n">mode</span><span class="o">=</span><span class="s1">&#39;Transform&#39;</span><span class="p">)</span><span class="o">.</span><span class="n">cuda</span><span class="p">()</span>
        
        <span class="n">model_dir</span> <span class="o">=</span> <span class="n">Path</span><span class="p">(</span><span class="n">params</span><span class="o">.</span><span class="n">data_files</span><span class="p">[</span><span class="s2">&quot;save_path&quot;</span><span class="p">])</span> <span class="o">/</span> <span class="sa">f</span><span class="s2">&quot;Data_</span><span class="si">{</span><span class="n">data_label</span><span class="si">}</span><span class="s2">/Repeat_</span><span class="si">{</span><span class="n">i</span><span class="si">}</span><span class="s2">&quot;</span>
        
        <span class="k">if</span> <span class="n">params</span><span class="o">.</span><span class="n">use_bestrank</span> <span class="o">==</span> <span class="kc">True</span><span class="p">:</span>
            <span class="n">modelpath</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">model_dir</span><span class="p">,</span><span class="s1">&#39;Best_rank_epoch.pth&#39;</span><span class="p">)</span>
        <span class="k">elif</span> <span class="n">params</span><span class="o">.</span><span class="n">use_bestvalid</span> <span class="o">==</span> <span class="kc">True</span><span class="p">:</span>
            <span class="n">modelpath</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">model_dir</span><span class="p">,</span><span class="s1">&#39;Best_validloss_epoch.pth&#39;</span><span class="p">)</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">modelpath</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">model_dir</span><span class="p">,</span><span class="s1">&#39;Last_epoch.pth&#39;</span><span class="p">)</span>

        <span class="k">if</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">exists</span><span class="p">(</span><span class="n">modelpath</span><span class="p">):</span>
            <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Transform model: </span><span class="si">{</span><span class="n">model_dir</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
            <span class="n">net</span><span class="o">.</span><span class="n">load_state_dict</span><span class="p">(</span><span class="n">torch</span><span class="o">.</span><span class="n">load</span><span class="p">(</span><span class="n">modelpath</span><span class="p">))</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">logger</span><span class="o">.</span><span class="n">warning</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Transform model: </span><span class="si">{</span><span class="n">modelpath</span><span class="si">}</span><span class="s2"> not exist&quot;</span><span class="p">)</span>
            <span class="k">continue</span>

        <span class="n">human_embeddings</span> <span class="o">=</span> <span class="n">extract_embeddings</span><span class="p">(</span><span class="n">human_data</span><span class="p">,</span> <span class="n">net</span><span class="o">.</span><span class="n">eval</span><span class="p">())</span>
        <span class="n">mouse_embeddings</span> <span class="o">=</span> <span class="n">extract_embeddings</span><span class="p">(</span><span class="n">mouse_data</span><span class="p">,</span> <span class="n">net</span><span class="o">.</span><span class="n">eval</span><span class="p">())</span>

        <span class="n">human_df</span> <span class="o">=</span> <span class="n">create_aligned_dataframe</span><span class="p">(</span><span class="n">human_embeddings</span><span class="p">,</span> <span class="n">human_regions</span><span class="p">)</span>
        <span class="n">mouse_df</span> <span class="o">=</span> <span class="n">create_aligned_dataframe</span><span class="p">(</span><span class="n">mouse_embeddings</span><span class="p">,</span> <span class="n">mouse_regions</span><span class="p">)</span>

        <span class="n">save_path</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">model_dir</span><span class="p">,</span><span class="s1">&#39;Results&#39;</span><span class="p">)</span>
        <span class="n">os</span><span class="o">.</span><span class="n">makedirs</span><span class="p">(</span><span class="n">save_path</span><span class="p">,</span><span class="n">exist_ok</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
        <span class="n">human_df</span><span class="o">.</span><span class="n">to_csv</span><span class="p">(</span><span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">save_path</span><span class="p">,</span><span class="s1">&#39;human_embedding.csv&#39;</span><span class="p">))</span>
        <span class="n">mouse_df</span><span class="o">.</span><span class="n">to_csv</span><span class="p">(</span><span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">save_path</span><span class="p">,</span><span class="s1">&#39;mouse_embedding.csv&#39;</span><span class="p">))</span></div>

<div class="viewcode-block" id="independent_test"><a class="viewcode-back" href="../../../../api/dnn_api.html#pipeline.code.dnn.analyzer.independent_test">[docs]</a><span class="k">def</span><span class="w"> </span><span class="nf">independent_test</span><span class="p">(</span><span class="n">params</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Evaluate the trained model on an independent test dataset.</span>

<span class="sd">    Loads the saved models from repeated training runs, runs inference on the test data,</span>
<span class="sd">    saves predictions and confusion matrix to CSV files.</span>

<span class="sd">    Parameters</span>
<span class="sd">    ----------</span>
<span class="sd">    params : object</span>

<span class="sd">        Configuration object with the following attributes:</span>
<span class="sd">        </span>
<span class="sd">        - data_files : dict</span>
<span class="sd">            Paths to independent test data and saved model weights.</span>
<span class="sd">            Expected keys include &#39;independent_test_path&#39;, &#39;independent_data_path&#39;, &#39;independent_s_path&#39;.</span>
<span class="sd">        - repeat_n : int</span>
<span class="sd">            Number of repeated training runs/models to evaluate.</span>
<span class="sd">        - input_units, hidden_units1, hidden_units2, hidden_units3, output_units : int</span>
<span class="sd">            Model architecture hyperparameters.</span>
<span class="sd">        - use_bestrank, use_bestvalid : bool</span>
<span class="sd">            Flags to select which checkpoint to load for evaluation.</span>

<span class="sd">    Returns</span>
<span class="sd">    -------</span>
<span class="sd">    None</span>
<span class="sd">        Saves prediction results and confusion matrices as CSV files.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">logger</span> <span class="o">=</span> <span class="n">get_logger</span><span class="p">()</span>
    <span class="n">X</span><span class="p">,</span> <span class="n">y</span><span class="p">,</span> <span class="n">labels</span><span class="p">,</span> <span class="n">_</span> <span class="p">,</span><span class="n">_</span> <span class="o">=</span> <span class="n">generate_matrix</span><span class="p">(</span><span class="n">params</span><span class="o">.</span><span class="n">data_files</span><span class="p">[</span><span class="s1">&#39;independent_test_path&#39;</span><span class="p">],</span> <span class="n">params</span><span class="p">)</span>
    <span class="n">test_loader</span> <span class="o">=</span> <span class="n">create_dataloader</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">concatenate</span><span class="p">((</span><span class="n">X</span><span class="p">,</span> <span class="n">y</span><span class="p">),</span> <span class="n">axis</span><span class="o">=</span><span class="mi">1</span><span class="p">),</span> <span class="mi">1</span><span class="p">,</span> <span class="n">shuffle</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">drop_last</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>

    <span class="n">data_label</span> <span class="o">=</span> <span class="n">Path</span><span class="p">(</span><span class="n">params</span><span class="o">.</span><span class="n">data_files</span><span class="p">[</span><span class="s1">&#39;independent_data_path&#39;</span><span class="p">])</span><span class="o">.</span><span class="n">stem</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s2">&quot;_&quot;</span><span class="p">)[</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span>
        
    <span class="k">for</span> <span class="n">repeat</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">params</span><span class="o">.</span><span class="n">repeat_n</span><span class="p">):</span>

        <span class="n">model_dir</span> <span class="o">=</span> <span class="n">Path</span><span class="p">(</span><span class="n">params</span><span class="o">.</span><span class="n">data_files</span><span class="p">[</span><span class="s2">&quot;independent_s_path&quot;</span><span class="p">])</span> <span class="o">/</span> <span class="sa">f</span><span class="s2">&quot;Data_</span><span class="si">{</span><span class="n">data_label</span><span class="si">}</span><span class="s2">/Repeat_</span><span class="si">{</span><span class="n">repeat</span><span class="si">}</span><span class="s2">&quot;</span>
        <span class="n">save_path</span> <span class="o">=</span> <span class="n">model_dir</span> <span class="o">/</span> <span class="s1">&#39;Results&#39;</span>
        <span class="n">model_path</span> <span class="o">=</span> <span class="n">model_dir</span> <span class="o">/</span> <span class="p">(</span><span class="s1">&#39;Best_rank_epoch.pth&#39;</span> <span class="k">if</span> <span class="n">params</span><span class="o">.</span><span class="n">use_bestrank</span> <span class="k">else</span> 
                                <span class="s1">&#39;Best_validloss_epoch.pth&#39;</span> <span class="k">if</span> <span class="n">params</span><span class="o">.</span><span class="n">use_bestvalid</span> <span class="k">else</span> 
                                <span class="s1">&#39;Last_epoch.pth&#39;</span><span class="p">)</span>

        <span class="k">if</span> <span class="ow">not</span> <span class="n">model_path</span><span class="o">.</span><span class="n">exists</span><span class="p">():</span>
            <span class="n">logger</span><span class="o">.</span><span class="n">warning</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Model not found: </span><span class="si">{</span><span class="n">model_path</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
            <span class="k">continue</span>
            
        <span class="n">model</span> <span class="o">=</span> <span class="n">ClassifierModule</span><span class="p">(</span>
            <span class="n">input_units</span><span class="o">=</span><span class="n">params</span><span class="o">.</span><span class="n">input_units</span><span class="p">,</span>
            <span class="n">output_units</span><span class="o">=</span><span class="n">params</span><span class="o">.</span><span class="n">output_units</span><span class="p">,</span>
            <span class="n">hidden_units1</span><span class="o">=</span><span class="n">params</span><span class="o">.</span><span class="n">hidden_units1</span><span class="p">,</span>
            <span class="n">hidden_units2</span><span class="o">=</span><span class="n">params</span><span class="o">.</span><span class="n">hidden_units2</span><span class="p">,</span>
            <span class="n">hidden_units3</span><span class="o">=</span><span class="n">params</span><span class="o">.</span><span class="n">hidden_units3</span><span class="p">,</span>
            <span class="n">mode</span><span class="o">=</span><span class="s1">&#39;Test&#39;</span>
        <span class="p">)</span><span class="o">.</span><span class="n">cuda</span><span class="p">()</span><span class="o">.</span><span class="n">eval</span><span class="p">()</span>
        <span class="n">model</span><span class="o">.</span><span class="n">load_state_dict</span><span class="p">(</span><span class="n">torch</span><span class="o">.</span><span class="n">load</span><span class="p">(</span><span class="n">model_path</span><span class="p">))</span>

        <span class="n">gt</span><span class="p">,</span> <span class="n">pred</span> <span class="o">=</span> <span class="p">[],</span> <span class="p">[]</span>
        <span class="k">with</span> <span class="n">torch</span><span class="o">.</span><span class="n">no_grad</span><span class="p">():</span>
            <span class="k">for</span> <span class="n">inputs</span><span class="p">,</span> <span class="n">targets</span> <span class="ow">in</span> <span class="n">test_loader</span><span class="p">:</span>
                <span class="n">outputs</span> <span class="o">=</span> <span class="n">model</span><span class="p">(</span><span class="n">inputs</span><span class="o">.</span><span class="n">cuda</span><span class="p">()</span><span class="o">.</span><span class="n">float</span><span class="p">())</span>
                <span class="n">pred</span><span class="o">.</span><span class="n">extend</span><span class="p">(</span><span class="n">torch</span><span class="o">.</span><span class="n">argmax</span><span class="p">(</span><span class="n">outputs</span><span class="o">.</span><span class="n">softmax</span><span class="p">(</span><span class="mi">1</span><span class="p">),</span> <span class="mi">1</span><span class="p">)</span><span class="o">.</span><span class="n">cpu</span><span class="p">()</span><span class="o">.</span><span class="n">numpy</span><span class="p">())</span>
                <span class="n">gt</span><span class="o">.</span><span class="n">extend</span><span class="p">(</span><span class="n">targets</span><span class="o">.</span><span class="n">cpu</span><span class="p">()</span><span class="o">.</span><span class="n">numpy</span><span class="p">())</span>

        <span class="n">save_path</span><span class="o">.</span><span class="n">mkdir</span><span class="p">(</span><span class="n">parents</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">exist_ok</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
        <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">({</span><span class="s1">&#39;gt&#39;</span><span class="p">:</span> <span class="n">gt</span><span class="p">,</span> <span class="s1">&#39;pred&#39;</span><span class="p">:</span> <span class="n">pred</span><span class="p">})</span><span class="o">.</span><span class="n">to_csv</span><span class="p">(</span><span class="n">save_path</span><span class="o">/</span><span class="s1">&#39;predictions.csv&#39;</span><span class="p">,</span> <span class="n">index</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>

        <span class="n">conf_matrix</span> <span class="o">=</span> <span class="n">confusion_matrix</span><span class="p">(</span><span class="n">gt</span><span class="p">,</span> <span class="n">pred</span><span class="p">)</span>

        <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span><span class="n">conf_matrix</span><span class="o">/</span><span class="n">conf_matrix</span><span class="o">.</span><span class="n">sum</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="n">keepdims</span><span class="o">=</span><span class="kc">True</span><span class="p">),</span>
                    <span class="n">index</span><span class="o">=</span><span class="n">np</span><span class="o">.</span><span class="n">unique</span><span class="p">(</span><span class="n">labels</span><span class="p">),</span>
                    <span class="n">columns</span><span class="o">=</span><span class="n">np</span><span class="o">.</span><span class="n">unique</span><span class="p">(</span><span class="n">labels</span><span class="p">))</span><span class="o">.</span><span class="n">to_csv</span><span class="p">(</span><span class="n">save_path</span><span class="o">/</span><span class="s1">&#39;confusion_matrix.csv&#39;</span><span class="p">)</span></div>

</pre></div>

           </div>
          </div>
          <footer>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright 2025, TransBrain.</p>
  </div>

  Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    provided by <a href="https://readthedocs.org">Read the Docs</a>.
   

</footer>
        </div>
      </div>
    </section>
  </div>
  <script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>