

<!DOCTYPE html>
<html class="writer-html5" lang="en">
<head>
  <meta charset="utf-8" /><meta name="viewport" content="width=device-width, initial-scale=1" />

  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>Transcriptional Data Process &mdash; TransBrain 0.1 documentation</title>
      <link rel="stylesheet" type="text/css" href="../_static/pygments.css?v=03e43079" />
      <link rel="stylesheet" type="text/css" href="../_static/css/theme.css?v=e59714d7" />

  
      <script src="../_static/jquery.js?v=5d32c60e"></script>
      <script src="../_static/_sphinx_javascript_frameworks_compat.js?v=2cd50e6c"></script>
      <script data-url_root="../" id="documentation_options" src="../_static/documentation_options.js?v=e031e9a9"></script>
      <script src="../_static/doctools.js?v=888ff710"></script>
      <script src="../_static/sphinx_highlight.js?v=4825356b"></script>
      <script crossorigin="anonymous" integrity="sha256-Ae2Vz/4ePdIu6ZyI/5ZGsYnb+m0JlOmKPjt6XZ9JJkA=" src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.4/require.min.js"></script>
    <script src="../_static/js/theme.js"></script>
    <link rel="index" title="Index" href="../genindex.html" />
    <link rel="search" title="Search" href="../search.html" />
    <link rel="next" title="Data Integration" href="data_integration_api.html" />
    <link rel="prev" title="Development Functions" href="utils.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >

          
          
          <a href="../index.html" class="icon icon-home">
            TransBrain
              <img src="../_static/logo_final.png" class="logo" alt="Logo"/>
          </a>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" aria-label="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <p class="caption" role="heading"><span class="caption-text">Setup &amp; Usage</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../installation.html">Installation</a><ul>
<li class="toctree-l2"><a class="reference internal" href="../installation.html#installation-options">Installation Options</a></li>
<li class="toctree-l2"><a class="reference internal" href="../installation.html#python-dependencies">Python Dependencies</a></li>
<li class="toctree-l2"><a class="reference internal" href="../installation.html#unit-test">Unit Test</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="../user.html">User Guide</a><ul>
<li class="toctree-l2"><a class="reference internal" href="../user.html#id2">Online Mapping</a></li>
<li class="toctree-l2"><a class="reference internal" href="../user.html#id3">General Use</a></li>
<li class="toctree-l2"><a class="reference internal" href="../user.html#id4">Advanced Use</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="../atlas.html">Supported Atlases</a><ul>
<li class="toctree-l2"><a class="reference internal" href="../atlas.html#human-brain-atlases">Human Brain Atlases</a></li>
<li class="toctree-l2"><a class="reference internal" href="../atlas.html#mouse-brain-atlas">Mouse Brain Atlas</a></li>
</ul>
</li>
</ul>
<p class="caption" role="heading"><span class="caption-text">Tutorials</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../methods/method_glossory.html">Method Glossory</a><ul>
<li class="toctree-l2"><a class="reference internal" href="../methods/method_glossory.html#datasets">Datasets</a></li>
<li class="toctree-l2"><a class="reference internal" href="../methods/method_glossory.html#methods-or-concepts">Methods or Concepts</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="../online_mapping.html">Online Mapping</a></li>
<li class="toctree-l1"><a class="reference internal" href="../tutorials/notebooks/basic_usage/basic_use_mouse_to_human.html">Basic use 1: Mouse to human</a><ul>
<li class="toctree-l2"><a class="reference internal" href="../tutorials/notebooks/basic_usage/basic_use_mouse_to_human.html#Step-1:-Prepare-input-data">Step 1: Prepare input data</a><ul>
<li class="toctree-l3"><a class="reference internal" href="../tutorials/notebooks/basic_usage/basic_use_mouse_to_human.html#Visualizing-mouse-phenotype-data">Visualizing mouse phenotype data</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="../tutorials/notebooks/basic_usage/basic_use_mouse_to_human.html#Step-2:-Initialization">Step 2: Initialization</a></li>
<li class="toctree-l2"><a class="reference internal" href="../tutorials/notebooks/basic_usage/basic_use_mouse_to_human.html#Step-3:-Mapping">Step 3: Mapping</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="../tutorials/notebooks/basic_usage/basic_use_human_to_mouse.html">Basic use 2: Human to mouse</a><ul>
<li class="toctree-l2"><a class="reference internal" href="../tutorials/notebooks/basic_usage/basic_use_human_to_mouse.html#Step-1:-Prepare-input-data">Step 1: Prepare input data</a><ul>
<li class="toctree-l3"><a class="reference internal" href="../tutorials/notebooks/basic_usage/basic_use_human_to_mouse.html#Visualizing-human-phenotype-data">Visualizing human phenotype data</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="../tutorials/notebooks/basic_usage/basic_use_human_to_mouse.html#Step-2:-Initialization">Step 2: Initialization</a></li>
<li class="toctree-l2"><a class="reference internal" href="../tutorials/notebooks/basic_usage/basic_use_human_to_mouse.html#Step-3:-Mapping">Step 3: Mapping</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="../tutorials/notebooks/translation/gradient_spectrum/fMRI_gradients.html">Case 1: Analyzing the conservation of resting-fMRI gradients</a><ul>
<li class="toctree-l2"><a class="reference internal" href="../tutorials/notebooks/translation/gradient_spectrum/fMRI_gradients.html#Introduction">Introduction</a></li>
<li class="toctree-l2"><a class="reference internal" href="../tutorials/notebooks/translation/gradient_spectrum/fMRI_gradients.html#Mapping-mouse-gradient-to-human">Mapping mouse gradient to human</a></li>
<li class="toctree-l2"><a class="reference internal" href="../tutorials/notebooks/translation/gradient_spectrum/fMRI_gradients.html#Compare-with-human-Gradient">Compare with human Gradient</a></li>
<li class="toctree-l2"><a class="reference internal" href="../tutorials/notebooks/translation/gradient_spectrum/fMRI_gradients.html#Cross-species-correlation">Cross-species correlation</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="../tutorials/notebooks/translation/optogenetic_annotation/optogenetic_circuits.html">Case 2: Annotating optogenetic circuits</a><ul>
<li class="toctree-l2"><a class="reference internal" href="../tutorials/notebooks/translation/optogenetic_annotation/optogenetic_circuits.html#Introduction">Introduction</a><ul>
<li class="toctree-l3"><a class="reference internal" href="../tutorials/notebooks/translation/optogenetic_annotation/optogenetic_circuits.html#Data">Data</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="../tutorials/notebooks/translation/optogenetic_annotation/optogenetic_circuits.html#Translating-mouse-stimulated-pattern-to-human">Translating mouse stimulated pattern to human</a></li>
<li class="toctree-l2"><a class="reference internal" href="../tutorials/notebooks/translation/optogenetic_annotation/optogenetic_circuits.html#Annotating-using-NeuroSynth.">Annotating using NeuroSynth.</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="../tutorials/notebooks/translation/autism_mutation/gene_mutations.html">Case 3: Linking gene mutations to imaging phenotype deviations</a><ul>
<li class="toctree-l2"><a class="reference internal" href="../tutorials/notebooks/translation/autism_mutation/gene_mutations.html#Introduction">Introduction</a><ul>
<li class="toctree-l3"><a class="reference internal" href="../tutorials/notebooks/translation/autism_mutation/gene_mutations.html#Data">Data</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="../tutorials/notebooks/translation/autism_mutation/gene_mutations.html#Translating-mouse-mutation-patterns-to-human">Translating mouse mutation patterns to human</a></li>
<li class="toctree-l2"><a class="reference internal" href="../tutorials/notebooks/translation/autism_mutation/gene_mutations.html#Calculate-individual-gene-risk-score">Calculate individual gene risk score</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="../building.html">Advanced Use</a><ul>
<li class="toctree-l2"><a class="reference internal" href="../pipeline/run_pipeline/integration.html">Transcriptional Data Integration</a><ul>
<li class="toctree-l3"><a class="reference internal" href="../pipeline/run_pipeline/integration.html#Introduction">Introduction</a><ul>
<li class="toctree-l4"><a class="reference internal" href="../pipeline/run_pipeline/integration.html#AHBA-data">AHBA data</a></li>
<li class="toctree-l4"><a class="reference internal" href="../pipeline/run_pipeline/integration.html#Human-single-nucleus-data">Human single-nucleus data</a></li>
<li class="toctree-l4"><a class="reference internal" href="../pipeline/run_pipeline/integration.html#Mouse-spatial-transcriptomic-data">Mouse spatial transcriptomic data</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="../pipeline/run_pipeline/integration.html#Selection-of-cortical-stable-genes">Selection of cortical stable genes</a></li>
<li class="toctree-l3"><a class="reference internal" href="../pipeline/run_pipeline/integration.html#Selection-of-subcortical-stable-genes">Selection of subcortical stable genes</a></li>
<li class="toctree-l3"><a class="reference internal" href="../pipeline/run_pipeline/integration.html#Selection-of-all-stable-genes">Selection of all stable genes</a></li>
<li class="toctree-l3"><a class="reference internal" href="../pipeline/run_pipeline/integration.html#Thresholds-of-cortex">Thresholds of cortex</a></li>
<li class="toctree-l3"><a class="reference internal" href="../pipeline/run_pipeline/integration.html#Thresholds-of-subcortex">Thresholds of subcortex</a></li>
<li class="toctree-l3"><a class="reference internal" href="../pipeline/run_pipeline/integration.html#Thresholds-of-all">Thresholds of all</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="../pipeline/run_pipeline/dnn.html">DNN Model Training</a><ul>
<li class="toctree-l3"><a class="reference internal" href="../pipeline/run_pipeline/dnn.html#Introduction">Introduction</a><ul>
<li class="toctree-l4"><a class="reference internal" href="../pipeline/run_pipeline/dnn.html#Model-architecture">Model architecture</a></li>
<li class="toctree-l4"><a class="reference internal" href="../pipeline/run_pipeline/dnn.html#Region-specific-embeddings">Region-specific embeddings</a></li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="../pipeline/run_pipeline/graph_embedding.html">Graph-based Random Walk</a><ul>
<li class="toctree-l3"><a class="reference internal" href="../pipeline/run_pipeline/graph_embedding.html#Introduction">Introduction</a><ul>
<li class="toctree-l4"><a class="reference internal" href="../pipeline/run_pipeline/graph_embedding.html#Human-diffusion-based-tractography">Human diffusion-based tractography</a></li>
<li class="toctree-l4"><a class="reference internal" href="../pipeline/run_pipeline/graph_embedding.html#Mouse-Tracer">Mouse Tracer</a></li>
<li class="toctree-l4"><a class="reference internal" href="../pipeline/run_pipeline/graph_embedding.html#Embedding-space-generated-from-Human-Mouse-graph">Embedding space generated from Human-Mouse graph</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="../pipeline/run_pipeline/graph_embedding.html#Define-human-and-mouse-ROIs">Define human and mouse ROIs</a></li>
<li class="toctree-l3"><a class="reference internal" href="../pipeline/run_pipeline/graph_embedding.html#Load-anatomical-hierachy-constraint-information">Load anatomical hierachy constraint information</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="../methods/mapping.html">Bidirectional Mapping</a></li>
</ul>
</li>
</ul>
<p class="caption" role="heading"><span class="caption-text">API</span></p>
<ul class="current">
<li class="toctree-l1"><a class="reference internal" href="external_api.html">API Reference</a><ul>
<li class="toctree-l2"><a class="reference internal" href="fetch_api.html">Fetching Atlases</a><ul>
<li class="toctree-l3"><a class="reference internal" href="fetch_api.html#transbrain.atlas.fetch_human_atlas"><code class="docutils literal notranslate"><span class="pre">fetch_human_atlas()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="fetch_api.html#transbrain.atlas.fetch_mouse_atlas"><code class="docutils literal notranslate"><span class="pre">fetch_mouse_atlas()</span></code></a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="base_api.html">Base Functions</a><ul>
<li class="toctree-l3"><a class="reference internal" href="base_api.html#transbrain.base.get_region_phenotypes"><code class="docutils literal notranslate"><span class="pre">get_region_phenotypes()</span></code></a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="mapping_api.html">Mapping Functions</a><ul>
<li class="toctree-l3"><a class="reference internal" href="mapping_api.html#transbrain.trans.SpeciesTrans"><code class="docutils literal notranslate"><span class="pre">SpeciesTrans</span></code></a><ul>
<li class="toctree-l4"><a class="reference internal" href="mapping_api.html#transbrain.trans.SpeciesTrans.atlas_type"><code class="docutils literal notranslate"><span class="pre">SpeciesTrans.atlas_type</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="mapping_api.html#transbrain.trans.SpeciesTrans.regions"><code class="docutils literal notranslate"><span class="pre">SpeciesTrans.regions</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="mapping_api.html#transbrain.trans.SpeciesTrans.embeddings"><code class="docutils literal notranslate"><span class="pre">SpeciesTrans.embeddings</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="mapping_api.html#transbrain.trans.SpeciesTrans._load_embeddings"><code class="docutils literal notranslate"><span class="pre">SpeciesTrans._load_embeddings()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="mapping_api.html#transbrain.trans.SpeciesTrans._dual_mapping"><code class="docutils literal notranslate"><span class="pre">SpeciesTrans._dual_mapping()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="mapping_api.html#transbrain.trans.SpeciesTrans._translate"><code class="docutils literal notranslate"><span class="pre">SpeciesTrans._translate()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="mapping_api.html#transbrain.trans.SpeciesTrans.mouse_to_human"><code class="docutils literal notranslate"><span class="pre">SpeciesTrans.mouse_to_human()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="mapping_api.html#transbrain.trans.SpeciesTrans.human_to_mouse"><code class="docutils literal notranslate"><span class="pre">SpeciesTrans.human_to_mouse()</span></code></a></li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="vis_api.html">Visualization Functions</a><ul>
<li class="toctree-l3"><a class="reference internal" href="vis_api.html#transbrain.vis.map_phenotype_to_nifti"><code class="docutils literal notranslate"><span class="pre">map_phenotype_to_nifti()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="vis_api.html#transbrain.vis.plot_mouse_phenotype"><code class="docutils literal notranslate"><span class="pre">plot_mouse_phenotype()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="vis_api.html#transbrain.vis.plot_human_phenotype"><code class="docutils literal notranslate"><span class="pre">plot_human_phenotype()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="vis_api.html#transbrain.surf.plot_surface_phenotype"><code class="docutils literal notranslate"><span class="pre">plot_surface_phenotype()</span></code></a></li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l1 current"><a class="reference internal" href="utils.html">Development Functions</a><ul class="current">
<li class="toctree-l2 current"><a class="current reference internal" href="#">Transcriptional Data Process</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#pipeline.code.dpreprocess.sn_.Preprocess_Sn"><code class="docutils literal notranslate"><span class="pre">Preprocess_Sn</span></code></a><ul>
<li class="toctree-l4"><a class="reference internal" href="#pipeline.code.dpreprocess.sn_.Preprocess_Sn.dataset_qc"><code class="docutils literal notranslate"><span class="pre">Preprocess_Sn.dataset_qc()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#pipeline.code.dpreprocess.sn_.Preprocess_Sn.dataset_normalized"><code class="docutils literal notranslate"><span class="pre">Preprocess_Sn.dataset_normalized()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#pipeline.code.dpreprocess.sn_.Preprocess_Sn.dataset_common_genes"><code class="docutils literal notranslate"><span class="pre">Preprocess_Sn.dataset_common_genes()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#pipeline.code.dpreprocess.sn_.Preprocess_Sn.dataset_batch_correct"><code class="docutils literal notranslate"><span class="pre">Preprocess_Sn.dataset_batch_correct()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#pipeline.code.dpreprocess.sn_.Preprocess_Sn.dataset_smooth"><code class="docutils literal notranslate"><span class="pre">Preprocess_Sn.dataset_smooth()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#pipeline.code.dpreprocess.sn_.Preprocess_Sn.dataset_scaled"><code class="docutils literal notranslate"><span class="pre">Preprocess_Sn.dataset_scaled()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#pipeline.code.dpreprocess.sn_.Preprocess_Sn.preprocess_pipeline"><code class="docutils literal notranslate"><span class="pre">Preprocess_Sn.preprocess_pipeline()</span></code></a></li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="data_integration_api.html">Data Integration</a><ul>
<li class="toctree-l3"><a class="reference internal" href="data_integration_api.html#pipeline.code.dintegrate.utils.stable_gene_filter"><code class="docutils literal notranslate"><span class="pre">stable_gene_filter()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="data_integration_api.html#pipeline.code.dintegrate.utils.calculate_thresholds"><code class="docutils literal notranslate"><span class="pre">calculate_thresholds()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="data_integration_api.html#pipeline.code.dintegrate.main.DataIntegrator"><code class="docutils literal notranslate"><span class="pre">DataIntegrator</span></code></a><ul>
<li class="toctree-l4"><a class="reference internal" href="data_integration_api.html#pipeline.code.dintegrate.main.DataIntegrator.__init__"><code class="docutils literal notranslate"><span class="pre">DataIntegrator.__init__()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="data_integration_api.html#pipeline.code.dintegrate.main.DataIntegrator.run"><code class="docutils literal notranslate"><span class="pre">DataIntegrator.run()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="data_integration_api.html#pipeline.code.dintegrate.main.DataIntegrator.pool_construction"><code class="docutils literal notranslate"><span class="pre">DataIntegrator.pool_construction()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="data_integration_api.html#pipeline.code.dintegrate.main.DataIntegrator.clean_pool"><code class="docutils literal notranslate"><span class="pre">DataIntegrator.clean_pool()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="data_integration_api.html#pipeline.code.dintegrate.main.DataIntegrator.weighted_average"><code class="docutils literal notranslate"><span class="pre">DataIntegrator.weighted_average()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="data_integration_api.html#pipeline.code.dintegrate.main.DataIntegrator.downsample_data"><code class="docutils literal notranslate"><span class="pre">DataIntegrator.downsample_data()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="data_integration_api.html#pipeline.code.dintegrate.main.DataIntegrator.generate_integrated_anndata"><code class="docutils literal notranslate"><span class="pre">DataIntegrator.generate_integrated_anndata()</span></code></a></li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="dnn_api.html">DNN</a><ul>
<li class="toctree-l3"><a class="reference internal" href="dnn_api.html#pipeline.code.dnn.analyzer.calculate_rank"><code class="docutils literal notranslate"><span class="pre">calculate_rank()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="dnn_api.html#pipeline.code.dnn.analyzer.transform_human_mouse"><code class="docutils literal notranslate"><span class="pre">transform_human_mouse()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="dnn_api.html#pipeline.code.dnn.analyzer.independent_test"><code class="docutils literal notranslate"><span class="pre">independent_test()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="dnn_api.html#pipeline.code.dnn.pipeline.DnnPipeline"><code class="docutils literal notranslate"><span class="pre">DnnPipeline</span></code></a><ul>
<li class="toctree-l4"><a class="reference internal" href="dnn_api.html#pipeline.code.dnn.pipeline.DnnPipeline.cfg"><code class="docutils literal notranslate"><span class="pre">DnnPipeline.cfg</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="dnn_api.html#pipeline.code.dnn.pipeline.DnnPipeline.logger"><code class="docutils literal notranslate"><span class="pre">DnnPipeline.logger</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="dnn_api.html#pipeline.code.dnn.pipeline.DnnPipeline.execute"><code class="docutils literal notranslate"><span class="pre">DnnPipeline.execute()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="dnn_api.html#pipeline.code.dnn.pipeline.DnnPipeline.train"><code class="docutils literal notranslate"><span class="pre">DnnPipeline.train()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="dnn_api.html#pipeline.code.dnn.pipeline.DnnPipeline.trans"><code class="docutils literal notranslate"><span class="pre">DnnPipeline.trans()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="dnn_api.html#pipeline.code.dnn.pipeline.DnnPipeline.test"><code class="docutils literal notranslate"><span class="pre">DnnPipeline.test()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="dnn_api.html#pipeline.code.dnn.pipeline.DnnPipeline._process_single_dataset"><code class="docutils literal notranslate"><span class="pre">DnnPipeline._process_single_dataset()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="dnn_api.html#pipeline.code.dnn.pipeline.DnnPipeline._train_single_model"><code class="docutils literal notranslate"><span class="pre">DnnPipeline._train_single_model()</span></code></a></li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="graph_walk_api.html">Graph Walk</a><ul>
<li class="toctree-l3"><a class="reference internal" href="graph_walk_api.html#pipeline.code.graph_walk.run.generate_embeddings"><code class="docutils literal notranslate"><span class="pre">generate_embeddings()</span></code></a></li>
</ul>
</li>
</ul>
</li>
</ul>
<p class="caption" role="heading"><span class="caption-text">Other</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../faq.html">Frequently Asked Questions</a></li>
<li class="toctree-l1"><a class="reference internal" href="../citing.html">Citing TransBrain</a></li>
</ul>

        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../index.html">TransBrain</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="../index.html" class="icon icon-home" aria-label="Home"></a></li>
          <li class="breadcrumb-item"><a href="utils.html">Development Functions</a></li>
      <li class="breadcrumb-item active">Transcriptional Data Process</li>
      <li class="wy-breadcrumbs-aside">
            <a href="../_sources/api/data_process_api.rst.txt" rel="nofollow"> View page source</a>
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  <section id="transcriptional-data-process">
<h1>Transcriptional Data Process<a class="headerlink" href="#transcriptional-data-process" title="Permalink to this heading"></a></h1>
<dl class="py function">
<dt class="sig sig-object py">
<span class="sig-prename descclassname"><span class="pre">pipeline.code.dpreprocess.ahba_.</span></span><span class="sig-name descname"><span class="pre">extract_AHBA_data</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">atlas_path</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">atlas_info_path</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">lr_mirror</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'bidirectional'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">gene_norm</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'srs'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">sample_norm</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'srs'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">return_report</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">return_counts</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">return_donors</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">norm_matched</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ibf_threshold</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">region_agg</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pipeline/code/dpreprocess/ahba_.html#extract_AHBA_data"><span class="viewcode-link"><span class="pre">[source]</span></span></a></dt>
<dd><p>Extracts Allen Human Brain Atlas (AHBA) gene expression data based on a given brain atlas.
For more information, please ref <a class="reference external" href="https://abagen.readthedocs.io/en/stable/generated/abagen.get_expression_data.html">https://abagen.readthedocs.io/en/stable/generated/abagen.get_expression_data.html</a></p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>atlas_path</strong> (<em>str</em>) – Path to the human brain atlas NIfTI file.</p></li>
<li><p><strong>atlas_info_path</strong> (<em>str</em>) – Path to a CSV file containing atlas region information with columns:
‘Anatomical Name’ and ‘Atlas Index’.</p></li>
<li><p><strong>lr_mirror</strong> (<em>{'left'</em><em>, </em><em>'right'</em><em>, </em><em>'bidirectional'}</em><em>, </em><em>optional</em>) – How to mirror samples across hemispheres. Default is ‘bidirectional’.</p></li>
<li><p><strong>gene_norm</strong> (<em>str</em><em>, </em><em>optional</em>) – Method by which to normalize microarray expression values for each donor. Default is ‘srs’.</p></li>
<li><p><strong>sample_norm</strong> (<em>str</em><em>, </em><em>optional</em>) – Method by which to normalize microarray expression values for each sample. Default is ‘srs’.</p></li>
<li><p><strong>return_report</strong> (<em>bool</em><em>, </em><em>optional</em>) – Whether to return a string containing longform text describing the processing procedures used to generate the expression DataFrames returned by this function. Default is True.</p></li>
<li><p><strong>return_counts</strong> (<em>bool</em><em>, </em><em>optional</em>) – Whether to return dataframe containing information on how many samples were assigned to each parcel in atlas for each donor. Default is True.</p></li>
<li><p><strong>return_donors</strong> (<em>bool</em><em>, </em><em>optional</em>) – Whether to return donor-level expression arrays instead of aggregating expression across donors with provided agg_metric. Default is True.</p></li>
<li><p><strong>norm_matched</strong> (<em>bool</em><em>, </em><em>optional</em>) – Whether to perform gene normalization (gene_norm) across only those samples matched to regions in atlas instead of all available samples. Default is False.</p></li>
<li><p><strong>ibf_threshold</strong> (<em>float</em><em>, </em><em>optional</em>) – Intensity-based filtering threshold. Default is 0.</p></li>
<li><p><strong>region_agg</strong> (<em>str</em><em> or </em><em>None</em><em>, </em><em>optional</em>) – Mechanism by which to reduce sample-level expression data into region-level expression.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>If <cite>return_donors=True</cite>, returns a dictionary where keys are donor IDs and values are gene expression DataFrames (regions × genes).
If <cite>return_donors=False</cite>, returns a single aggregated DataFrame.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>dict[str, pandas.DataFrame] or pandas.DataFrame</p>
</dd>
</dl>
<p class="rubric">Examples</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">expr_data</span> <span class="o">=</span> <span class="n">extract_AHBA_data</span><span class="p">(</span>
<span class="gp">... </span>    <span class="s1">&#39;/path/to/atlas.nii.gz&#39;</span><span class="p">,</span>
<span class="gp">... </span>    <span class="s1">&#39;/path/to/atlas_info.csv&#39;</span>
<span class="gp">... </span><span class="p">)</span>
</pre></div>
</div>
</dd></dl>

<hr class="docutils" />
<dl class="py class">
<dt class="sig sig-object py" id="pipeline.code.dpreprocess.sn_.Preprocess_Sn">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">pipeline.code.dpreprocess.sn_.</span></span><span class="sig-name descname"><span class="pre">Preprocess_Sn</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">adata</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">min_genes</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">200</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">min_cells</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">3</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">total_UMIs</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">800</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">log_base</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">2</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">target_sum</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">10000.0</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">exclude_highly_expressed</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">max_fraction</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0.05</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">n_jobs</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">60</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">regress_out</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">combat</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">num_samples</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">100</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">max_value</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">10</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pipeline/code/dpreprocess/sn_.html#Preprocess_Sn"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pipeline.code.dpreprocess.sn_.Preprocess_Sn" title="Permalink to this definition"></a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">object</span></code></p>
<p>Preprocessing pipeline for single-nucleus RNA-seq data based on the Scanpy framework.</p>
<p>Implements quality control, normalization, batch correction, gene selection,
intra-regional smoothing, and scaling, with optional saving of the processed data.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>adata</strong> (<em>AnnData</em><em>, </em><em>optional</em>) – Initial AnnData object containing expression matrix.</p></li>
<li><p><strong>min_genes</strong> (<em>int</em><em>, </em><em>optional</em>) – Minimum number of genes expressed per cell (default is 200).</p></li>
<li><p><strong>min_cells</strong> (<em>int</em><em>, </em><em>optional</em>) – Minimum number of cells a gene must be expressed in (default is 3).</p></li>
<li><p><strong>total_UMIs</strong> (<em>int</em><em>, </em><em>optional</em>) – Minimum total UMI count required per cell (default is 800).</p></li>
<li><p><strong>log_base</strong> (<em>float</em><em>, </em><em>optional</em>) – Base of logarithm used in log transformation (default is 2).</p></li>
<li><p><strong>target_sum</strong> (<em>float</em><em>, </em><em>optional</em>) – Target total count after normalization (default is 1e4).</p></li>
<li><p><strong>exclude_highly_expressed</strong> (<em>bool</em><em>, </em><em>optional</em>) – Whether to exclude highly expressed genes during normalization (default is False).</p></li>
<li><p><strong>max_fraction</strong> (<em>float</em><em>, </em><em>optional</em>) – Maximum expression fraction to define ‘highly expressed’ genes (default is 0.05).</p></li>
<li><p><strong>n_jobs</strong> (<em>int</em><em>, </em><em>optional</em>) – Number of threads for parallel computation (default is 60).</p></li>
<li><p><strong>regress_out</strong> (<em>bool</em><em>, </em><em>optional</em>) – Whether to regress out technical covariates (e.g., total UMIs) (default is False).</p></li>
<li><p><strong>combat</strong> (<em>bool</em><em>, </em><em>optional</em>) – Whether to perform batch effect correction using ComBat (default is False).</p></li>
<li><p><strong>num_samples</strong> (<em>int</em><em>, </em><em>optional</em>) – Number of resampling iterations for intra-region smoothing (default is None).</p></li>
<li><p><strong>max_value</strong> (<em>float</em><em>, </em><em>optional</em>) – Maximum value for scaling (clipping) of gene expression (default is 10).</p></li>
</ul>
</dd>
</dl>
<dl class="py method">
<dt class="sig sig-object py" id="pipeline.code.dpreprocess.sn_.Preprocess_Sn.dataset_qc">
<span class="sig-name descname"><span class="pre">dataset_qc</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pipeline/code/dpreprocess/sn_.html#Preprocess_Sn.dataset_qc"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pipeline.code.dpreprocess.sn_.Preprocess_Sn.dataset_qc" title="Permalink to this definition"></a></dt>
<dd><p>Perform quality control filtering on the dataset.</p>
<p>This function filters cells and genes based on quality metrics:</p>
<blockquote>
<div><ul class="simple">
<li><p>Calculates total UMIs per cell and adds it to <cite>adata.obs</cite>.</p></li>
<li><p>Filters out cells with fewer than <cite>min_genes</cite> expressed genes.</p></li>
<li><p>Filters out genes expressed in fewer than <cite>min_cells</cite> cells.</p></li>
<li><p>Removes cells with total UMIs below a threshold <cite>total_UMIs</cite>.</p></li>
<li><p>Removes genes with zero expression across all remaining cells.</p></li>
</ul>
</div></blockquote>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>The filtered AnnData object.</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p>AnnData</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pipeline.code.dpreprocess.sn_.Preprocess_Sn.dataset_normalized">
<span class="sig-name descname"><span class="pre">dataset_normalized</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pipeline/code/dpreprocess/sn_.html#Preprocess_Sn.dataset_normalized"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pipeline.code.dpreprocess.sn_.Preprocess_Sn.dataset_normalized" title="Permalink to this definition"></a></dt>
<dd><p>Normalize the single-nucleus RNA-seq dataset.</p>
<p>This method performs total-count normalization to scale counts per cell
to a common target sum, optionally excluding highly expressed genes,
followed by log-transformation with a specified logarithm base.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>The normalized AnnData object.</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p>AnnData</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pipeline.code.dpreprocess.sn_.Preprocess_Sn.dataset_common_genes">
<span class="sig-name descname"><span class="pre">dataset_common_genes</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">gene_list</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pipeline/code/dpreprocess/sn_.html#Preprocess_Sn.dataset_common_genes"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pipeline.code.dpreprocess.sn_.Preprocess_Sn.dataset_common_genes" title="Permalink to this definition"></a></dt>
<dd><p>Filter the dataset to keep only genes common to the provided gene list.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>gene_list</strong> (<em>list</em><em>, </em><em>optional</em>) – List of gene names to intersect with the dataset’s genes. If None,
no filtering is applied (default is None).</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>The filtered AnnData object containing only the common genes.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>AnnData</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pipeline.code.dpreprocess.sn_.Preprocess_Sn.dataset_batch_correct">
<span class="sig-name descname"><span class="pre">dataset_batch_correct</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pipeline/code/dpreprocess/sn_.html#Preprocess_Sn.dataset_batch_correct"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pipeline.code.dpreprocess.sn_.Preprocess_Sn.dataset_batch_correct" title="Permalink to this definition"></a></dt>
<dd><p>Perform batch effect correction.
This includes optional regression of total UMIs and batch correction using Combat.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>The AnnData object after batch correction.</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p>AnnData</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pipeline.code.dpreprocess.sn_.Preprocess_Sn.dataset_smooth">
<span class="sig-name descname"><span class="pre">dataset_smooth</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pipeline/code/dpreprocess/sn_.html#Preprocess_Sn.dataset_smooth"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pipeline.code.dpreprocess.sn_.Preprocess_Sn.dataset_smooth" title="Permalink to this definition"></a></dt>
<dd><p>Perform intra-regional gene expression smoothing.</p>
<p>For each region, perform smoothing by averaging gene expression over random samples of cells within that region.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>The smoothed AnnData object.</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p>AnnData</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pipeline.code.dpreprocess.sn_.Preprocess_Sn.dataset_scaled">
<span class="sig-name descname"><span class="pre">dataset_scaled</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">cortical_regions</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">subcortical_regions</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">scale_type</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'split'</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pipeline/code/dpreprocess/sn_.html#Preprocess_Sn.dataset_scaled"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pipeline.code.dpreprocess.sn_.Preprocess_Sn.dataset_scaled" title="Permalink to this definition"></a></dt>
<dd><p>Scale gene expression data, optionally splitting into cortical and subcortical regions.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>cortical_regions</strong> (<em>list</em><em>, </em><em>optional</em>) – List of cortical region names to subset and scale separately.</p></li>
<li><p><strong>subcortical_regions</strong> (<em>list</em><em>, </em><em>optional</em>) – List of subcortical region names to subset and scale separately.</p></li>
<li><p><strong>scale_type</strong> (<em>{'split'</em><em>, </em><em>'global'}</em><em>, </em><em>optional</em>) – Scaling method to apply. ‘split’ scales cortical and subcortical regions separately,
‘global’ scales the whole dataset at once. Default is ‘split’.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>The scaled AnnData object.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>AnnData</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pipeline.code.dpreprocess.sn_.Preprocess_Sn.preprocess_pipeline">
<span class="sig-name descname"><span class="pre">preprocess_pipeline</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">adata</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">min_genes</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">200</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">min_cells</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">3</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">total_UMIs</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">800</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">log_base</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">2</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">target_sum</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">10000.0</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">exclude_highly_expressed</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">max_fraction</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0.05</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">n_jobs</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">60</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">regress_out</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">combat</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">num_samples</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">100</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">max_value</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">10</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">dataset_path</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">gene_list</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">cortical_regions</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">subcortical_regions</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">scale_type</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'split'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">save_path</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">steps</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pipeline/code/dpreprocess/sn_.html#Preprocess_Sn.preprocess_pipeline"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pipeline.code.dpreprocess.sn_.Preprocess_Sn.preprocess_pipeline" title="Permalink to this definition"></a></dt>
<dd><p>Preprocess single-nucleus RNA-seq data with flexible pipeline steps.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>adata</strong> (<em>AnnData</em><em>, </em><em>optional</em>) – Input AnnData object (.h5ad) containing single-nucleus data.
If None, data will be loaded from <cite>dataset_path</cite>.</p></li>
<li><p><strong>min_genes</strong> (<em>int</em><em>, </em><em>default=200</em>) – Minimum number of genes required per cell. Cells with fewer genes will be filtered out.</p></li>
<li><p><strong>min_cells</strong> (<em>int</em><em>, </em><em>default=3</em>) – Minimum number of cells in which a gene must be detected.
Genes detected in fewer cells will be removed.</p></li>
<li><p><strong>total_UMIs</strong> (<em>int</em><em>, </em><em>default=800</em>) – Minimum total UMI counts per cell. Cells below this threshold will be filtered out.</p></li>
<li><p><strong>log_base</strong> (<em>int</em><em> or </em><em>float</em><em>, </em><em>default=2</em>) – Base of logarithm used for log transformation.</p></li>
<li><p><strong>target_sum</strong> (<em>float</em><em>, </em><em>default=1e4</em>) – Target sum for count normalization. After normalization, each cell’s counts sum to this value.</p></li>
<li><p><strong>exclude_highly_expressed</strong> (<em>bool</em><em>, </em><em>default=True</em>) – Whether to exclude highly expressed genes during normalization to reduce technical artifacts.</p></li>
<li><p><strong>max_fraction</strong> (<em>float</em><em>, </em><em>default=0.05</em>) – Maximum fraction of counts that can come from a single gene to consider it for exclusion.</p></li>
<li><p><strong>n_jobs</strong> (<em>int</em><em>, </em><em>default=60</em>) – Number of parallel jobs for computation.</p></li>
<li><p><strong>regress_out</strong> (<em>bool</em><em>, </em><em>default=True</em>) – Whether to regress out technical covariates (e.g., total counts).</p></li>
<li><p><strong>combat</strong> (<em>bool</em><em>, </em><em>default=False</em>) – Whether to apply ComBat batch correction.</p></li>
<li><p><strong>num_samples</strong> (<em>int</em><em>, </em><em>default=100</em>) – Number of samples for downsampling/bootstrap during smoothing.</p></li>
<li><p><strong>max_value</strong> (<em>float</em><em>, </em><em>default=10</em>) – Maximum clip threshold for transformed values to avoid extreme outliers.</p></li>
<li><p><strong>dataset_path</strong> (<em>str</em><em>, </em><em>optional</em>) – Path to the single-nucleus dataset in .h5ad format, used if <cite>adata</cite> is None.</p></li>
<li><p><strong>gene_list</strong> (<em>list</em><em> of </em><em>str</em><em>, </em><em>optional</em>) – List of common genes to filter on.</p></li>
<li><p><strong>cortical_regions</strong> (<em>list</em><em> of </em><em>str</em><em>, </em><em>optional</em>) – List of cortical region names in the dataset.</p></li>
<li><p><strong>subcortical_regions</strong> (<em>list</em><em> of </em><em>str</em><em>, </em><em>optional</em>) – List of subcortical region names in the dataset.</p></li>
<li><p><strong>scale_type</strong> (<em>{'split'</em><em>, </em><em>'all'}</em><em>, </em><em>default='split'</em>) – Normalization method for scaling the dataset. ‘split’ scales regions separately,
‘all’ scales all regions together.</p></li>
<li><p><strong>save_path</strong> (<em>str</em><em>, </em><em>optional</em>) – File path to save the processed data.</p></li>
<li><p><strong>steps</strong> (<em>dict</em><em> of </em><em>str to bool</em><em>, </em><em>optional</em>) – <dl class="simple">
<dt>Dictionary controlling execution of processing steps. Keys and defaults:</dt><dd><ul>
<li><p>’qc’ (bool): Quality control (default: True)</p></li>
<li><p>’normalize’ (bool): Data normalization (default: True)</p></li>
<li><p>’filter_genes’ (bool): Gene filtering (default: True)</p></li>
<li><p>’batch_correct’ (bool): Batch effect correction (default: True)</p></li>
<li><p>’smooth’ (bool): Data smoothing (default: True)</p></li>
<li><p>’scale’ (bool): Region-specific scaling (default: True)</p></li>
<li><p>’save’ (bool): Save processed data (default: True)</p></li>
</ul>
</dd>
</dl>
</p></li>
<li><p><strong>**kwargs</strong> – Additional optional parameters.</p></li>
</ul>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p>None</p>
</dd>
</dl>
</dd></dl>

</dd></dl>

</section>


           </div>
          </div>
          <footer><div class="rst-footer-buttons" role="navigation" aria-label="Footer">
        <a href="utils.html" class="btn btn-neutral float-left" title="Development Functions" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
        <a href="data_integration_api.html" class="btn btn-neutral float-right" title="Data Integration" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right" aria-hidden="true"></span></a>
    </div>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright 2025, TransBrain.</p>
  </div>

  Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    provided by <a href="https://readthedocs.org">Read the Docs</a>.
   

</footer>
        </div>
      </div>
    </section>
  </div>
  <script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>