

<!DOCTYPE html>
<html class="writer-html5" lang="en">
<head>
  <meta charset="utf-8" />
  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>pipeline.code.dpreprocess.ahba_ &mdash; TransBrain 0.1 documentation</title>
      <link rel="stylesheet" type="text/css" href="../../../../_static/pygments.css?v=03e43079" />
      <link rel="stylesheet" type="text/css" href="../../../../_static/css/theme.css?v=e59714d7" />

  
      <script src="../../../../_static/jquery.js?v=5d32c60e"></script>
      <script src="../../../../_static/_sphinx_javascript_frameworks_compat.js?v=2cd50e6c"></script>
      <script data-url_root="../../../../" id="documentation_options" src="../../../../_static/documentation_options.js?v=e031e9a9"></script>
      <script src="../../../../_static/doctools.js?v=888ff710"></script>
      <script src="../../../../_static/sphinx_highlight.js?v=4825356b"></script>
      <script crossorigin="anonymous" integrity="sha256-Ae2Vz/4ePdIu6ZyI/5ZGsYnb+m0JlOmKPjt6XZ9JJkA=" src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.4/require.min.js"></script>
    <script src="../../../../_static/js/theme.js"></script>
    <link rel="index" title="Index" href="../../../../genindex.html" />
    <link rel="search" title="Search" href="../../../../search.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >

          
          
          <a href="../../../../index.html" class="icon icon-home">
            TransBrain
              <img src="../../../../_static/logo_final.png" class="logo" alt="Logo"/>
          </a>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../../../../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" aria-label="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <p class="caption" role="heading"><span class="caption-text">Setup &amp; Usage</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../../../../installation.html">Installation</a><ul>
<li class="toctree-l2"><a class="reference internal" href="../../../../installation.html#installation-options">Installation Options</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../../installation.html#python-dependencies">Python Dependencies</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../../installation.html#unit-test">Unit Test</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="../../../../user.html">User Guide</a><ul>
<li class="toctree-l2"><a class="reference internal" href="../../../../user.html#id2">Online Mapping</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../../user.html#id3">General Use</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../../user.html#id4">Advanced Use</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="../../../../atlas.html">Supported Atlases</a><ul>
<li class="toctree-l2"><a class="reference internal" href="../../../../atlas.html#human-brain-atlases">Human Brain Atlases</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../../atlas.html#mouse-brain-atlas">Mouse Brain Atlas</a></li>
</ul>
</li>
</ul>
<p class="caption" role="heading"><span class="caption-text">Tutorials</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../../../../methods/method_glossory.html">Method Glossory</a><ul>
<li class="toctree-l2"><a class="reference internal" href="../../../../methods/method_glossory.html#datasets">Datasets</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../../methods/method_glossory.html#methods-or-concepts">Methods or Concepts</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="../../../../online_mapping.html">Online Mapping</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../../tutorials/notebooks/basic_usage/basic_use_mouse_to_human.html">Basic use 1: Mouse to human</a><ul>
<li class="toctree-l2"><a class="reference internal" href="../../../../tutorials/notebooks/basic_usage/basic_use_mouse_to_human.html#Step-1:-Prepare-input-data">Step 1: Prepare input data</a><ul>
<li class="toctree-l3"><a class="reference internal" href="../../../../tutorials/notebooks/basic_usage/basic_use_mouse_to_human.html#Visualizing-mouse-phenotype-data">Visualizing mouse phenotype data</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="../../../../tutorials/notebooks/basic_usage/basic_use_mouse_to_human.html#Step-2:-Initialization">Step 2: Initialization</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../../tutorials/notebooks/basic_usage/basic_use_mouse_to_human.html#Step-3:-Mapping">Step 3: Mapping</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="../../../../tutorials/notebooks/basic_usage/basic_use_human_to_mouse.html">Basic use 2: Human to mouse</a><ul>
<li class="toctree-l2"><a class="reference internal" href="../../../../tutorials/notebooks/basic_usage/basic_use_human_to_mouse.html#Step-1:-Prepare-input-data">Step 1: Prepare input data</a><ul>
<li class="toctree-l3"><a class="reference internal" href="../../../../tutorials/notebooks/basic_usage/basic_use_human_to_mouse.html#Visualizing-human-phenotype-data">Visualizing human phenotype data</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="../../../../tutorials/notebooks/basic_usage/basic_use_human_to_mouse.html#Step-2:-Initialization">Step 2: Initialization</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../../tutorials/notebooks/basic_usage/basic_use_human_to_mouse.html#Step-3:-Mapping">Step 3: Mapping</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="../../../../tutorials/notebooks/translation/gradient_spectrum/fMRI_gradients.html">Case 1: Analyzing the conservation of resting-fMRI gradients</a><ul>
<li class="toctree-l2"><a class="reference internal" href="../../../../tutorials/notebooks/translation/gradient_spectrum/fMRI_gradients.html#Introduction">Introduction</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../../tutorials/notebooks/translation/gradient_spectrum/fMRI_gradients.html#Mapping-mouse-gradient-to-human">Mapping mouse gradient to human</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../../tutorials/notebooks/translation/gradient_spectrum/fMRI_gradients.html#Compare-with-human-Gradient">Compare with human Gradient</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../../tutorials/notebooks/translation/gradient_spectrum/fMRI_gradients.html#Cross-species-correlation">Cross-species correlation</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="../../../../tutorials/notebooks/translation/optogenetic_annotation/optogenetic_circuits.html">Case 2: Annotating optogenetic circuits</a><ul>
<li class="toctree-l2"><a class="reference internal" href="../../../../tutorials/notebooks/translation/optogenetic_annotation/optogenetic_circuits.html#Introduction">Introduction</a><ul>
<li class="toctree-l3"><a class="reference internal" href="../../../../tutorials/notebooks/translation/optogenetic_annotation/optogenetic_circuits.html#Data">Data</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="../../../../tutorials/notebooks/translation/optogenetic_annotation/optogenetic_circuits.html#Translating-mouse-stimulated-pattern-to-human">Translating mouse stimulated pattern to human</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../../tutorials/notebooks/translation/optogenetic_annotation/optogenetic_circuits.html#Annotating-using-NeuroSynth.">Annotating using NeuroSynth.</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="../../../../tutorials/notebooks/translation/autism_mutation/gene_mutations.html">Case 3: Linking gene mutations to imaging phenotype deviations</a><ul>
<li class="toctree-l2"><a class="reference internal" href="../../../../tutorials/notebooks/translation/autism_mutation/gene_mutations.html#Introduction">Introduction</a><ul>
<li class="toctree-l3"><a class="reference internal" href="../../../../tutorials/notebooks/translation/autism_mutation/gene_mutations.html#Data">Data</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="../../../../tutorials/notebooks/translation/autism_mutation/gene_mutations.html#Translating-mouse-mutation-patterns-to-human">Translating mouse mutation patterns to human</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../../tutorials/notebooks/translation/autism_mutation/gene_mutations.html#Calculate-individual-gene-risk-score">Calculate individual gene risk score</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="../../../../building.html">Advanced Use</a><ul>
<li class="toctree-l2"><a class="reference internal" href="../../../../pipeline/run_pipeline/integration.html">Transcriptional Data Integration</a><ul>
<li class="toctree-l3"><a class="reference internal" href="../../../../pipeline/run_pipeline/integration.html#Introduction">Introduction</a><ul>
<li class="toctree-l4"><a class="reference internal" href="../../../../pipeline/run_pipeline/integration.html#AHBA-data">AHBA data</a></li>
<li class="toctree-l4"><a class="reference internal" href="../../../../pipeline/run_pipeline/integration.html#Human-single-nucleus-data">Human single-nucleus data</a></li>
<li class="toctree-l4"><a class="reference internal" href="../../../../pipeline/run_pipeline/integration.html#Mouse-spatial-transcriptomic-data">Mouse spatial transcriptomic data</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="../../../../pipeline/run_pipeline/integration.html#Selection-of-cortical-stable-genes">Selection of cortical stable genes</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../../../pipeline/run_pipeline/integration.html#Selection-of-subcortical-stable-genes">Selection of subcortical stable genes</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../../../pipeline/run_pipeline/integration.html#Selection-of-all-stable-genes">Selection of all stable genes</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../../../pipeline/run_pipeline/integration.html#Thresholds-of-cortex">Thresholds of cortex</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../../../pipeline/run_pipeline/integration.html#Thresholds-of-subcortex">Thresholds of subcortex</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../../../pipeline/run_pipeline/integration.html#Thresholds-of-all">Thresholds of all</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="../../../../pipeline/run_pipeline/dnn.html">DNN Model Training</a><ul>
<li class="toctree-l3"><a class="reference internal" href="../../../../pipeline/run_pipeline/dnn.html#Introduction">Introduction</a><ul>
<li class="toctree-l4"><a class="reference internal" href="../../../../pipeline/run_pipeline/dnn.html#Model-architecture">Model architecture</a></li>
<li class="toctree-l4"><a class="reference internal" href="../../../../pipeline/run_pipeline/dnn.html#Region-specific-embeddings">Region-specific embeddings</a></li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="../../../../pipeline/run_pipeline/graph_embedding.html">Graph-based Random Walk</a><ul>
<li class="toctree-l3"><a class="reference internal" href="../../../../pipeline/run_pipeline/graph_embedding.html#Introduction">Introduction</a><ul>
<li class="toctree-l4"><a class="reference internal" href="../../../../pipeline/run_pipeline/graph_embedding.html#Human-diffusion-based-tractography">Human diffusion-based tractography</a></li>
<li class="toctree-l4"><a class="reference internal" href="../../../../pipeline/run_pipeline/graph_embedding.html#Mouse-Tracer">Mouse Tracer</a></li>
<li class="toctree-l4"><a class="reference internal" href="../../../../pipeline/run_pipeline/graph_embedding.html#Embedding-space-generated-from-Human-Mouse-graph">Embedding space generated from Human-Mouse graph</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="../../../../pipeline/run_pipeline/graph_embedding.html#Define-human-and-mouse-ROIs">Define human and mouse ROIs</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../../../pipeline/run_pipeline/graph_embedding.html#Load-anatomical-hierachy-constraint-information">Load anatomical hierachy constraint information</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="../../../../methods/mapping.html">Bidirectional Mapping</a></li>
</ul>
</li>
</ul>
<p class="caption" role="heading"><span class="caption-text">API</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../../../../api/external_api.html">API Reference</a><ul>
<li class="toctree-l2"><a class="reference internal" href="../../../../api/fetch_api.html">Fetching Atlases</a><ul>
<li class="toctree-l3"><a class="reference internal" href="../../../../api/fetch_api.html#transbrain.atlas.fetch_human_atlas"><code class="docutils literal notranslate"><span class="pre">fetch_human_atlas()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="../../../../api/fetch_api.html#transbrain.atlas.fetch_mouse_atlas"><code class="docutils literal notranslate"><span class="pre">fetch_mouse_atlas()</span></code></a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="../../../../api/base_api.html">Base Functions</a><ul>
<li class="toctree-l3"><a class="reference internal" href="../../../../api/base_api.html#transbrain.base.get_region_phenotypes"><code class="docutils literal notranslate"><span class="pre">get_region_phenotypes()</span></code></a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="../../../../api/mapping_api.html">Mapping Functions</a><ul>
<li class="toctree-l3"><a class="reference internal" href="../../../../api/mapping_api.html#transbrain.trans.SpeciesTrans"><code class="docutils literal notranslate"><span class="pre">SpeciesTrans</span></code></a><ul>
<li class="toctree-l4"><a class="reference internal" href="../../../../api/mapping_api.html#transbrain.trans.SpeciesTrans.atlas_type"><code class="docutils literal notranslate"><span class="pre">SpeciesTrans.atlas_type</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="../../../../api/mapping_api.html#transbrain.trans.SpeciesTrans.regions"><code class="docutils literal notranslate"><span class="pre">SpeciesTrans.regions</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="../../../../api/mapping_api.html#transbrain.trans.SpeciesTrans.embeddings"><code class="docutils literal notranslate"><span class="pre">SpeciesTrans.embeddings</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="../../../../api/mapping_api.html#transbrain.trans.SpeciesTrans._load_embeddings"><code class="docutils literal notranslate"><span class="pre">SpeciesTrans._load_embeddings()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="../../../../api/mapping_api.html#transbrain.trans.SpeciesTrans._dual_mapping"><code class="docutils literal notranslate"><span class="pre">SpeciesTrans._dual_mapping()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="../../../../api/mapping_api.html#transbrain.trans.SpeciesTrans._translate"><code class="docutils literal notranslate"><span class="pre">SpeciesTrans._translate()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="../../../../api/mapping_api.html#transbrain.trans.SpeciesTrans.mouse_to_human"><code class="docutils literal notranslate"><span class="pre">SpeciesTrans.mouse_to_human()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="../../../../api/mapping_api.html#transbrain.trans.SpeciesTrans.human_to_mouse"><code class="docutils literal notranslate"><span class="pre">SpeciesTrans.human_to_mouse()</span></code></a></li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="../../../../api/vis_api.html">Visualization Functions</a><ul>
<li class="toctree-l3"><a class="reference internal" href="../../../../api/vis_api.html#transbrain.vis.map_phenotype_to_nifti"><code class="docutils literal notranslate"><span class="pre">map_phenotype_to_nifti()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="../../../../api/vis_api.html#transbrain.vis.plot_mouse_phenotype"><code class="docutils literal notranslate"><span class="pre">plot_mouse_phenotype()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="../../../../api/vis_api.html#transbrain.vis.plot_human_phenotype"><code class="docutils literal notranslate"><span class="pre">plot_human_phenotype()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="../../../../api/vis_api.html#transbrain.surf.plot_surface_phenotype"><code class="docutils literal notranslate"><span class="pre">plot_surface_phenotype()</span></code></a></li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="../../../../api/utils.html">Development Functions</a><ul>
<li class="toctree-l2"><a class="reference internal" href="../../../../api/data_process_api.html">Transcriptional Data Process</a><ul>
<li class="toctree-l3"><a class="reference internal" href="../../../../api/data_process_api.html#pipeline.code.dpreprocess.sn_.Preprocess_Sn"><code class="docutils literal notranslate"><span class="pre">Preprocess_Sn</span></code></a><ul>
<li class="toctree-l4"><a class="reference internal" href="../../../../api/data_process_api.html#pipeline.code.dpreprocess.sn_.Preprocess_Sn.dataset_qc"><code class="docutils literal notranslate"><span class="pre">Preprocess_Sn.dataset_qc()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="../../../../api/data_process_api.html#pipeline.code.dpreprocess.sn_.Preprocess_Sn.dataset_normalized"><code class="docutils literal notranslate"><span class="pre">Preprocess_Sn.dataset_normalized()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="../../../../api/data_process_api.html#pipeline.code.dpreprocess.sn_.Preprocess_Sn.dataset_common_genes"><code class="docutils literal notranslate"><span class="pre">Preprocess_Sn.dataset_common_genes()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="../../../../api/data_process_api.html#pipeline.code.dpreprocess.sn_.Preprocess_Sn.dataset_batch_correct"><code class="docutils literal notranslate"><span class="pre">Preprocess_Sn.dataset_batch_correct()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="../../../../api/data_process_api.html#pipeline.code.dpreprocess.sn_.Preprocess_Sn.dataset_smooth"><code class="docutils literal notranslate"><span class="pre">Preprocess_Sn.dataset_smooth()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="../../../../api/data_process_api.html#pipeline.code.dpreprocess.sn_.Preprocess_Sn.dataset_scaled"><code class="docutils literal notranslate"><span class="pre">Preprocess_Sn.dataset_scaled()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="../../../../api/data_process_api.html#pipeline.code.dpreprocess.sn_.Preprocess_Sn.preprocess_pipeline"><code class="docutils literal notranslate"><span class="pre">Preprocess_Sn.preprocess_pipeline()</span></code></a></li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="../../../../api/data_integration_api.html">Data Integration</a><ul>
<li class="toctree-l3"><a class="reference internal" href="../../../../api/data_integration_api.html#pipeline.code.dintegrate.utils.stable_gene_filter"><code class="docutils literal notranslate"><span class="pre">stable_gene_filter()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="../../../../api/data_integration_api.html#pipeline.code.dintegrate.utils.calculate_thresholds"><code class="docutils literal notranslate"><span class="pre">calculate_thresholds()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="../../../../api/data_integration_api.html#pipeline.code.dintegrate.main.DataIntegrator"><code class="docutils literal notranslate"><span class="pre">DataIntegrator</span></code></a><ul>
<li class="toctree-l4"><a class="reference internal" href="../../../../api/data_integration_api.html#pipeline.code.dintegrate.main.DataIntegrator.__init__"><code class="docutils literal notranslate"><span class="pre">DataIntegrator.__init__()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="../../../../api/data_integration_api.html#pipeline.code.dintegrate.main.DataIntegrator.run"><code class="docutils literal notranslate"><span class="pre">DataIntegrator.run()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="../../../../api/data_integration_api.html#pipeline.code.dintegrate.main.DataIntegrator.pool_construction"><code class="docutils literal notranslate"><span class="pre">DataIntegrator.pool_construction()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="../../../../api/data_integration_api.html#pipeline.code.dintegrate.main.DataIntegrator.clean_pool"><code class="docutils literal notranslate"><span class="pre">DataIntegrator.clean_pool()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="../../../../api/data_integration_api.html#pipeline.code.dintegrate.main.DataIntegrator.weighted_average"><code class="docutils literal notranslate"><span class="pre">DataIntegrator.weighted_average()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="../../../../api/data_integration_api.html#pipeline.code.dintegrate.main.DataIntegrator.downsample_data"><code class="docutils literal notranslate"><span class="pre">DataIntegrator.downsample_data()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="../../../../api/data_integration_api.html#pipeline.code.dintegrate.main.DataIntegrator.generate_integrated_anndata"><code class="docutils literal notranslate"><span class="pre">DataIntegrator.generate_integrated_anndata()</span></code></a></li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="../../../../api/dnn_api.html">DNN</a><ul>
<li class="toctree-l3"><a class="reference internal" href="../../../../api/dnn_api.html#pipeline.code.dnn.analyzer.calculate_rank"><code class="docutils literal notranslate"><span class="pre">calculate_rank()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="../../../../api/dnn_api.html#pipeline.code.dnn.analyzer.transform_human_mouse"><code class="docutils literal notranslate"><span class="pre">transform_human_mouse()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="../../../../api/dnn_api.html#pipeline.code.dnn.analyzer.independent_test"><code class="docutils literal notranslate"><span class="pre">independent_test()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="../../../../api/dnn_api.html#pipeline.code.dnn.pipeline.DnnPipeline"><code class="docutils literal notranslate"><span class="pre">DnnPipeline</span></code></a><ul>
<li class="toctree-l4"><a class="reference internal" href="../../../../api/dnn_api.html#pipeline.code.dnn.pipeline.DnnPipeline.cfg"><code class="docutils literal notranslate"><span class="pre">DnnPipeline.cfg</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="../../../../api/dnn_api.html#pipeline.code.dnn.pipeline.DnnPipeline.logger"><code class="docutils literal notranslate"><span class="pre">DnnPipeline.logger</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="../../../../api/dnn_api.html#pipeline.code.dnn.pipeline.DnnPipeline.execute"><code class="docutils literal notranslate"><span class="pre">DnnPipeline.execute()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="../../../../api/dnn_api.html#pipeline.code.dnn.pipeline.DnnPipeline.train"><code class="docutils literal notranslate"><span class="pre">DnnPipeline.train()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="../../../../api/dnn_api.html#pipeline.code.dnn.pipeline.DnnPipeline.trans"><code class="docutils literal notranslate"><span class="pre">DnnPipeline.trans()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="../../../../api/dnn_api.html#pipeline.code.dnn.pipeline.DnnPipeline.test"><code class="docutils literal notranslate"><span class="pre">DnnPipeline.test()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="../../../../api/dnn_api.html#pipeline.code.dnn.pipeline.DnnPipeline._process_single_dataset"><code class="docutils literal notranslate"><span class="pre">DnnPipeline._process_single_dataset()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="../../../../api/dnn_api.html#pipeline.code.dnn.pipeline.DnnPipeline._train_single_model"><code class="docutils literal notranslate"><span class="pre">DnnPipeline._train_single_model()</span></code></a></li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="../../../../api/graph_walk_api.html">Graph Walk</a><ul>
<li class="toctree-l3"><a class="reference internal" href="../../../../api/graph_walk_api.html#pipeline.code.graph_walk.run.generate_embeddings"><code class="docutils literal notranslate"><span class="pre">generate_embeddings()</span></code></a></li>
</ul>
</li>
</ul>
</li>
</ul>
<p class="caption" role="heading"><span class="caption-text">Other</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../../../../faq.html">Frequently Asked Questions</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../../citing.html">Citing TransBrain</a></li>
</ul>

        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../../../../index.html">TransBrain</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="../../../../index.html" class="icon icon-home" aria-label="Home"></a></li>
          <li class="breadcrumb-item"><a href="../../../index.html">Module code</a></li>
      <li class="breadcrumb-item active">pipeline.code.dpreprocess.ahba_</li>
      <li class="wy-breadcrumbs-aside">
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  <h1>Source code for pipeline.code.dpreprocess.ahba_</h1><div class="highlight"><pre>
<span></span><span class="kn">import</span><span class="w"> </span><span class="nn">warnings</span>
<span class="n">warnings</span><span class="o">.</span><span class="n">filterwarnings</span><span class="p">(</span><span class="s1">&#39;ignore&#39;</span><span class="p">)</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">abagen</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">pandas</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">pd</span>
<span class="kn">from</span><span class="w"> </span><span class="nn">nilearn</span><span class="w"> </span><span class="kn">import</span> <span class="n">image</span>

<span class="kn">import</span><span class="w"> </span><span class="nn">os</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">itertools</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">anndata</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">scanpy</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">sc</span>

<span class="n">micro_path</span><span class="o">=</span><span class="s1">&#39;../../datasets/microarray&#39;</span> <span class="c1"># AHBA microarray</span>

<div class="viewcode-block" id="extract_AHBA_data"><a class="viewcode-back" href="../../../../api/data_process_api.html#pipeline.code.dpreprocess.ahba_.extract_AHBA_data">[docs]</a><span class="k">def</span><span class="w"> </span><span class="nf">extract_AHBA_data</span><span class="p">(</span><span class="n">atlas_path</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span><span class="n">atlas_info_path</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span><span class="n">lr_mirror</span><span class="o">=</span><span class="s1">&#39;bidirectional&#39;</span><span class="p">,</span><span class="n">gene_norm</span><span class="o">=</span><span class="s1">&#39;srs&#39;</span><span class="p">,</span><span class="n">sample_norm</span><span class="o">=</span><span class="s1">&#39;srs&#39;</span><span class="p">,</span><span class="n">return_report</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
                      <span class="n">return_counts</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span><span class="n">return_donors</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span><span class="n">norm_matched</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span><span class="n">ibf_threshold</span><span class="o">=</span><span class="mi">0</span><span class="p">,</span><span class="n">region_agg</span><span class="o">=</span><span class="kc">None</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">:</span>
<span class="w">    </span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Extracts Allen Human Brain Atlas (AHBA) gene expression data based on a given brain atlas.</span>
<span class="sd">    For more information, please ref https://abagen.readthedocs.io/en/stable/generated/abagen.get_expression_data.html</span>

<span class="sd">    Parameters</span>
<span class="sd">    ----------</span>
<span class="sd">    atlas_path : str</span>
<span class="sd">        Path to the human brain atlas NIfTI file.</span>
<span class="sd">    atlas_info_path : str</span>
<span class="sd">        Path to a CSV file containing atlas region information with columns:</span>
<span class="sd">        &#39;Anatomical Name&#39; and &#39;Atlas Index&#39;.</span>
<span class="sd">    lr_mirror : {&#39;left&#39;, &#39;right&#39;, &#39;bidirectional&#39;}, optional</span>
<span class="sd">        How to mirror samples across hemispheres. Default is &#39;bidirectional&#39;.</span>
<span class="sd">    gene_norm : str, optional</span>
<span class="sd">        Method by which to normalize microarray expression values for each donor. Default is &#39;srs&#39;.</span>
<span class="sd">    sample_norm : str, optional</span>
<span class="sd">        Method by which to normalize microarray expression values for each sample. Default is &#39;srs&#39;.</span>
<span class="sd">    return_report : bool, optional</span>
<span class="sd">        Whether to return a string containing longform text describing the processing procedures used to generate the expression DataFrames returned by this function. Default is True.</span>
<span class="sd">    return_counts : bool, optional</span>
<span class="sd">        Whether to return dataframe containing information on how many samples were assigned to each parcel in atlas for each donor. Default is True.</span>
<span class="sd">    return_donors : bool, optional</span>
<span class="sd">        Whether to return donor-level expression arrays instead of aggregating expression across donors with provided agg_metric. Default is True.</span>
<span class="sd">    norm_matched : bool, optional</span>
<span class="sd">        Whether to perform gene normalization (gene_norm) across only those samples matched to regions in atlas instead of all available samples. Default is False.</span>
<span class="sd">    ibf_threshold : float, optional</span>
<span class="sd">        Intensity-based filtering threshold. Default is 0.</span>
<span class="sd">    region_agg : str or None, optional</span>
<span class="sd">        Mechanism by which to reduce sample-level expression data into region-level expression.</span>

<span class="sd">    Returns</span>
<span class="sd">    -------</span>
<span class="sd">    dict[str, pandas.DataFrame] or pandas.DataFrame</span>
<span class="sd">        If `return_donors=True`, returns a dictionary where keys are donor IDs and values are gene expression DataFrames (regions × genes).</span>
<span class="sd">        If `return_donors=False`, returns a single aggregated DataFrame.</span>

<span class="sd">    Examples</span>
<span class="sd">    --------</span>
<span class="sd">    &gt;&gt;&gt; expr_data = extract_AHBA_data(</span>
<span class="sd">    ...     &#39;/path/to/atlas.nii.gz&#39;,</span>
<span class="sd">    ...     &#39;/path/to/atlas_info.csv&#39;</span>
<span class="sd">    ... ) </span>
<span class="sd">    &quot;&quot;&quot;</span>
    
    <span class="c1"># Load atlas and region information</span>
    <span class="n">atlas</span> <span class="o">=</span> <span class="n">image</span><span class="o">.</span><span class="n">load_img</span><span class="p">(</span><span class="n">atlas_path</span><span class="p">)</span>
    <span class="n">atlas_info</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="n">atlas_info_path</span><span class="p">)</span>

    <span class="n">atlas_dict</span> <span class="o">=</span> <span class="p">(</span>
        <span class="n">atlas_info</span>
        <span class="o">.</span><span class="n">set_index</span><span class="p">(</span><span class="s1">&#39;Atlas Index&#39;</span><span class="p">)[</span><span class="s1">&#39;Anatomical Name&#39;</span><span class="p">]</span>
        <span class="o">.</span><span class="n">to_dict</span><span class="p">()</span>
    <span class="p">)</span>
    <span class="n">atlas_dict</span><span class="o">.</span><span class="n">setdefault</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="s1">&#39;other&#39;</span><span class="p">)</span>

    <span class="c1"># Get expression data from abagen</span>
    <span class="n">expr_raw</span> <span class="o">=</span> <span class="n">abagen</span><span class="o">.</span><span class="n">get_expression_data</span><span class="p">(</span>
        <span class="n">atlas</span><span class="p">,</span> <span class="n">data_dir</span><span class="o">=</span><span class="n">micro_path</span><span class="p">,</span> <span class="n">lr_mirror</span><span class="o">=</span><span class="n">lr_mirror</span><span class="p">,</span> <span class="n">gene_norm</span><span class="o">=</span><span class="n">gene_norm</span><span class="p">,</span>
        <span class="n">sample_norm</span><span class="o">=</span><span class="n">sample_norm</span><span class="p">,</span> <span class="n">return_report</span><span class="o">=</span><span class="n">return_report</span><span class="p">,</span> <span class="n">return_counts</span><span class="o">=</span><span class="n">return_counts</span><span class="p">,</span>
        <span class="n">return_donors</span><span class="o">=</span><span class="n">return_donors</span><span class="p">,</span> <span class="n">norm_matched</span><span class="o">=</span><span class="n">norm_matched</span><span class="p">,</span> <span class="n">ibf_threshold</span><span class="o">=</span><span class="n">ibf_threshold</span><span class="p">,</span>
        <span class="n">region_agg</span><span class="o">=</span><span class="n">region_agg</span>
    <span class="p">)</span>

    <span class="c1"># Process each donor&#39;s data</span>
    <span class="n">processed_data</span> <span class="o">=</span> <span class="p">{}</span>
    <span class="k">for</span> <span class="n">donor_id</span><span class="p">,</span> <span class="n">expr_data</span> <span class="ow">in</span> <span class="n">expr_raw</span><span class="o">.</span><span class="n">items</span><span class="p">():</span>
        <span class="n">index_values</span> <span class="o">=</span> <span class="p">(</span>
            <span class="n">expr_data</span><span class="o">.</span><span class="n">index</span><span class="o">.</span><span class="n">get_level_values</span><span class="p">(</span><span class="mi">0</span><span class="p">)</span>
            <span class="k">if</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">expr_data</span><span class="o">.</span><span class="n">index</span><span class="p">,</span> <span class="n">pd</span><span class="o">.</span><span class="n">MultiIndex</span><span class="p">)</span>
            <span class="k">else</span> <span class="n">expr_data</span><span class="o">.</span><span class="n">index</span>
        <span class="p">)</span>

        <span class="n">region_labels</span> <span class="o">=</span> <span class="n">index_values</span><span class="o">.</span><span class="n">map</span><span class="p">(</span><span class="n">atlas_dict</span><span class="p">)</span><span class="o">.</span><span class="n">fillna</span><span class="p">(</span><span class="s1">&#39;other&#39;</span><span class="p">)</span>
        <span class="n">processed_data</span><span class="p">[</span><span class="n">donor_id</span><span class="p">]</span> <span class="o">=</span> <span class="n">expr_data</span><span class="o">.</span><span class="n">set_index</span><span class="p">(</span><span class="n">region_labels</span><span class="o">.</span><span class="n">values</span><span class="p">)</span>

    <span class="k">return</span> <span class="n">processed_data</span></div>
</pre></div>

           </div>
          </div>
          <footer>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright 2025, TransBrain.</p>
  </div>

  Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    provided by <a href="https://readthedocs.org">Read the Docs</a>.
   

</footer>
        </div>
      </div>
    </section>
  </div>
  <script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>