

<!DOCTYPE html>
<html class="writer-html5" lang="en">
<head>
  <meta charset="utf-8" /><meta name="viewport" content="width=device-width, initial-scale=1" />

  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>Method Glossory &mdash; TransBrain 0.1 documentation</title>
      <link rel="stylesheet" type="text/css" href="../_static/pygments.css?v=03e43079" />
      <link rel="stylesheet" type="text/css" href="../_static/css/theme.css?v=e59714d7" />

  
      <script src="../_static/jquery.js?v=5d32c60e"></script>
      <script src="../_static/_sphinx_javascript_frameworks_compat.js?v=2cd50e6c"></script>
      <script data-url_root="../" id="documentation_options" src="../_static/documentation_options.js?v=e031e9a9"></script>
      <script src="../_static/doctools.js?v=888ff710"></script>
      <script src="../_static/sphinx_highlight.js?v=4825356b"></script>
      <script crossorigin="anonymous" integrity="sha256-Ae2Vz/4ePdIu6ZyI/5ZGsYnb+m0JlOmKPjt6XZ9JJkA=" src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.4/require.min.js"></script>
    <script src="../_static/js/theme.js"></script>
    <link rel="index" title="Index" href="../genindex.html" />
    <link rel="search" title="Search" href="../search.html" />
    <link rel="next" title="Online Mapping" href="../online_mapping.html" />
    <link rel="prev" title="Supported Atlases" href="../atlas.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >

          
          
          <a href="../index.html" class="icon icon-home">
            TransBrain
              <img src="../_static/logo_final.png" class="logo" alt="Logo"/>
          </a>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" aria-label="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <p class="caption" role="heading"><span class="caption-text">Setup &amp; Usage</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../installation.html">Installation</a><ul>
<li class="toctree-l2"><a class="reference internal" href="../installation.html#installation-options">Installation Options</a></li>
<li class="toctree-l2"><a class="reference internal" href="../installation.html#python-dependencies">Python Dependencies</a></li>
<li class="toctree-l2"><a class="reference internal" href="../installation.html#unit-test">Unit Test</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="../user.html">User Guide</a><ul>
<li class="toctree-l2"><a class="reference internal" href="../user.html#id2">Online Mapping</a></li>
<li class="toctree-l2"><a class="reference internal" href="../user.html#id3">General Use</a></li>
<li class="toctree-l2"><a class="reference internal" href="../user.html#id4">Advanced Use</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="../atlas.html">Supported Atlases</a><ul>
<li class="toctree-l2"><a class="reference internal" href="../atlas.html#human-brain-atlases">Human Brain Atlases</a></li>
<li class="toctree-l2"><a class="reference internal" href="../atlas.html#mouse-brain-atlas">Mouse Brain Atlas</a></li>
</ul>
</li>
</ul>
<p class="caption" role="heading"><span class="caption-text">Tutorials</span></p>
<ul class="current">
<li class="toctree-l1 current"><a class="current reference internal" href="#">Method Glossory</a><ul>
<li class="toctree-l2"><a class="reference internal" href="#datasets">Datasets</a></li>
<li class="toctree-l2"><a class="reference internal" href="#methods-or-concepts">Methods or Concepts</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="../online_mapping.html">Online Mapping</a></li>
<li class="toctree-l1"><a class="reference internal" href="../tutorials/notebooks/basic_usage/basic_use_mouse_to_human.html">Basic use 1: Mouse to human</a><ul>
<li class="toctree-l2"><a class="reference internal" href="../tutorials/notebooks/basic_usage/basic_use_mouse_to_human.html#Step-1:-Prepare-input-data">Step 1: Prepare input data</a><ul>
<li class="toctree-l3"><a class="reference internal" href="../tutorials/notebooks/basic_usage/basic_use_mouse_to_human.html#Visualizing-mouse-phenotype-data">Visualizing mouse phenotype data</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="../tutorials/notebooks/basic_usage/basic_use_mouse_to_human.html#Step-2:-Initialization">Step 2: Initialization</a></li>
<li class="toctree-l2"><a class="reference internal" href="../tutorials/notebooks/basic_usage/basic_use_mouse_to_human.html#Step-3:-Mapping">Step 3: Mapping</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="../tutorials/notebooks/basic_usage/basic_use_human_to_mouse.html">Basic use 2: Human to mouse</a><ul>
<li class="toctree-l2"><a class="reference internal" href="../tutorials/notebooks/basic_usage/basic_use_human_to_mouse.html#Step-1:-Prepare-input-data">Step 1: Prepare input data</a><ul>
<li class="toctree-l3"><a class="reference internal" href="../tutorials/notebooks/basic_usage/basic_use_human_to_mouse.html#Visualizing-human-phenotype-data">Visualizing human phenotype data</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="../tutorials/notebooks/basic_usage/basic_use_human_to_mouse.html#Step-2:-Initialization">Step 2: Initialization</a></li>
<li class="toctree-l2"><a class="reference internal" href="../tutorials/notebooks/basic_usage/basic_use_human_to_mouse.html#Step-3:-Mapping">Step 3: Mapping</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="../tutorials/notebooks/translation/gradient_spectrum/fMRI_gradients.html">Case 1: Analyzing the conservation of resting-fMRI gradients</a><ul>
<li class="toctree-l2"><a class="reference internal" href="../tutorials/notebooks/translation/gradient_spectrum/fMRI_gradients.html#Introduction">Introduction</a></li>
<li class="toctree-l2"><a class="reference internal" href="../tutorials/notebooks/translation/gradient_spectrum/fMRI_gradients.html#Mapping-mouse-gradient-to-human">Mapping mouse gradient to human</a></li>
<li class="toctree-l2"><a class="reference internal" href="../tutorials/notebooks/translation/gradient_spectrum/fMRI_gradients.html#Compare-with-human-Gradient">Compare with human Gradient</a></li>
<li class="toctree-l2"><a class="reference internal" href="../tutorials/notebooks/translation/gradient_spectrum/fMRI_gradients.html#Cross-species-correlation">Cross-species correlation</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="../tutorials/notebooks/translation/optogenetic_annotation/optogenetic_circuits.html">Case 2: Annotating optogenetic circuits</a><ul>
<li class="toctree-l2"><a class="reference internal" href="../tutorials/notebooks/translation/optogenetic_annotation/optogenetic_circuits.html#Introduction">Introduction</a><ul>
<li class="toctree-l3"><a class="reference internal" href="../tutorials/notebooks/translation/optogenetic_annotation/optogenetic_circuits.html#Data">Data</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="../tutorials/notebooks/translation/optogenetic_annotation/optogenetic_circuits.html#Translating-mouse-stimulated-pattern-to-human">Translating mouse stimulated pattern to human</a></li>
<li class="toctree-l2"><a class="reference internal" href="../tutorials/notebooks/translation/optogenetic_annotation/optogenetic_circuits.html#Annotating-using-NeuroSynth.">Annotating using NeuroSynth.</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="../tutorials/notebooks/translation/autism_mutation/gene_mutations.html">Case 3: Linking gene mutations to imaging phenotype deviations</a><ul>
<li class="toctree-l2"><a class="reference internal" href="../tutorials/notebooks/translation/autism_mutation/gene_mutations.html#Introduction">Introduction</a><ul>
<li class="toctree-l3"><a class="reference internal" href="../tutorials/notebooks/translation/autism_mutation/gene_mutations.html#Data">Data</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="../tutorials/notebooks/translation/autism_mutation/gene_mutations.html#Translating-mouse-mutation-patterns-to-human">Translating mouse mutation patterns to human</a></li>
<li class="toctree-l2"><a class="reference internal" href="../tutorials/notebooks/translation/autism_mutation/gene_mutations.html#Calculate-individual-gene-risk-score">Calculate individual gene risk score</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="../building.html">Advanced Use</a><ul>
<li class="toctree-l2"><a class="reference internal" href="../pipeline/run_pipeline/integration.html">Transcriptional Data Integration</a><ul>
<li class="toctree-l3"><a class="reference internal" href="../pipeline/run_pipeline/integration.html#Introduction">Introduction</a><ul>
<li class="toctree-l4"><a class="reference internal" href="../pipeline/run_pipeline/integration.html#AHBA-data">AHBA data</a></li>
<li class="toctree-l4"><a class="reference internal" href="../pipeline/run_pipeline/integration.html#Human-single-nucleus-data">Human single-nucleus data</a></li>
<li class="toctree-l4"><a class="reference internal" href="../pipeline/run_pipeline/integration.html#Mouse-spatial-transcriptomic-data">Mouse spatial transcriptomic data</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="../pipeline/run_pipeline/integration.html#Selection-of-cortical-stable-genes">Selection of cortical stable genes</a></li>
<li class="toctree-l3"><a class="reference internal" href="../pipeline/run_pipeline/integration.html#Selection-of-subcortical-stable-genes">Selection of subcortical stable genes</a></li>
<li class="toctree-l3"><a class="reference internal" href="../pipeline/run_pipeline/integration.html#Selection-of-all-stable-genes">Selection of all stable genes</a></li>
<li class="toctree-l3"><a class="reference internal" href="../pipeline/run_pipeline/integration.html#Thresholds-of-cortex">Thresholds of cortex</a></li>
<li class="toctree-l3"><a class="reference internal" href="../pipeline/run_pipeline/integration.html#Thresholds-of-subcortex">Thresholds of subcortex</a></li>
<li class="toctree-l3"><a class="reference internal" href="../pipeline/run_pipeline/integration.html#Thresholds-of-all">Thresholds of all</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="../pipeline/run_pipeline/dnn.html">DNN Model Training</a><ul>
<li class="toctree-l3"><a class="reference internal" href="../pipeline/run_pipeline/dnn.html#Introduction">Introduction</a><ul>
<li class="toctree-l4"><a class="reference internal" href="../pipeline/run_pipeline/dnn.html#Model-architecture">Model architecture</a></li>
<li class="toctree-l4"><a class="reference internal" href="../pipeline/run_pipeline/dnn.html#Region-specific-embeddings">Region-specific embeddings</a></li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="../pipeline/run_pipeline/graph_embedding.html">Graph-based Random Walk</a><ul>
<li class="toctree-l3"><a class="reference internal" href="../pipeline/run_pipeline/graph_embedding.html#Introduction">Introduction</a><ul>
<li class="toctree-l4"><a class="reference internal" href="../pipeline/run_pipeline/graph_embedding.html#Human-diffusion-based-tractography">Human diffusion-based tractography</a></li>
<li class="toctree-l4"><a class="reference internal" href="../pipeline/run_pipeline/graph_embedding.html#Mouse-Tracer">Mouse Tracer</a></li>
<li class="toctree-l4"><a class="reference internal" href="../pipeline/run_pipeline/graph_embedding.html#Embedding-space-generated-from-Human-Mouse-graph">Embedding space generated from Human-Mouse graph</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="../pipeline/run_pipeline/graph_embedding.html#Define-human-and-mouse-ROIs">Define human and mouse ROIs</a></li>
<li class="toctree-l3"><a class="reference internal" href="../pipeline/run_pipeline/graph_embedding.html#Load-anatomical-hierachy-constraint-information">Load anatomical hierachy constraint information</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="mapping.html">Bidirectional Mapping</a></li>
</ul>
</li>
</ul>
<p class="caption" role="heading"><span class="caption-text">API</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../api/external_api.html">API Reference</a><ul>
<li class="toctree-l2"><a class="reference internal" href="../api/fetch_api.html">Fetching Atlases</a><ul>
<li class="toctree-l3"><a class="reference internal" href="../api/fetch_api.html#transbrain.atlas.fetch_human_atlas"><code class="docutils literal notranslate"><span class="pre">fetch_human_atlas()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="../api/fetch_api.html#transbrain.atlas.fetch_mouse_atlas"><code class="docutils literal notranslate"><span class="pre">fetch_mouse_atlas()</span></code></a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="../api/base_api.html">Base Functions</a><ul>
<li class="toctree-l3"><a class="reference internal" href="../api/base_api.html#transbrain.base.get_region_phenotypes"><code class="docutils literal notranslate"><span class="pre">get_region_phenotypes()</span></code></a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="../api/mapping_api.html">Mapping Functions</a><ul>
<li class="toctree-l3"><a class="reference internal" href="../api/mapping_api.html#transbrain.trans.SpeciesTrans"><code class="docutils literal notranslate"><span class="pre">SpeciesTrans</span></code></a><ul>
<li class="toctree-l4"><a class="reference internal" href="../api/mapping_api.html#transbrain.trans.SpeciesTrans.atlas_type"><code class="docutils literal notranslate"><span class="pre">SpeciesTrans.atlas_type</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="../api/mapping_api.html#transbrain.trans.SpeciesTrans.regions"><code class="docutils literal notranslate"><span class="pre">SpeciesTrans.regions</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="../api/mapping_api.html#transbrain.trans.SpeciesTrans.embeddings"><code class="docutils literal notranslate"><span class="pre">SpeciesTrans.embeddings</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="../api/mapping_api.html#transbrain.trans.SpeciesTrans._load_embeddings"><code class="docutils literal notranslate"><span class="pre">SpeciesTrans._load_embeddings()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="../api/mapping_api.html#transbrain.trans.SpeciesTrans._dual_mapping"><code class="docutils literal notranslate"><span class="pre">SpeciesTrans._dual_mapping()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="../api/mapping_api.html#transbrain.trans.SpeciesTrans._translate"><code class="docutils literal notranslate"><span class="pre">SpeciesTrans._translate()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="../api/mapping_api.html#transbrain.trans.SpeciesTrans.mouse_to_human"><code class="docutils literal notranslate"><span class="pre">SpeciesTrans.mouse_to_human()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="../api/mapping_api.html#transbrain.trans.SpeciesTrans.human_to_mouse"><code class="docutils literal notranslate"><span class="pre">SpeciesTrans.human_to_mouse()</span></code></a></li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="../api/vis_api.html">Visualization Functions</a><ul>
<li class="toctree-l3"><a class="reference internal" href="../api/vis_api.html#transbrain.vis.map_phenotype_to_nifti"><code class="docutils literal notranslate"><span class="pre">map_phenotype_to_nifti()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="../api/vis_api.html#transbrain.vis.plot_mouse_phenotype"><code class="docutils literal notranslate"><span class="pre">plot_mouse_phenotype()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="../api/vis_api.html#transbrain.vis.plot_human_phenotype"><code class="docutils literal notranslate"><span class="pre">plot_human_phenotype()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="../api/vis_api.html#transbrain.surf.plot_surface_phenotype"><code class="docutils literal notranslate"><span class="pre">plot_surface_phenotype()</span></code></a></li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="../api/utils.html">Development Functions</a><ul>
<li class="toctree-l2"><a class="reference internal" href="../api/data_process_api.html">Transcriptional Data Process</a><ul>
<li class="toctree-l3"><a class="reference internal" href="../api/data_process_api.html#pipeline.code.dpreprocess.sn_.Preprocess_Sn"><code class="docutils literal notranslate"><span class="pre">Preprocess_Sn</span></code></a><ul>
<li class="toctree-l4"><a class="reference internal" href="../api/data_process_api.html#pipeline.code.dpreprocess.sn_.Preprocess_Sn.dataset_qc"><code class="docutils literal notranslate"><span class="pre">Preprocess_Sn.dataset_qc()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="../api/data_process_api.html#pipeline.code.dpreprocess.sn_.Preprocess_Sn.dataset_normalized"><code class="docutils literal notranslate"><span class="pre">Preprocess_Sn.dataset_normalized()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="../api/data_process_api.html#pipeline.code.dpreprocess.sn_.Preprocess_Sn.dataset_common_genes"><code class="docutils literal notranslate"><span class="pre">Preprocess_Sn.dataset_common_genes()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="../api/data_process_api.html#pipeline.code.dpreprocess.sn_.Preprocess_Sn.dataset_batch_correct"><code class="docutils literal notranslate"><span class="pre">Preprocess_Sn.dataset_batch_correct()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="../api/data_process_api.html#pipeline.code.dpreprocess.sn_.Preprocess_Sn.dataset_smooth"><code class="docutils literal notranslate"><span class="pre">Preprocess_Sn.dataset_smooth()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="../api/data_process_api.html#pipeline.code.dpreprocess.sn_.Preprocess_Sn.dataset_scaled"><code class="docutils literal notranslate"><span class="pre">Preprocess_Sn.dataset_scaled()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="../api/data_process_api.html#pipeline.code.dpreprocess.sn_.Preprocess_Sn.preprocess_pipeline"><code class="docutils literal notranslate"><span class="pre">Preprocess_Sn.preprocess_pipeline()</span></code></a></li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="../api/data_integration_api.html">Data Integration</a><ul>
<li class="toctree-l3"><a class="reference internal" href="../api/data_integration_api.html#pipeline.code.dintegrate.utils.stable_gene_filter"><code class="docutils literal notranslate"><span class="pre">stable_gene_filter()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="../api/data_integration_api.html#pipeline.code.dintegrate.utils.calculate_thresholds"><code class="docutils literal notranslate"><span class="pre">calculate_thresholds()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="../api/data_integration_api.html#pipeline.code.dintegrate.main.DataIntegrator"><code class="docutils literal notranslate"><span class="pre">DataIntegrator</span></code></a><ul>
<li class="toctree-l4"><a class="reference internal" href="../api/data_integration_api.html#pipeline.code.dintegrate.main.DataIntegrator.__init__"><code class="docutils literal notranslate"><span class="pre">DataIntegrator.__init__()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="../api/data_integration_api.html#pipeline.code.dintegrate.main.DataIntegrator.run"><code class="docutils literal notranslate"><span class="pre">DataIntegrator.run()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="../api/data_integration_api.html#pipeline.code.dintegrate.main.DataIntegrator.pool_construction"><code class="docutils literal notranslate"><span class="pre">DataIntegrator.pool_construction()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="../api/data_integration_api.html#pipeline.code.dintegrate.main.DataIntegrator.clean_pool"><code class="docutils literal notranslate"><span class="pre">DataIntegrator.clean_pool()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="../api/data_integration_api.html#pipeline.code.dintegrate.main.DataIntegrator.weighted_average"><code class="docutils literal notranslate"><span class="pre">DataIntegrator.weighted_average()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="../api/data_integration_api.html#pipeline.code.dintegrate.main.DataIntegrator.downsample_data"><code class="docutils literal notranslate"><span class="pre">DataIntegrator.downsample_data()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="../api/data_integration_api.html#pipeline.code.dintegrate.main.DataIntegrator.generate_integrated_anndata"><code class="docutils literal notranslate"><span class="pre">DataIntegrator.generate_integrated_anndata()</span></code></a></li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="../api/dnn_api.html">DNN</a><ul>
<li class="toctree-l3"><a class="reference internal" href="../api/dnn_api.html#pipeline.code.dnn.analyzer.calculate_rank"><code class="docutils literal notranslate"><span class="pre">calculate_rank()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="../api/dnn_api.html#pipeline.code.dnn.analyzer.transform_human_mouse"><code class="docutils literal notranslate"><span class="pre">transform_human_mouse()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="../api/dnn_api.html#pipeline.code.dnn.analyzer.independent_test"><code class="docutils literal notranslate"><span class="pre">independent_test()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="../api/dnn_api.html#pipeline.code.dnn.pipeline.DnnPipeline"><code class="docutils literal notranslate"><span class="pre">DnnPipeline</span></code></a><ul>
<li class="toctree-l4"><a class="reference internal" href="../api/dnn_api.html#pipeline.code.dnn.pipeline.DnnPipeline.cfg"><code class="docutils literal notranslate"><span class="pre">DnnPipeline.cfg</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="../api/dnn_api.html#pipeline.code.dnn.pipeline.DnnPipeline.logger"><code class="docutils literal notranslate"><span class="pre">DnnPipeline.logger</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="../api/dnn_api.html#pipeline.code.dnn.pipeline.DnnPipeline.execute"><code class="docutils literal notranslate"><span class="pre">DnnPipeline.execute()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="../api/dnn_api.html#pipeline.code.dnn.pipeline.DnnPipeline.train"><code class="docutils literal notranslate"><span class="pre">DnnPipeline.train()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="../api/dnn_api.html#pipeline.code.dnn.pipeline.DnnPipeline.trans"><code class="docutils literal notranslate"><span class="pre">DnnPipeline.trans()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="../api/dnn_api.html#pipeline.code.dnn.pipeline.DnnPipeline.test"><code class="docutils literal notranslate"><span class="pre">DnnPipeline.test()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="../api/dnn_api.html#pipeline.code.dnn.pipeline.DnnPipeline._process_single_dataset"><code class="docutils literal notranslate"><span class="pre">DnnPipeline._process_single_dataset()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="../api/dnn_api.html#pipeline.code.dnn.pipeline.DnnPipeline._train_single_model"><code class="docutils literal notranslate"><span class="pre">DnnPipeline._train_single_model()</span></code></a></li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="../api/graph_walk_api.html">Graph Walk</a><ul>
<li class="toctree-l3"><a class="reference internal" href="../api/graph_walk_api.html#pipeline.code.graph_walk.run.generate_embeddings"><code class="docutils literal notranslate"><span class="pre">generate_embeddings()</span></code></a></li>
</ul>
</li>
</ul>
</li>
</ul>
<p class="caption" role="heading"><span class="caption-text">Other</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../faq.html">Frequently Asked Questions</a></li>
<li class="toctree-l1"><a class="reference internal" href="../citing.html">Citing TransBrain</a></li>
</ul>

        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../index.html">TransBrain</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="../index.html" class="icon icon-home" aria-label="Home"></a></li>
      <li class="breadcrumb-item active">Method Glossory</li>
      <li class="wy-breadcrumbs-aside">
            <a href="../_sources/methods/method_glossory.rst.txt" rel="nofollow"> View page source</a>
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  <section id="method-glossory">
<span id="glossory"></span><h1>Method Glossory<a class="headerlink" href="#method-glossory" title="Permalink to this heading"></a></h1>
<p>To better help users understand the terms involved in the TransBrain methodology, we provide a glossary with additional links, covering: <strong>Datasets</strong> and <strong>Methods or Concepts</strong>.</p>
<section id="datasets">
<h2>Datasets<a class="headerlink" href="#datasets" title="Permalink to this heading"></a></h2>
<dl class="simple glossary">
<dt id="term-AHBA-data">AHBA data<a class="headerlink" href="#term-AHBA-data" title="Permalink to this term"></a></dt><dd><p>The Allen Human Brain Atlas (AHBA) is a microarray dataset of gene expression patterns in the human brain, derived from post-mortem tissue samples. It provides detailed maps of where specific genes are expressed across different regions of the brain, helping to understand brain function and structure.</p>
<ul class="simple">
<li><p>link: <a class="reference external" href="http://human.brain-map.org">http://human.brain-map.org</a></p></li>
<li><p>article: <a class="reference external" href="https://www.nature.com/articles/nature11405">https://www.nature.com/articles/nature11405</a></p></li>
</ul>
</dd>
<dt id="term-Human-single-nucleus-data">Human single-nucleus data<a class="headerlink" href="#term-Human-single-nucleus-data" title="Permalink to this term"></a></dt><dd><p>Single-nucleus RNA sequencing (snRNA-seq) data from human brain tissue provided by the Linnarsson laboratory. This dataset includes single-nucleus transcriptomic data from multiple brain regions of four adult human donors.</p>
<ul class="simple">
<li><p>link: <a class="reference external" href="https://github.com/linnarsson-lab/adult-human-brain?tab=readme-ov-file">https://github.com/linnarsson-lab/adult-human-brain?tab=readme-ov-file</a></p></li>
<li><p>article: <a class="reference external" href="https://www.science.org/doi/abs/10.1126/science.add7046">https://www.science.org/doi/abs/10.1126/science.add7046</a></p></li>
</ul>
</dd>
<dt id="term-Mouse-spatial-transcriptomic-data">Mouse spatial transcriptomic data<a class="headerlink" href="#term-Mouse-spatial-transcriptomic-data" title="Permalink to this term"></a></dt><dd><p>Whole-brain spatial transcriptomic (ST) sequencing data from three adult male mice from Ortiz et al.’s study.</p>
<ul class="simple">
<li><p>link: <a class="reference external" href="http://molecularatlas.org">http://molecularatlas.org</a></p></li>
<li><p>article: <a class="reference external" href="https://www.science.org/doi/full/10.1126/sciadv.abb3446">https://www.science.org/doi/full/10.1126/sciadv.abb3446</a></p></li>
</ul>
</dd>
<dt id="term-Diffusion-Weighted-Imaging">Diffusion Weighted Imaging<a class="headerlink" href="#term-Diffusion-Weighted-Imaging" title="Permalink to this term"></a></dt><dd><p>Diffusion Weighted Imaging is an advanced MRI technique that maps the diffusion of water molecules along white matter fibers in the brain. By mapping the direction and density of water diffusion, it helps identify the connections between different brain regions, allowing researchers to study the brain’s structural network and its organization.</p>
</dd>
<dt id="term-Human-Connectome-Project-Young-Adult-HCP-YA">Human Connectome Project - Young Adult (HCP-YA)<a class="headerlink" href="#term-Human-Connectome-Project-Young-Adult-HCP-YA" title="Permalink to this term"></a></dt><dd><p>The Human Connectome Project - Young Adult (HCP-YA) is a comprehensive neuroimaging study aimed at mapping the human brain’s structural and functional connectivity in healthy young adults. Conducted by the WU-Minn Consortium, the project involved over 1,200 participants aged 22-35 and utilized advanced imaging techniques to provide detailed insights into brain organization.</p>
<ul class="simple">
<li><p>link <a class="reference external" href="https://www.humanconnectome.org/study/hcp-young-adult/data-releases">https://www.humanconnectome.org/study/hcp-young-adult/data-releases</a></p></li>
</ul>
</dd>
<dt id="term-Mouse-tracer-data">Mouse tracer data<a class="headerlink" href="#term-Mouse-tracer-data" title="Permalink to this term"></a></dt><dd><p>Tracer technology is a neuroscience method that uses molecular markers to map neural pathways by injecting a tracer into a brain region, which is then transported along neurons to track their connectivity.
The mouse tracer data used in our project was sourced from the Allen Mouse Connectivity Atlas.</p>
<ul class="simple">
<li><p>link <a class="reference external" href="https://connectivity.brain-map.org/">https://connectivity.brain-map.org/</a></p></li>
</ul>
</dd>
<dt id="term-Human-resting-state-fMRI">Human resting-state fMRI<a class="headerlink" href="#term-Human-resting-state-fMRI" title="Permalink to this term"></a></dt><dd><p>Resting-state fMRI is a neuroimaging technique that measures brain activity through blood oxygen level changes while the subject is at rest. It helps understand the brain’s intrinsic networks and is commonly used in studies of brain function and connectivity. The data used here were obtained from the HCP Young Adult 1200 Subjects Data Release.</p>
<ul class="simple">
<li><p>link: <a class="reference external" href="https://www.humanconnectome.org/study/hcp-young-adult/data-releases">https://www.humanconnectome.org/study/hcp-young-adult/data-releases</a></p></li>
</ul>
</dd>
<dt id="term-Mouse-resting-state-fMRI">Mouse resting-state fMRI<a class="headerlink" href="#term-Mouse-resting-state-fMRI" title="Permalink to this term"></a></dt><dd><p>The resting-state fMRI data for awake mice, made publicly available by Gozzi et al., were obtained from 10 adult male mice (aged &lt; 6 months) using a 7T Bruker MRI scanner.</p>
<ul class="simple">
<li><p>link: <a class="reference external" href="https://data.mendeley.com/datasets/np2fx99hn6/2">https://data.mendeley.com/datasets/np2fx99hn6/2</a></p></li>
<li><p>article: <a class="reference external" href="https://www.cell.com/current-biology/fulltext/S0960-9822(21)01691-2">https://www.cell.com/current-biology/fulltext/S0960-9822(21)01691-2</a></p></li>
</ul>
</dd>
<dt id="term-ABIDE-I-dataset">ABIDE I dataset<a class="headerlink" href="#term-ABIDE-I-dataset" title="Permalink to this term"></a></dt><dd><p>The ABIDE I (Autism Brain Imaging Data Exchange I) is a large-scale collection of neuroimaging dataset aimed at understanding the neural basis of Autism Spectrum Disorder (ASD).</p>
<ul class="simple">
<li><p>link <a class="reference external" href="https://fcon_1000.projects.nitrc.org/indi/abide/abide_I.html">https://fcon_1000.projects.nitrc.org/indi/abide/abide_I.html</a></p></li>
</ul>
</dd>
<dt id="term-Mouse-structural-imaging-dataset">Mouse structural imaging dataset<a class="headerlink" href="#term-Mouse-structural-imaging-dataset" title="Permalink to this term"></a></dt><dd><p>A large set of autism mouse models, composed of separate cohorts from various laboratories, and scanned at the Mouse Imaging Centre in Toronto.</p>
<ul class="simple">
<li><p>link: <a class="reference external" href="https://www.braincode.ca/content/public-data-releases#dr001">https://www.braincode.ca/content/public-data-releases#dr001</a>.</p></li>
<li><p>article: <a class="reference external" href="https://www.nature.com/articles/mp201498">https://www.nature.com/articles/mp201498</a></p></li>
</ul>
</dd>
</dl>
</section>
<section id="methods-or-concepts">
<h2>Methods or Concepts<a class="headerlink" href="#methods-or-concepts" title="Permalink to this heading"></a></h2>
<dl class="simple glossary">
<dt id="term-Microarray">Microarray<a class="headerlink" href="#term-Microarray" title="Permalink to this term"></a></dt><dd><p>Microarray technology measures gene expression levels across thousands of genes simultaneously using fixed probes on a solid surface, offering high-throughput gene expression analysis.</p>
</dd>
<dt id="term-Single-nucleus-RNA-sequencing">Single-nucleus RNA sequencing<a class="headerlink" href="#term-Single-nucleus-RNA-sequencing" title="Permalink to this term"></a></dt><dd><p>snRNA-seq analyzes gene expression at the individual cell level, offering insights into the molecular diversity of different brain cell types.</p>
</dd>
<dt id="term-Spatial-transcriptomic-data">Spatial transcriptomic data<a class="headerlink" href="#term-Spatial-transcriptomic-data" title="Permalink to this term"></a></dt><dd><p>Spatial transcriptomics combines gene expression analysis with tissue spatial context, enabling the mapping of gene activity across tissue sections while preserving their spatial organization.</p>
</dd>
<dt id="term-Brainnetome-atlas">Brainnetome atlas<a class="headerlink" href="#term-Brainnetome-atlas" title="Permalink to this term"></a></dt><dd><p>The Brainnetome Atlas is a human brain atlas constructed based on structural connectivity, which divides the brain into 246 regions</p>
<ul class="simple">
<li><p>link: <a class="reference external" href="https://atlas.brainnetome.org/">https://atlas.brainnetome.org/</a></p></li>
<li><p>article: <a class="reference external" href="https://www.science.org/doi/full/10.1126/sciadv.abb3446">https://www.science.org/doi/full/10.1126/sciadv.abb3446</a></p></li>
</ul>
</dd>
<dt id="term-Desikan-Killiany-atlas">Desikan-Killiany atlas<a class="headerlink" href="#term-Desikan-Killiany-atlas" title="Permalink to this term"></a></dt><dd><p>The Desikan-Killiany (DK) Atlas is a widely used brain atlas for human neuroimaging studies, particularly in structural MRI. It divides the brain into distinct regions based on anatomical landmarks.</p>
<ul class="simple">
<li><p>link: <a class="reference external" href="https://surfer.nmr.mgh.harvard.edu/fswiki/CorticalParcellation">https://surfer.nmr.mgh.harvard.edu/fswiki/CorticalParcellation</a></p></li>
</ul>
</dd>
<dt id="term-AAL-atlas">AAL atlas<a class="headerlink" href="#term-AAL-atlas" title="Permalink to this term"></a></dt><dd><p>Automated Anatomical Labeling (AAL) Atlas is a widely used brain atlas that provides a standardized map for labeling anatomical regions of the human brain. It divides the brain into predefined regions of interest (ROIs) based on structural MRI scans.</p>
<ul class="simple">
<li><p>link: <a class="reference external" href="https://www.gin.cnrs.fr/en/tools/aal/">https://www.gin.cnrs.fr/en/tools/aal/</a></p></li>
</ul>
</dd>
<dt id="term-Allen-Mouse-Brain-Atlas">Allen Mouse Brain Atlas<a class="headerlink" href="#term-Allen-Mouse-Brain-Atlas" title="Permalink to this term"></a></dt><dd><p>The Allen Mouse Brain Atlas CCFv3 (Common Coordinate Framework version 3) is a 3D reference brain atlas for the mouse brain. The CCFv3 space is built by aligning high-resolution brain images from multiple mice to a single reference brain, allowing for precise localization of brain regions.</p>
<ul class="simple">
<li><p>link: <a class="reference external" href="https://atlas.brain-map.org/">https://atlas.brain-map.org/</a></p></li>
<li><p>article: <a class="reference external" href="https://www.cell.com/cell/fulltext/S0092-8674(20)30402-5?dgcid=raven_jbs_aip_email">https://www.cell.com/cell/fulltext/S0092-8674(20)30402-5?dgcid=raven_jbs_aip_email</a></p></li>
</ul>
</dd>
<dt id="term-Cross-species-homology-mapping">Cross-species homology mapping<a class="headerlink" href="#term-Cross-species-homology-mapping" title="Permalink to this term"></a></dt><dd><p>Cross-species homology region mapping involves identifying conserved genomic regions between different species that share similar functions or evolutionary origins.</p>
</dd>
<dt id="term-Neural-network">Neural network<a class="headerlink" href="#term-Neural-network" title="Permalink to this term"></a></dt><dd><p>A neural network is a computational model inspired by the way biological neural networks in the brain process information. It consists of layers of interconnected nodes (neurons) that work together to solve complex tasks such as classification, regression, and pattern recognition.</p>
</dd>
<dt id="term-Embeddings">Embeddings<a class="headerlink" href="#term-Embeddings" title="Permalink to this term"></a></dt><dd><p>Embeddings represent high-dimensional data in a lower-dimensional space, while preserving important relationships and patterns. They also help capture semantic similarities, allowing for better generalization and improved performance.</p>
</dd>
<dt id="term-Gene-enrichment-analysis">Gene enrichment analysis<a class="headerlink" href="#term-Gene-enrichment-analysis" title="Permalink to this term"></a></dt><dd><p>Gene enrichment analysis identifies biological functions or pathways that are overrepresented in a specific gene set, helping to reveal the underlying molecular processes or disease mechanisms.</p>
</dd>
<dt id="term-Human-Mouse-Graph">Human-Mouse Graph<a class="headerlink" href="#term-Human-Mouse-Graph" title="Permalink to this term"></a></dt><dd><p>The Human-Mouse graph is a graph structure used to represent the similarities between brain regions in humans and mice. In this graph, brain regions from both species are represented as nodes, and the edges (connections) between nodes are weighted based on transcriptional similarity and structural connectivity.</p>
</dd>
<dt id="term-Random-Walk">Random Walk<a class="headerlink" href="#term-Random-Walk" title="Permalink to this term"></a></dt><dd><p>A random walk explores a graph to learn node embeddings based on their neighborhood structures. In TransBrain, Node2Vec algorithm was used, with the probability of moving to a neighbor depending on the graph structure and walk parameters.</p>
</dd>
<dt id="term-Dual-regression">Dual regression<a class="headerlink" href="#term-Dual-regression" title="Permalink to this term"></a></dt><dd><p>Dual regression is a method used in neuroimaging to project group-level patterns (such as brain networks derived from Independent Component Analysis, ICA) onto individual subject data. Previous work have also employed this method to translate brain phenotypes across species using  gene expression data.</p>
<ul class="simple">
<li><p>link: <a class="reference external" href="https://web.mit.edu/fsl_v5.0.10/fsl/doc/wiki/DualRegression.html">https://web.mit.edu/fsl_v5.0.10/fsl/doc/wiki/DualRegression.html</a></p></li>
<li><p>article: <a class="reference external" href="https://www.nature.com/articles/s41380-021-01298-5">https://www.nature.com/articles/s41380-021-01298-5</a></p></li>
</ul>
</dd>
<dt id="term-Phenotype">Phenotype<a class="headerlink" href="#term-Phenotype" title="Permalink to this term"></a></dt><dd><p>In neuroscience, phenotype refers to the observable characteristics or traits of the brain, including brain structure, function, and individual behavior. Phenotypic traits can include aspects such as cognitive abilities, neural connectivity, or responses to stimuli, and they are often studied to understand the biological basis of diseases or neurological disorders.</p>
</dd>
<dt id="term-Functional-gradients">Functional gradients<a class="headerlink" href="#term-Functional-gradients" title="Permalink to this term"></a></dt><dd><p>Functional gradients refer to inherent patterns of brain activity or connectivity that smoothly transition across different brain regions, reflecting the gradual shift in neural function and cognitive processes within the brain’s organization.</p>
<ul class="simple">
<li><p>article <a class="reference external" href="https://www.pnas.org/doi/abs/10.1073/pnas.1608282113">https://www.pnas.org/doi/abs/10.1073/pnas.1608282113</a></p></li>
</ul>
</dd>
<dt id="term-Co-activation-patterns-CAPs">Co-activation patterns (CAPs)<a class="headerlink" href="#term-Co-activation-patterns-CAPs" title="Permalink to this term"></a></dt><dd><p>Co-activation patterns (CAPs) are brain activity patterns where different regions are activated together, used to study functional brain networks and their collaboration during tasks or rest.</p>
</dd>
<dt id="term-Optogenetic-fMRI">Optogenetic fMRI<a class="headerlink" href="#term-Optogenetic-fMRI" title="Permalink to this term"></a></dt><dd><p>Optogenetic fMRI combines optogenetics and functional MRI to control and monitor brain activity. By using light to stimulate specific neurons that have been genetically modified to express light-sensitive proteins, researchers can map the causal relationship between neural activity and brain function in real time.</p>
</dd>
<dt id="term-Neurosynth">Neurosynth<a class="headerlink" href="#term-Neurosynth" title="Permalink to this term"></a></dt><dd><p>Neurosynth is a meta-analytical platform capable of generating keyword-based statistical functional maps that display activation patterns of specific cognitive processes or psychological terms.</p>
<ul class="simple">
<li><p>link <a class="reference external" href="https://neurosynth.org">https://neurosynth.org</a></p></li>
</ul>
</dd>
<dt id="term-Deformation-based-morphology">Deformation-based morphology<a class="headerlink" href="#term-Deformation-based-morphology" title="Permalink to this term"></a></dt><dd><p>Deformation-based morphology (DBM) is a technique used to study structural brain changes by analyzing the spatial deformation of brain images. It involves mapping brain structures between individual subjects and a common template to detect differences in brain volume.</p>
</dd>
<dt id="term-Normative-model">Normative model<a class="headerlink" href="#term-Normative-model" title="Permalink to this term"></a></dt><dd><p>Normative model is a statistical method that represents typical brain structure or function in a healthy population. It serves as a baseline or reference to compare individual data, helping identify deviations that may indicate abnormalities, such as in neurological or psychiatric disorders.</p>
</dd>
</dl>
</section>
</section>


           </div>
          </div>
          <footer><div class="rst-footer-buttons" role="navigation" aria-label="Footer">
        <a href="../atlas.html" class="btn btn-neutral float-left" title="Supported Atlases" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
        <a href="../online_mapping.html" class="btn btn-neutral float-right" title="Online Mapping" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right" aria-hidden="true"></span></a>
    </div>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright 2025, TransBrain.</p>
  </div>

  Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    provided by <a href="https://readthedocs.org">Read the Docs</a>.
   

</footer>
        </div>
      </div>
    </section>
  </div>
  <script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>