

<!DOCTYPE html>
<html class="writer-html5" lang="en">
<head>
  <meta charset="utf-8" /><meta name="viewport" content="width=device-width, initial-scale=1" />

  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>General Usage &mdash; TransBrain 0.1 documentation</title>
      <link rel="stylesheet" type="text/css" href="_static/pygments.css?v=03e43079" />
      <link rel="stylesheet" type="text/css" href="_static/css/theme.css?v=e59714d7" />

  
      <script src="_static/jquery.js?v=5d32c60e"></script>
      <script src="_static/_sphinx_javascript_frameworks_compat.js?v=2cd50e6c"></script>
      <script data-url_root="./" id="documentation_options" src="_static/documentation_options.js?v=e031e9a9"></script>
      <script src="_static/doctools.js?v=888ff710"></script>
      <script src="_static/sphinx_highlight.js?v=4825356b"></script>
      <script crossorigin="anonymous" integrity="sha256-Ae2Vz/4ePdIu6ZyI/5ZGsYnb+m0JlOmKPjt6XZ9JJkA=" src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.4/require.min.js"></script>
    <script src="_static/js/theme.js"></script>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >

          
          
          <a href="index.html" class="icon icon-home">
            TransBrain
              <img src="_static/logo_final.png" class="logo" alt="Logo"/>
          </a>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" aria-label="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <p class="caption" role="heading"><span class="caption-text">Setup &amp; Usage</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="installation.html">Installation</a><ul>
<li class="toctree-l2"><a class="reference internal" href="installation.html#installation-options">Installation Options</a></li>
<li class="toctree-l2"><a class="reference internal" href="installation.html#python-dependencies">Python Dependencies</a></li>
<li class="toctree-l2"><a class="reference internal" href="installation.html#unit-test">Unit Test</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="user.html">User Guide</a><ul>
<li class="toctree-l2"><a class="reference internal" href="user.html#id2">General Use</a></li>
<li class="toctree-l2"><a class="reference internal" href="user.html#id3">Online Mapping</a></li>
<li class="toctree-l2"><a class="reference internal" href="user.html#id4">Advanced Use</a></li>
</ul>
</li>
</ul>
<p class="caption" role="heading"><span class="caption-text">Tutorials</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="building.html">Construction Pipeline</a><ul>
<li class="toctree-l2"><a class="reference internal" href="pipeline/run_pipeline/integration.html">Transcriptional Data Integration</a><ul>
<li class="toctree-l3"><a class="reference internal" href="pipeline/run_pipeline/integration.html#Selection-of-Cross-Modality-Stable-Genes">Selection of Cross-Modality Stable Genes</a><ul>
<li class="toctree-l4"><a class="reference internal" href="pipeline/run_pipeline/integration.html#The-shared-gene-space-between-the-​​AHBA,-sn-dataset-and-​​mouse-st-dataset​​">The shared gene space between the ​​AHBA, sn dataset and ​​mouse st dataset​​</a></li>
<li class="toctree-l4"><a class="reference internal" href="pipeline/run_pipeline/integration.html#Cortical-stable-genes">Cortical stable genes</a></li>
<li class="toctree-l4"><a class="reference internal" href="pipeline/run_pipeline/integration.html#Subcortical-stable-genes">Subcortical stable genes</a></li>
<li class="toctree-l4"><a class="reference internal" href="pipeline/run_pipeline/integration.html#All-stable-genes">All stable genes</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="pipeline/run_pipeline/integration.html#Determining-Similarity-Thresholds-for-Sample-Pool-Construction">Determining Similarity Thresholds for Sample Pool Construction</a><ul>
<li class="toctree-l4"><a class="reference internal" href="pipeline/run_pipeline/integration.html#Thresholds-of-cortex">Thresholds of cortex</a></li>
<li class="toctree-l4"><a class="reference internal" href="pipeline/run_pipeline/integration.html#Thresholds-of-subcortex">Thresholds of subcortex</a></li>
<li class="toctree-l4"><a class="reference internal" href="pipeline/run_pipeline/integration.html#Thresholds-of-all">Thresholds of all</a></li>
<li class="toctree-l4"><a class="reference internal" href="pipeline/run_pipeline/integration.html#Integrating-dataset">Integrating dataset</a></li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="pipeline/run_pipeline/dnn.html">DNN Model Training</a><ul>
<li class="toctree-l3"><a class="reference internal" href="pipeline/run_pipeline/dnn.html#cortical-train">cortical train</a></li>
<li class="toctree-l3"><a class="reference internal" href="pipeline/run_pipeline/dnn.html#subcortical-train">subcortical train</a></li>
<li class="toctree-l3"><a class="reference internal" href="pipeline/run_pipeline/dnn.html#all-train">all train</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="pipeline/run_pipeline/graph_embedding.html">Graph-based Random Walk</a><ul>
<li class="toctree-l3"><a class="reference internal" href="pipeline/run_pipeline/graph_embedding.html#Define-human-and-mouse-ROIs">Define human and mouse ROIs</a></li>
<li class="toctree-l3"><a class="reference internal" href="pipeline/run_pipeline/graph_embedding.html#Generate-human-and-mouse-graph">Generate human and mouse graph</a></li>
<li class="toctree-l3"><a class="reference internal" href="pipeline/run_pipeline/graph_embedding.html#Hierachy-constraint">Hierachy constraint</a></li>
<li class="toctree-l3"><a class="reference internal" href="pipeline/run_pipeline/graph_embedding.html#Generate-graph-embeddings">Generate graph embeddings</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="methods/mapping.html">Bidirectional Mapping</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="cases.html">Example Cases</a><ul>
<li class="toctree-l2"><a class="reference internal" href="tutorials/notebooks/tr_embeddings_analysis/embeddings_analysis.html">Analyzing cross-species similarity of transcriptional embeddings</a><ul>
<li class="toctree-l3"><a class="reference internal" href="tutorials/notebooks/tr_embeddings_analysis/embeddings_analysis.html#Human-Autocorrelation-Matrix">Human Autocorrelation Matrix</a></li>
<li class="toctree-l3"><a class="reference internal" href="tutorials/notebooks/tr_embeddings_analysis/embeddings_analysis.html#Cross-species-Similarity-Matrix">Cross-species Similarity Matrix</a></li>
<li class="toctree-l3"><a class="reference internal" href="tutorials/notebooks/tr_embeddings_analysis/embeddings_analysis.html#Potential-homologous-regions'-rank-of-Cross-species-Matrix">Potential homologous regions’ rank of Cross-species Matrix</a><ul>
<li class="toctree-l4"><a class="reference internal" href="tutorials/notebooks/tr_embeddings_analysis/embeddings_analysis.html#Cortical-rank">Cortical rank</a></li>
<li class="toctree-l4"><a class="reference internal" href="tutorials/notebooks/tr_embeddings_analysis/embeddings_analysis.html#Subcortical-rank">Subcortical rank</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="tutorials/notebooks/tr_embeddings_analysis/embeddings_analysis.html#Gradients-of-TRembeddings-conserved-across-species">Gradients of TRembeddings conserved across species</a></li>
<li class="toctree-l3"><a class="reference internal" href="tutorials/notebooks/tr_embeddings_analysis/embeddings_analysis.html#Load-cortical-embeddings">Load cortical embeddings</a></li>
<li class="toctree-l3"><a class="reference internal" href="tutorials/notebooks/tr_embeddings_analysis/embeddings_analysis.html#PCA-of-Cortical-embeddings">PCA of Cortical embeddings</a></li>
<li class="toctree-l3"><a class="reference internal" href="tutorials/notebooks/tr_embeddings_analysis/embeddings_analysis.html#Visulization">Visulization</a><ul>
<li class="toctree-l4"><a class="reference internal" href="tutorials/notebooks/tr_embeddings_analysis/embeddings_analysis.html#Human-Gradient1">Human Gradient1</a></li>
<li class="toctree-l4"><a class="reference internal" href="tutorials/notebooks/tr_embeddings_analysis/embeddings_analysis.html#Human-Gradient2">Human Gradient2</a></li>
<li class="toctree-l4"><a class="reference internal" href="tutorials/notebooks/tr_embeddings_analysis/embeddings_analysis.html#Human-Gradient3">Human Gradient3</a></li>
<li class="toctree-l4"><a class="reference internal" href="tutorials/notebooks/tr_embeddings_analysis/embeddings_analysis.html#Mouse-gradient1">Mouse gradient1</a></li>
<li class="toctree-l4"><a class="reference internal" href="tutorials/notebooks/tr_embeddings_analysis/embeddings_analysis.html#Mouse-gradient2">Mouse gradient2</a></li>
<li class="toctree-l4"><a class="reference internal" href="tutorials/notebooks/tr_embeddings_analysis/embeddings_analysis.html#Mouse-gradient3">Mouse gradient3</a></li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="tutorials/notebooks/translation/gradient_spectrum/fMRI_gradients.html">Inferring the conservation of resting-fMRI gradients across species</a><ul>
<li class="toctree-l3"><a class="reference internal" href="tutorials/notebooks/translation/gradient_spectrum/fMRI_gradients.html#translating-mouse-gradient-to-human">translating mouse gradient to human</a></li>
<li class="toctree-l3"><a class="reference internal" href="tutorials/notebooks/translation/gradient_spectrum/fMRI_gradients.html#visulization">visulization</a><ul>
<li class="toctree-l4"><a class="reference internal" href="tutorials/notebooks/translation/gradient_spectrum/fMRI_gradients.html#mouse-gradient">mouse gradient</a></li>
<li class="toctree-l4"><a class="reference internal" href="tutorials/notebooks/translation/gradient_spectrum/fMRI_gradients.html#mouse-gradients-mapped-to-human">mouse gradients mapped to human</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="tutorials/notebooks/translation/gradient_spectrum/fMRI_gradients.html#human-gradient">human gradient</a></li>
<li class="toctree-l3"><a class="reference internal" href="tutorials/notebooks/translation/gradient_spectrum/fMRI_gradients.html#correlation">correlation</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="tutorials/notebooks/translation/optogenetic_annotation/optogenetic_circuits.html">Annotating optogenetic circuits</a><ul>
<li class="toctree-l3"><a class="reference internal" href="tutorials/notebooks/translation/optogenetic_annotation/optogenetic_circuits.html#mouse-opto-phenotypes">mouse opto phenotypes</a></li>
<li class="toctree-l3"><a class="reference internal" href="tutorials/notebooks/translation/optogenetic_annotation/optogenetic_circuits.html#translating-mouse-ai-stimulated-pattern-to-human">translating mouse ai stimulated pattern to human</a></li>
<li class="toctree-l3"><a class="reference internal" href="tutorials/notebooks/translation/optogenetic_annotation/optogenetic_circuits.html#translating-mouse-drn-stimulated-pattern-to-human">translating mouse drn stimulated pattern to human</a><ul>
<li class="toctree-l4"><a class="reference internal" href="tutorials/notebooks/translation/optogenetic_annotation/optogenetic_circuits.html#annotating-using-NeuroSynth.">annotating using NeuroSynth.</a></li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="tutorials/notebooks/translation/autism_mutation/gene_mutations.html">Linking gene mutations to imaging phenotype deviations in autism</a><ul>
<li class="toctree-l3"><a class="reference internal" href="tutorials/notebooks/translation/autism_mutation/gene_mutations.html#load-deviation-pattern">load deviation pattern</a></li>
<li class="toctree-l3"><a class="reference internal" href="tutorials/notebooks/translation/autism_mutation/gene_mutations.html#translating-mouse-mutation-patterns-to-human">translating mouse mutation patterns to human</a></li>
<li class="toctree-l3"><a class="reference internal" href="tutorials/notebooks/translation/autism_mutation/gene_mutations.html#individual-gene-risk-scores-based-on-mouse-model">individual gene risk scores based on mouse model</a></li>
<li class="toctree-l3"><a class="reference internal" href="tutorials/notebooks/translation/autism_mutation/gene_mutations.html#load-AHBA-gene-expression">load AHBA gene expression</a></li>
<li class="toctree-l3"><a class="reference internal" href="tutorials/notebooks/translation/autism_mutation/gene_mutations.html#individual-gene-risk-scores-based-on-AHBA">individual gene risk scores based on AHBA</a></li>
<li class="toctree-l3"><a class="reference internal" href="tutorials/notebooks/translation/autism_mutation/gene_mutations.html#correlate-risk-scores">correlate risk scores</a></li>
<li class="toctree-l3"><a class="reference internal" href="tutorials/notebooks/translation/autism_mutation/gene_mutations.html#correlate-risk-scores-in-specific-gene">correlate risk scores in specific gene</a></li>
<li class="toctree-l3"><a class="reference internal" href="tutorials/notebooks/translation/autism_mutation/gene_mutations.html#correlating-sythetic-mutation-pattern-with-AHBA-gene-expression">correlating sythetic mutation pattern with AHBA gene expression</a></li>
<li class="toctree-l3"><a class="reference internal" href="tutorials/notebooks/translation/autism_mutation/gene_mutations.html#load-behavior-data">load behavior data</a></li>
</ul>
</li>
</ul>
</li>
</ul>
<p class="caption" role="heading"><span class="caption-text">API</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="api/utils.html">Utils</a><ul>
<li class="toctree-l2"><a class="reference internal" href="api/data_process_api.html">Transcriptional Data Process</a><ul>
<li class="toctree-l3"><a class="reference internal" href="api/data_process_api.html#pipeline.code.dpreprocess.sn_.Preprocess_Sn"><code class="docutils literal notranslate"><span class="pre">Preprocess_Sn</span></code></a><ul>
<li class="toctree-l4"><a class="reference internal" href="api/data_process_api.html#pipeline.code.dpreprocess.sn_.Preprocess_Sn.dataset_qc"><code class="docutils literal notranslate"><span class="pre">Preprocess_Sn.dataset_qc()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="api/data_process_api.html#pipeline.code.dpreprocess.sn_.Preprocess_Sn.dataset_normalized"><code class="docutils literal notranslate"><span class="pre">Preprocess_Sn.dataset_normalized()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="api/data_process_api.html#pipeline.code.dpreprocess.sn_.Preprocess_Sn.dataset_common_genes"><code class="docutils literal notranslate"><span class="pre">Preprocess_Sn.dataset_common_genes()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="api/data_process_api.html#pipeline.code.dpreprocess.sn_.Preprocess_Sn.dataset_batch_correct"><code class="docutils literal notranslate"><span class="pre">Preprocess_Sn.dataset_batch_correct()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="api/data_process_api.html#pipeline.code.dpreprocess.sn_.Preprocess_Sn.dataset_smooth"><code class="docutils literal notranslate"><span class="pre">Preprocess_Sn.dataset_smooth()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="api/data_process_api.html#pipeline.code.dpreprocess.sn_.Preprocess_Sn.dataset_scaled"><code class="docutils literal notranslate"><span class="pre">Preprocess_Sn.dataset_scaled()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="api/data_process_api.html#pipeline.code.dpreprocess.sn_.Preprocess_Sn.preprocess_pipeline"><code class="docutils literal notranslate"><span class="pre">Preprocess_Sn.preprocess_pipeline()</span></code></a></li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="api/data_integration_api.html">Data Integration</a><ul>
<li class="toctree-l3"><a class="reference internal" href="api/data_integration_api.html#pipeline.code.dintegrate.utils.stable_gene_filter"><code class="docutils literal notranslate"><span class="pre">stable_gene_filter()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="api/data_integration_api.html#pipeline.code.dintegrate.utils.calculate_thresholds"><code class="docutils literal notranslate"><span class="pre">calculate_thresholds()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="api/data_integration_api.html#pipeline.code.dintegrate.main.DataIntegrator"><code class="docutils literal notranslate"><span class="pre">DataIntegrator</span></code></a><ul>
<li class="toctree-l4"><a class="reference internal" href="api/data_integration_api.html#pipeline.code.dintegrate.main.DataIntegrator.__init__"><code class="docutils literal notranslate"><span class="pre">DataIntegrator.__init__()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="api/data_integration_api.html#pipeline.code.dintegrate.main.DataIntegrator.run"><code class="docutils literal notranslate"><span class="pre">DataIntegrator.run()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="api/data_integration_api.html#pipeline.code.dintegrate.main.DataIntegrator.pool_construction"><code class="docutils literal notranslate"><span class="pre">DataIntegrator.pool_construction()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="api/data_integration_api.html#pipeline.code.dintegrate.main.DataIntegrator.clean_pool"><code class="docutils literal notranslate"><span class="pre">DataIntegrator.clean_pool()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="api/data_integration_api.html#pipeline.code.dintegrate.main.DataIntegrator.weighted_average"><code class="docutils literal notranslate"><span class="pre">DataIntegrator.weighted_average()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="api/data_integration_api.html#pipeline.code.dintegrate.main.DataIntegrator.downsample_data"><code class="docutils literal notranslate"><span class="pre">DataIntegrator.downsample_data()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="api/data_integration_api.html#pipeline.code.dintegrate.main.DataIntegrator.generate_integrated_anndata"><code class="docutils literal notranslate"><span class="pre">DataIntegrator.generate_integrated_anndata()</span></code></a></li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="api/dnn_api.html">DNN</a><ul>
<li class="toctree-l3"><a class="reference internal" href="api/dnn_api.html#pipeline.code.dnn.analyzer.train_test_split"><code class="docutils literal notranslate"><span class="pre">train_test_split()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="api/dnn_api.html#pipeline.code.dnn.analyzer.calculate_rank"><code class="docutils literal notranslate"><span class="pre">calculate_rank()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="api/dnn_api.html#pipeline.code.dnn.analyzer.transform_human_mouse"><code class="docutils literal notranslate"><span class="pre">transform_human_mouse()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="api/dnn_api.html#pipeline.code.dnn.analyzer.independent_test"><code class="docutils literal notranslate"><span class="pre">independent_test()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="api/dnn_api.html#pipeline.code.dnn.pipeline.DnnPipeline"><code class="docutils literal notranslate"><span class="pre">DnnPipeline</span></code></a><ul>
<li class="toctree-l4"><a class="reference internal" href="api/dnn_api.html#pipeline.code.dnn.pipeline.DnnPipeline.cfg"><code class="docutils literal notranslate"><span class="pre">DnnPipeline.cfg</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="api/dnn_api.html#pipeline.code.dnn.pipeline.DnnPipeline.logger"><code class="docutils literal notranslate"><span class="pre">DnnPipeline.logger</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="api/dnn_api.html#pipeline.code.dnn.pipeline.DnnPipeline.execute"><code class="docutils literal notranslate"><span class="pre">DnnPipeline.execute()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="api/dnn_api.html#pipeline.code.dnn.pipeline.DnnPipeline.train"><code class="docutils literal notranslate"><span class="pre">DnnPipeline.train()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="api/dnn_api.html#pipeline.code.dnn.pipeline.DnnPipeline.trans"><code class="docutils literal notranslate"><span class="pre">DnnPipeline.trans()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="api/dnn_api.html#pipeline.code.dnn.pipeline.DnnPipeline.test"><code class="docutils literal notranslate"><span class="pre">DnnPipeline.test()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="api/dnn_api.html#pipeline.code.dnn.pipeline.DnnPipeline._process_single_dataset"><code class="docutils literal notranslate"><span class="pre">DnnPipeline._process_single_dataset()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="api/dnn_api.html#pipeline.code.dnn.pipeline.DnnPipeline._train_single_model"><code class="docutils literal notranslate"><span class="pre">DnnPipeline._train_single_model()</span></code></a></li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="api/graph_walk_api.html">Graph Walk</a><ul>
<li class="toctree-l3"><a class="reference internal" href="api/graph_walk_api.html#pipeline.code.graph_walk.run.generate_embeddings"><code class="docutils literal notranslate"><span class="pre">generate_embeddings()</span></code></a></li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="api/external_api.html">API Reference</a><ul>
<li class="toctree-l2"><a class="reference internal" href="api/fetch_api.html">Fetching Atlases</a><ul>
<li class="toctree-l3"><a class="reference internal" href="api/fetch_api.html#transbrain.atlas.fetch_human_atlas"><code class="docutils literal notranslate"><span class="pre">fetch_human_atlas()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="api/fetch_api.html#transbrain.atlas.fetch_mouse_atlas"><code class="docutils literal notranslate"><span class="pre">fetch_mouse_atlas()</span></code></a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="api/base_api.html">Base Functions</a><ul>
<li class="toctree-l3"><a class="reference internal" href="api/base_api.html#transbrain.base.get_region_phenotypes"><code class="docutils literal notranslate"><span class="pre">get_region_phenotypes()</span></code></a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="api/mapping_api.html">Mapping Functions</a><ul>
<li class="toctree-l3"><a class="reference internal" href="api/mapping_api.html#transbrain.trans.SpeciesTrans"><code class="docutils literal notranslate"><span class="pre">SpeciesTrans</span></code></a><ul>
<li class="toctree-l4"><a class="reference internal" href="api/mapping_api.html#transbrain.trans.SpeciesTrans.atlas_type"><code class="docutils literal notranslate"><span class="pre">SpeciesTrans.atlas_type</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="api/mapping_api.html#transbrain.trans.SpeciesTrans.regions"><code class="docutils literal notranslate"><span class="pre">SpeciesTrans.regions</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="api/mapping_api.html#transbrain.trans.SpeciesTrans.embeddings"><code class="docutils literal notranslate"><span class="pre">SpeciesTrans.embeddings</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="api/mapping_api.html#transbrain.trans.SpeciesTrans._load_embeddings"><code class="docutils literal notranslate"><span class="pre">SpeciesTrans._load_embeddings()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="api/mapping_api.html#transbrain.trans.SpeciesTrans._dual_mapping"><code class="docutils literal notranslate"><span class="pre">SpeciesTrans._dual_mapping()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="api/mapping_api.html#transbrain.trans.SpeciesTrans._translate"><code class="docutils literal notranslate"><span class="pre">SpeciesTrans._translate()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="api/mapping_api.html#transbrain.trans.SpeciesTrans.mouse_to_human"><code class="docutils literal notranslate"><span class="pre">SpeciesTrans.mouse_to_human()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="api/mapping_api.html#transbrain.trans.SpeciesTrans.human_to_mouse"><code class="docutils literal notranslate"><span class="pre">SpeciesTrans.human_to_mouse()</span></code></a></li>
</ul>
</li>
</ul>
</li>
</ul>
</li>
</ul>
<p class="caption" role="heading"><span class="caption-text">Other</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="faq.html">Frequently Asked Questions</a></li>
<li class="toctree-l1"><a class="reference internal" href="citing.html">Citing TransBrain</a></li>
</ul>

        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="index.html">TransBrain</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="index.html" class="icon icon-home" aria-label="Home"></a></li>
      <li class="breadcrumb-item active">General Usage</li>
      <li class="wy-breadcrumbs-aside">
            <a href="_sources/general_use.rst.txt" rel="nofollow"> View page source</a>
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  <section id="general-usage">
<span id="general-use"></span><h1>General Usage<a class="headerlink" href="#general-usage" title="Permalink to this heading"></a></h1>
<ul class="simple">
<li><p>Here is a brief demonstration on how to use TransBrain. Before getting started, please <a class="reference internal" href="installation.html"><span class="doc">install</span></a> TransBrain and run <code class="docutils literal notranslate"><span class="pre">test_transbrain.py</span></code> to ensure successful installation.</p></li>
<li><p>Refer to this <a class="reference internal" href="tests/basic_usage.html"><span class="doc">basic usage notebook</span></a> for a quick introduction to TransBrain’s mapping workflow.</p></li>
<li><p>Please see detailed documentation in <a class="reference internal" href="api/external_api.html#api"><span class="std std-ref">API Reference</span></a> and <a class="reference internal" href="cases.html#tutorial-section"><span class="std std-ref">Example Cases</span></a>.</p></li>
</ul>
<hr class="docutils" />
<section id="supported-atlases">
<h2>Supported Atlases<a class="headerlink" href="#supported-atlases" title="Permalink to this heading"></a></h2>
<ul class="simple">
<li><p>We provide several example data in the <code class="docutils literal notranslate"><span class="pre">exampledata</span></code> directory of the our <a class="reference external" href="https://github.com/ibpshangzheng/transbrain/tree/main/transbrain/exampledata">GitHub repository</a>.</p></li>
<li><p>You can download and check these files to learn the required input format. <strong>Note</strong> that when replacing with your own data, please <strong>strictly follow</strong> the format and region order in the provided template file. Mapping will <strong>fail</strong> if the structure is incorrect.</p></li>
<li><p>For detailed atlases information, please refer to our <a class="reference external" href="https://www.biorxiv.org/content/10.1101/2025.01.27.635016v1">paper</a> and <a class="reference external" href="https://github.com/ibpshangzheng/transbrain/tree/main/transbrain/atlas">transbrain/atlas</a></p></li>
<li><p>Support for additional atlases will be expanded in future updates.</p></li>
</ul>
<section id="human-brain-atlases">
<h3>Human Brain Atlases<a class="headerlink" href="#human-brain-atlases" title="Permalink to this heading"></a></h3>
<p>We currently provide the following the options.</p>
<ul class="simple">
<li><p><a class="reference external" href="https://atlas.brainnetome.org/">BN (Brainnetome Atlas)</a></p></li>
</ul>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>The naming of regions in the Brainnetome (BN) atlas are defined based on the anatomical locations from <a class="reference external" href="https://en.wikipedia.org/wiki/Brodmann_area">Brodmann atlas</a>. You can check the correspondence in the BN website or in this <a class="reference external" href="https://github.com/ibpshangzheng/Transbrain/blob/main/transbrain/atlas/BNA_subregions.xlsx">table</a> to help understand. For subcortical regions, we adopted a <a class="reference external" href="https://github.com/ibpshangzheng/transbrain/tree/main/transbrain/atlas">hybrid approach (22 ROIs)</a> that integrates the Brainnetome Atlas, the <a class="reference external" href="https://community.brain-map.org/t/allen-human-reference-atlas-3d-2020-new/405">Allen Brain Atlas</a>, and <a class="reference external" href="https://www.sciencedirect.com/science/article/abs/pii/S1053811913001237?via%3Dihub">public manual delineations</a>.</p>
</div>
<ul class="simple">
<li><p><a class="reference external" href="https://surfer.nmr.mgh.harvard.edu/fswiki/CorticalParcellation">DK (Desikan-Killiany Atlas)</a></p></li>
<li><p><a class="reference external" href="https://www.gin.cnrs.fr/en/tools/aal/">AAL (Automated Anatomical Labeling)</a></p></li>
</ul>
</section>
<section id="mouse-brain-atlas">
<h3>Mouse Brain Atlas<a class="headerlink" href="#mouse-brain-atlas" title="Permalink to this heading"></a></h3>
<ul class="simple">
<li><p><a class="reference external" href="https://atlas.brain-map.org/">CCFv3 (Allen Mouse Common Coordinate Framework v3)</a></p></li>
</ul>
</section>
</section>
<hr class="docutils" />
<section id="step-1-initialization">
<h2>Step 1: Initialization<a class="headerlink" href="#step-1-initialization" title="Permalink to this heading"></a></h2>
<p>Before starting the mapping, you need to initialize TransBrain by creating an instance of the <code class="docutils literal notranslate"><span class="pre">SpeciesTrans</span></code> class.</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span><span class="w"> </span><span class="nn">pandas</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">pd</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">transbrain</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">tb</span>

<span class="c1"># Initialize TransBrain for BN atlas</span>
<span class="n">Transformer</span> <span class="o">=</span> <span class="n">tb</span><span class="o">.</span><span class="n">trans</span><span class="o">.</span><span class="n">SpeciesTrans</span><span class="p">(</span><span class="s1">&#39;bn&#39;</span><span class="p">)</span>
</pre></div>
</div>
</section>
<hr class="docutils" />
<section id="step-2-prepare-input-data">
<h2>Step 2: Prepare input data<a class="headerlink" href="#step-2-prepare-input-data" title="Permalink to this heading"></a></h2>
<p>There are two ways to input data. One is that you already have ROI-level phenotype data as <code class="docutils literal notranslate"><span class="pre">CSV</span></code> table which follows the format and region order in the <a class="reference external" href="https://github.com/ibpshangzheng/transbrain/tree/main/transbrain/exampledata">provided template file</a>. The DataFrame contains two columns: <code class="docutils literal notranslate"><span class="pre">['Anatomical</span> <span class="pre">Name',</span> <span class="pre">'Phenotype']</span></code>.</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1"># Example mouse data</span>
<span class="n">mouse_phenotype</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="s1">&#39;/transbrain/exampledata/mouse/mouse_all_example_data.csv&#39;</span><span class="p">,</span><span class="n">index_col</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span>
</pre></div>
</div>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1"># Example human data</span>
<span class="n">human_phenotype</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="s1">&#39;/transbrain/exampledata/human/bn/human_bn_all_example_data.csv&#39;</span><span class="p">,</span><span class="n">index_col</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span>
</pre></div>
</div>
<p>Or if you source data is volumetric data in <code class="docutils literal notranslate"><span class="pre">.nii</span></code> or <code class="docutils literal notranslate"><span class="pre">.nii.gz</span></code> format that has been aligned to the atlas space required by TransBrain. You can use the following function to extract ROI-level data.</p>
<section id="fetch-atlas-in-transbrain">
<h3>Fetch atlas in TransBrain<a class="headerlink" href="#fetch-atlas-in-transbrain" title="Permalink to this heading"></a></h3>
<p>You can fetch the Human or Mouse atlas using the functions provided in the <cite>atlas</cite> module of TransBrain. These functions return both the atlas image and detailed region information for further mapping.</p>
<ul class="simple">
<li><p><code class="docutils literal notranslate"><span class="pre">atlas_type</span></code> can be choose from <code class="docutils literal notranslate"><span class="pre">['bn',</span> <span class="pre">'dk',</span> <span class="pre">'aal']</span></code>.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">region_type</span></code> can be set to <code class="docutils literal notranslate"><span class="pre">'cortex'</span></code>, <code class="docutils literal notranslate"><span class="pre">'subcortex'</span></code>, or <code class="docutils literal notranslate"><span class="pre">'all'</span></code> (the entire brain).</p></li>
<li><p>returns a <code class="docutils literal notranslate"><span class="pre">dictionary</span></code> containing the following keys: <code class="docutils literal notranslate"><span class="pre">[atlas,</span> <span class="pre">atlas_data,</span> <span class="pre">region_info,</span> <span class="pre">info_table]</span></code></p></li>
</ul>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1">#fetch human atlas</span>
<span class="n">human_atlas</span> <span class="o">=</span> <span class="n">tb</span><span class="o">.</span><span class="n">atlas</span><span class="o">.</span><span class="n">fetch_human_atlas</span><span class="p">(</span><span class="n">atlas_type</span><span class="o">=</span><span class="s1">&#39;bn&#39;</span><span class="p">,</span><span class="n">region_type</span><span class="o">=</span><span class="s1">&#39;cortex&#39;</span><span class="p">)</span>
</pre></div>
</div>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1">#fetch mouse atlas</span>
<span class="n">mouse_atlas</span> <span class="o">=</span> <span class="n">tb</span><span class="o">.</span><span class="n">atlas</span><span class="o">.</span><span class="n">fetch_mouse_atlas</span><span class="p">(</span><span class="n">region_type</span><span class="o">=</span><span class="s1">&#39;all&#39;</span><span class="p">)</span>
</pre></div>
</div>
</section>
<section id="get-phenotypes-from-image">
<h3>Get phenotypes from image<a class="headerlink" href="#get-phenotypes-from-image" title="Permalink to this heading"></a></h3>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1"># Get phenotypes in Human atlas used in TransBrain</span>
<span class="n">phenotype_nii_path</span> <span class="o">=</span> <span class="s1">&#39;/transbrain/exampledata/human/human_example_phenotype_data.nii.gz&#39;</span>
<span class="n">human_phenptype_extracted</span> <span class="o">=</span> <span class="n">tb</span><span class="o">.</span><span class="n">base</span><span class="o">.</span><span class="n">get_region_phenotypes</span><span class="p">(</span><span class="n">phenotype_nii_path</span><span class="p">,</span> <span class="n">atlas_dict</span> <span class="o">=</span> <span class="n">human_atlas</span><span class="p">)</span>
</pre></div>
</div>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1"># Get phenotypes in Mouse atlas used in TransBrain</span>
<span class="n">phenotype_nii_path</span> <span class="o">=</span> <span class="p">(</span><span class="s1">&#39;/transbrain/exampledata/mouse/mouse_example_phenotype_data.nii.gz&#39;</span><span class="p">)</span>
<span class="n">mouse_phenptype_extracted</span> <span class="o">=</span> <span class="n">tb</span><span class="o">.</span><span class="n">base</span><span class="o">.</span><span class="n">get_region_phenotypes</span><span class="p">(</span><span class="n">phenotype_nii_path</span><span class="p">,</span> <span class="n">atlas_dict</span> <span class="o">=</span> <span class="n">mouse_atlas</span><span class="p">)</span>
</pre></div>
</div>
</section>
</section>
<hr class="docutils" />
<section id="step-3-mapping">
<h2>Step 3: Mapping<a class="headerlink" href="#step-3-mapping" title="Permalink to this heading"></a></h2>
<section id="mouse-to-human-example">
<h3>Mouse to Human Example<a class="headerlink" href="#mouse-to-human-example" title="Permalink to this heading"></a></h3>
<p>This function supports several optional parameters:</p>
<ul class="simple">
<li><p><code class="docutils literal notranslate"><span class="pre">region_type</span></code> can be set to <code class="docutils literal notranslate"><span class="pre">'cortex'</span></code>, <code class="docutils literal notranslate"><span class="pre">'subcortex'</span></code>, or <code class="docutils literal notranslate"><span class="pre">'all'</span></code> (the entire brain).</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">normalize</span></code> determines whether to normalize the input data before mapping. Default is <code class="docutils literal notranslate"><span class="pre">True</span></code>, if your data has already been normalized, you can set it to <code class="docutils literal notranslate"><span class="pre">False</span></code>.</p></li>
</ul>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1"># Example from mouse to human</span>
<span class="n">mouse_phenotype_in_human</span> <span class="o">=</span> <span class="n">Transformer</span><span class="o">.</span><span class="n">mouse_to_human</span><span class="p">(</span>
    <span class="n">mouse_phenotype</span><span class="p">,</span>
    <span class="n">region_type</span><span class="o">=</span><span class="s1">&#39;all&#39;</span><span class="p">,</span>
    <span class="n">normalize</span><span class="o">=</span><span class="kc">True</span>
    <span class="p">)</span>
</pre></div>
</div>
</section>
<section id="human-to-mouse-example">
<h3>Human to Mouse Example<a class="headerlink" href="#human-to-mouse-example" title="Permalink to this heading"></a></h3>
<p>Mapping from human to mouse follows similar usage and requirements, but in the opposite direction.</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1"># Example from human to mouse</span>
<span class="n">human_phenotype_in_mouse</span> <span class="o">=</span> <span class="n">Transformer</span><span class="o">.</span><span class="n">human_to_mouse</span><span class="p">(</span>
    <span class="n">human_phenotype</span><span class="p">,</span>
    <span class="n">region_type</span><span class="o">=</span><span class="s1">&#39;all&#39;</span><span class="p">,</span>
    <span class="n">normalize</span><span class="o">=</span><span class="kc">True</span>
    <span class="p">)</span>
</pre></div>
</div>
</section>
</section>
<hr class="docutils" />
<section id="get-graph-embeddings">
<h2>Get Graph Embeddings<a class="headerlink" href="#get-graph-embeddings" title="Permalink to this heading"></a></h2>
<ul class="simple">
<li><p>Load the graph embedding matrix obtained from the construction progress, which serves as the foundation for training dual-regression mapping model.</p></li>
<li><p>This step is not required if you only want to use our precomputed matrices for mapping, as it has already been integrated into the function above.</p></li>
</ul>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1"># Get graph embeddings for BN atlas</span>
<span class="n">Transformer</span> <span class="o">=</span> <span class="n">tb</span><span class="o">.</span><span class="n">trans</span><span class="o">.</span><span class="n">SpeciesTrans</span><span class="p">(</span><span class="s1">&#39;bn&#39;</span><span class="p">)</span>
<span class="n">Human_Mouse_embedding_bn</span> <span class="o">=</span> <span class="n">Transformer</span><span class="o">.</span><span class="n">_load_embeddings</span><span class="p">()</span>
</pre></div>
</div>
</section>
</section>


           </div>
          </div>
          <footer>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright 2025, TransBrain Developers.</p>
  </div>

  Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    provided by <a href="https://readthedocs.org">Read the Docs</a>.
   

</footer>
        </div>
      </div>
    </section>
  </div>
  <script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>